// @troykelly/openclaw-projects/dist/api-client.js
/**
 * HTTP API client for openclaw-projects backend.
 * Handles authentication, request/response formatting, error handling,
 * retry logic with exponential backoff, and timeout handling.
 */
import { createLogger } from './logger.js';
/**
 * Sleep for a specified duration.
 */
function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
/**
 * Generate a unique request ID for tracing.
 */
function generateRequestId() {
    return crypto.randomUUID();
}
/**
 * Calculate retry delay with exponential backoff and jitter.
 */
function calculateRetryDelay(attempt, baseDelay = 1000, maxDelay = 10000) {
    // Exponential backoff: 1s, 2s, 4s, 8s...
    const exponentialDelay = baseDelay * 2 ** attempt;
    // Add jitter (±25%) to prevent thundering herd
    const jitter = exponentialDelay * (0.75 + Math.random() * 0.5);
    return Math.min(jitter, maxDelay);
}
/**
 * Check if an error is retryable.
 */
function isRetryableStatus(status) {
    // Retry on 5xx server errors, 429 rate-limited, and network errors (status 0)
    return status === 0 || status === 429 || (status >= 500 && status < 600);
}
/**
 * Map HTTP status to error code.
 */
function getErrorCode(status) {
    if (status === 401 || status === 403)
        return 'AUTH_ERROR';
    if (status === 404)
        return 'NOT_FOUND';
    if (status === 429)
        return 'RATE_LIMITED';
    if (status >= 500)
        return 'SERVER_ERROR';
    if (status === 0)
        return 'NETWORK_ERROR';
    return 'CLIENT_ERROR';
}
/**
 * HTTP API client for the openclaw-projects backend.
 */
export class ApiClient {
    baseUrl;
    apiKey;
    logger;
    timeout;
    maxRetries;
    constructor(options) {
        // Ensure URL doesn't have trailing slash
        this.baseUrl = options.config.apiUrl.replace(/\/$/, '');
        this.apiKey = options.config.apiKey;
        this.logger = options.logger ?? createLogger('api-client');
        this.timeout = options.config.timeout;
        this.maxRetries = options.config.maxRetries;
    }
    /**
     * Makes an authenticated request to the API with retry logic.
     */
    async request(method, path, body, options) {
        const url = `${this.baseUrl}${path}`;
        const requestId = generateRequestId();
        const timeout = options?.timeout ?? this.timeout;
        let lastError = null;
        for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
            if (attempt > 0) {
                // For 429 responses, prefer the server-specified Retry-After delay
                const delay = lastError?.status === 429 && lastError.retryAfter ? lastError.retryAfter * 1000 : calculateRetryDelay(attempt - 1);
                this.logger.debug(`Retrying request (attempt ${attempt + 1}/${this.maxRetries + 1})`, {
                    path,
                    delay,
                });
                await sleep(delay);
            }
            try {
                const result = await this.executeRequest(method, url, body, requestId, options, timeout);
                if (result.success) {
                    return result;
                }
                lastError = result.error;
                if (!isRetryableStatus(result.error.status)) {
                    return result;
                }
            }
            catch (error) {
                // Handle timeout or network errors
                const isAbort = error instanceof Error && error.name === 'AbortError';
                lastError = {
                    status: 0,
                    message: error instanceof Error ? error.message : 'Unknown error',
                    code: isAbort ? 'TIMEOUT' : 'NETWORK_ERROR',
                };
                this.logger.error('API request failed', {
                    method,
                    path,
                    requestId,
                    error: lastError.message,
                });
                // If the caller's signal triggered the abort, stop immediately —
                // retrying a deliberately-cancelled request wastes resources (#2221).
                if (isAbort && options?.signal?.aborted) {
                    return { success: false, error: lastError };
                }
            }
        }
        return {
            success: false,
            error: lastError ?? {
                status: 0,
                message: 'Request failed after retries',
                code: 'NETWORK_ERROR',
            },
        };
    }
    /**
     * Execute a single request with timeout handling.
     */
    async executeRequest(method, url, body, requestId, options, timeout) {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), timeout);
        // Forward external abort signal to our internal controller (#2221).
        // When a hook timeout fires, this cancels the in-flight HTTP request
        // instead of letting it run to the API client's own (longer) timeout.
        if (options?.signal) {
            if (options.signal.aborted) {
                controller.abort();
            }
            else {
                options.signal.addEventListener('abort', () => controller.abort(), { once: true });
            }
        }
        try {
            const headers = {
                'X-Request-Id': requestId,
            };
            // Only set Content-Type for requests that have a body.
            // Fastify rejects Content-Type: application/json with an empty body.
            if (body !== undefined) {
                headers['Content-Type'] = 'application/json';
            }
            if (this.apiKey) {
                headers.Authorization = `Bearer ${this.apiKey}`;
            }
            if (options?.user_id) {
                headers['X-Agent-Id'] = options.user_id;
            }
            if (options?.user_email) {
                headers['X-User-Email'] = options.user_email;
            }
            // Namespace scoping — ensures by-ID operations target the correct namespace (#1760)
            if (options?.namespace) {
                headers['X-Namespace'] = options.namespace;
            }
            // Mark request as coming from an agent for privacy filtering
            if (options?.isAgent) {
                headers['X-OpenClaw-Agent'] = options.user_id || 'plugin-agent';
            }
            const response = await fetch(url, {
                method,
                headers,
                body: body ? JSON.stringify(body) : undefined,
                signal: controller.signal,
            });
            clearTimeout(timeoutId);
            if (!response.ok) {
                const errorBody = await response.json().catch(() => ({}));
                const retryAfter = response.headers.get('Retry-After');
                return {
                    success: false,
                    error: {
                        status: response.status,
                        message: errorBody.message || response.statusText,
                        code: getErrorCode(response.status),
                        details: errorBody.details,
                        retryAfter: retryAfter ? Number.parseInt(retryAfter, 10) : undefined,
                    },
                };
            }
            // Handle no content
            if (response.status === 204) {
                return { success: true, data: undefined };
            }
            const data = (await response.json());
            return { success: true, data };
        }
        catch (error) {
            clearTimeout(timeoutId);
            throw error;
        }
    }
    /** GET request */
    async get(path, options) {
        return this.request('GET', path, undefined, options);
    }
    /** POST request */
    async post(path, body, options) {
        return this.request('POST', path, body, options);
    }
    /** PUT request */
    async put(path, body, options) {
        return this.request('PUT', path, body, options);
    }
    /** PATCH request */
    async patch(path, body, options) {
        return this.request('PATCH', path, body, options);
    }
    /** DELETE request */
    async delete(path, options) {
        return this.request('DELETE', path, undefined, options);
    }
    /**
     * Health check endpoint.
     */
    async healthCheck() {
        const start = Date.now();
        const result = await this.get('/health');
        const latencyMs = Date.now() - start;
        return {
            healthy: result.success,
            latencyMs,
        };
    }
}
/**
 * Creates a new API client instance.
 */
export function createApiClient(options) {
    return new ApiClient(options);
}
