// @troykelly/openclaw-projects/dist/cli.js
/**
 * CLI command handlers for the OpenClaw plugin.
 * These handlers can be registered with an OpenClaw CLI API.
 */
/** Default limit for recall command */
const DEFAULT_RECALL_LIMIT = 5;
/** User scoping mode descriptions */
const SCOPING_DESCRIPTIONS = {
    agent: 'Memories are scoped by agent ID. One agent = one user. All sessions for this agent share the same memories.',
    identity: 'Memories are scoped by canonical identity from identity links. Useful when a single user may interact through multiple agents.',
    session: 'Memories are scoped by full session key. Most isolated mode - each session has separate memories.',
};
/**
 * Creates the status command handler.
 * Tests API connectivity and displays health information.
 */
export function createStatusCommand(ctx) {
    const { client, logger, config } = ctx;
    return async () => {
        logger.info('CLI status command invoked', { user_id: ctx.user_id });
        try {
            const response = await client.healthCheck();
            if (response.healthy) {
                return {
                    success: true,
                    message: `API is healthy (latency: ${response.latencyMs}ms)`,
                    data: {
                        healthy: true,
                        latencyMs: response.latencyMs,
                        apiUrl: config.apiUrl,
                    },
                };
            }
            else {
                return {
                    success: false,
                    message: 'API is unhealthy',
                    data: {
                        healthy: false,
                        latencyMs: response.latencyMs,
                        apiUrl: config.apiUrl,
                    },
                };
            }
        }
        catch (error) {
            logger.error('CLI status command error', {
                error: error instanceof Error ? error.message : String(error),
            });
            return {
                success: false,
                message: `Connection error: ${error instanceof Error ? error.message : 'Unknown error'}`,
            };
        }
    };
}
/**
 * Creates the users command handler.
 * Displays current user scoping configuration.
 */
export function createUsersCommand(ctx) {
    const { logger, config, user_id } = ctx;
    return async () => {
        logger.info('CLI users command invoked', { user_id });
        const scopingMode = config.userScoping ?? 'agent';
        const description = SCOPING_DESCRIPTIONS[scopingMode] ?? 'Unknown scoping mode';
        return {
            success: true,
            message: `User scoping mode: ${scopingMode}`,
            data: {
                scopingMode,
                description,
                currentUserId: user_id,
            },
        };
    };
}
/**
 * Creates the recall command handler.
 * Searches memories using the /api/memories/search endpoint.
 */
export function createRecallCommand(ctx) {
    const { client, logger, config, user_id } = ctx;
    return async (options) => {
        const { query, limit = config.maxRecallMemories ?? DEFAULT_RECALL_LIMIT } = options;
        // Log without query content for privacy
        logger.info('CLI recall command invoked', {
            user_id,
            queryLength: query?.length ?? 0,
            limit,
        });
        if (!query || query.trim() === '') {
            return {
                success: false,
                message: 'Error: query is required',
            };
        }
        try {
            const queryParams = new URLSearchParams({
                q: query.substring(0, 500), // Limit query length
                limit: String(limit),
            });
            const response = await client.get(`/memories/search?${queryParams.toString()}`, { user_id });
            if (!response.success) {
                logger.error('CLI recall command API error', {
                    user_id,
                    status: response.error.status,
                    code: response.error.code,
                });
                return {
                    success: false,
                    message: `API error: ${response.error.message}`,
                };
            }
            // Map API field names (results/type) to CLI interface (memories/category)
            const memories = (response.data.results ?? []).map((m) => ({
                id: m.id,
                content: m.content,
                score: m.score ?? 0,
                category: m.type,
            }));
            return {
                success: true,
                message: `Found ${memories.length} memories`,
                data: {
                    memories,
                    query,
                    limit,
                },
            };
        }
        catch (error) {
            logger.error('CLI recall command error', {
                user_id,
                error: error instanceof Error ? error.message : String(error),
            });
            return {
                success: false,
                message: `Error: ${error instanceof Error ? error.message : 'Unknown error'}`,
            };
        }
    };
}
/**
 * Creates all CLI command handlers.
 */
export function createCliCommands(ctx) {
    return {
        status: createStatusCommand(ctx),
        users: createUsersCommand(ctx),
        recall: createRecallCommand(ctx),
    };
}
