// @troykelly/openclaw-projects/dist/context.js
/**
 * Context extraction utilities for OpenClaw plugin.
 * Extracts user, agent, and session context from the OpenClaw runtime.
 */
/**
 * Extracts user context from OpenClaw runtime context.
 * Returns undefined if user information is not available.
 */
export function extractUserContext(runtimeContext) {
    if (!runtimeContext || typeof runtimeContext !== 'object') {
        return undefined;
    }
    const ctx = runtimeContext;
    const user = ctx.user;
    if (!user || typeof user.id !== 'string') {
        return undefined;
    }
    return {
        user_id: user.id,
        display_name: typeof user.display_name === 'string' ? user.display_name : undefined,
        email: typeof user.email === 'string' ? user.email : undefined,
    };
}
/**
 * Extracts agent context from OpenClaw runtime context.
 */
export function extractAgentContext(runtimeContext) {
    const defaultAgent = {
        agentId: 'unknown',
        name: 'Unknown Agent',
    };
    if (!runtimeContext || typeof runtimeContext !== 'object') {
        return defaultAgent;
    }
    const ctx = runtimeContext;
    const agent = ctx.agent;
    if (!agent) {
        return defaultAgent;
    }
    return {
        agentId: typeof agent.id === 'string' ? agent.id : 'unknown',
        name: typeof agent.name === 'string' ? agent.name : 'Unknown Agent',
        version: typeof agent.version === 'string' ? agent.version : undefined,
    };
}
/**
 * Extracts session context from OpenClaw runtime context.
 */
export function extractSessionContext(runtimeContext) {
    const defaultSession = {
        sessionId: crypto.randomUUID(),
        startedAt: new Date(),
    };
    if (!runtimeContext || typeof runtimeContext !== 'object') {
        return defaultSession;
    }
    const ctx = runtimeContext;
    const session = ctx.session;
    if (!session) {
        return defaultSession;
    }
    return {
        sessionId: typeof session.id === 'string' ? session.id : crypto.randomUUID(),
        startedAt: session.startedAt instanceof Date ? session.startedAt : typeof session.startedAt === 'string' ? new Date(session.startedAt) : new Date(),
        thread_id: typeof session.thread_id === 'string' ? session.thread_id : undefined,
    };
}
/**
 * Extracts complete plugin context from OpenClaw runtime.
 */
export function extractContext(runtimeContext) {
    return {
        user: extractUserContext(runtimeContext),
        agent: extractAgentContext(runtimeContext),
        session: extractSessionContext(runtimeContext),
    };
}
/** Maximum length for agent IDs (matches PostgreSQL identifier limits) */
export const MAX_AGENT_ID_LENGTH = 63;
/** Allowed characters in session keys: alphanumeric, colon, hyphen, underscore */
const SESSION_KEY_REGEX = /^[a-zA-Z0-9:_-]+$/;
/** Maximum length for session keys */
const MAX_SESSION_KEY_LENGTH = 500;
/**
 * Validates a session key format.
 * - Max length: 500 characters
 * - Allowed characters: alphanumeric, colon, hyphen, underscore
 */
export function validateSessionKey(sessionKey) {
    if (!sessionKey || sessionKey.length === 0) {
        return false;
    }
    if (sessionKey.length > MAX_SESSION_KEY_LENGTH) {
        return false;
    }
    return SESSION_KEY_REGEX.test(sessionKey);
}
/**
 * Parse agentId from session key.
 * Session key format: agent:<agentId>:<channel>:...
 * Returns "unknown" for invalid or missing keys.
 */
export function parseAgentIdFromSessionKey(sessionKey) {
    if (!sessionKey || sessionKey.length === 0) {
        return 'unknown';
    }
    // Validate session key format first
    if (!validateSessionKey(sessionKey)) {
        return 'unknown';
    }
    // Parse format: agent:<agentId>:<channel>:...
    const parts = sessionKey.split(':');
    if (parts.length < 3 || parts[0] !== 'agent') {
        return 'unknown';
    }
    const agentId = parts[1];
    if (!agentId || agentId.length === 0) {
        return 'unknown';
    }
    if (agentId.length > MAX_AGENT_ID_LENGTH) {
        return 'unknown';
    }
    return agentId;
}
/**
 * Resolve the effective agent ID from all available sources.
 *
 * Priority:
 * 1. Explicit config agentId (always wins)
 * 2. Hook context agentId (per-session, from gateway)
 * 3. Parsed from session key (fallback parsing)
 * 4. Existing state value (may be resolved from a previous hook call)
 * 5. "unknown" (last resort)
 *
 * Issue #1644: Agent ID must be resolved per-session from hook context,
 * not at plugin registration time from api.runtime.
 */
export function resolveAgentId(hookCtx, configAgentId, currentStateValue) {
    // 1. Explicit config (always wins)
    if (configAgentId)
        return configAgentId;
    // 2. Hook context agentId (skip if "unknown")
    if (hookCtx.agentId && hookCtx.agentId !== 'unknown')
        return hookCtx.agentId;
    // 3. Parse from session key
    const fromSession = parseAgentIdFromSessionKey(hookCtx.sessionKey);
    if (fromSession !== 'unknown')
        return fromSession;
    // 4. Existing state (may already be resolved from a previous hook call)
    if (currentStateValue !== 'unknown')
        return currentStateValue;
    // 5. Last resort
    return 'unknown';
}
/**
 * Get the user scope key based on configured scoping mode.
 * @param context - User scoping context
 * @param scopeMode - "agent" | "identity" | "session"
 * @returns Scope key to pass to backend API
 */
export function getUserScopeKey(context, scopeMode) {
    switch (scopeMode) {
        case 'agent':
            return context.agentId || 'unknown';
        case 'identity':
            // Prefer identity key if available, fall back to agent
            return context.identityKey || context.agentId || 'unknown';
        case 'session':
            // Prefer full session key if available, fall back to agent
            return context.sessionKey || context.agentId || 'unknown';
        default:
            return context.agentId || 'unknown';
    }
}
