// @troykelly/openclaw-projects/dist/index.js
/**
 * OpenClaw Projects Plugin
 *
 * This plugin provides memory management, projects, todos, and contacts
 * integration for OpenClaw agents.
 *
 * Registration pattern:
 * - OpenClaw 2026 API: `export default (api) => { ... }` (recommended)
 * - Legacy API: `register(ctx)` returns plugin instance
 */
// Re-export the OpenClaw 2026 API default export
export { default, registerOpenClaw, schemas } from './register-openclaw.js';
import { validateConfig, validateRawConfig, redactConfig } from './config.js';
import { createLogger } from './logger.js';
import { createApiClient } from './api-client.js';
import { extractContext, getUserScopeKey } from './context.js';
import { createAutoRecallHook, createAutoCaptureHook, createHealthCheck, } from './hooks.js';
import { createCliCommands } from './cli.js';
import { createMemoryRecallTool, createMemoryStoreTool, createMemoryForgetTool, createProjectListTool, createProjectGetTool, createProjectCreateTool, createTodoListTool, createTodoCreateTool, createTodoCompleteTool, createContactSearchTool, createContactGetTool, createContactCreateTool, } from './tools/index.js';
/**
 * Creates a plugin instance from resolved configuration.
 * Internal helper used by both sync and async registration.
 */
function createPluginInstance(config, logger, runtime) {
    // Create API client
    const apiClient = createApiClient({ config, logger });
    // Extract context
    const context = extractContext(runtime);
    // Determine user scope key based on config
    const user_id = getUserScopeKey({
        agentId: context.agent.agentId,
        sessionKey: context.session.sessionId,
    }, config.userScoping ?? 'agent');
    // Create tools
    const tools = {
        memoryRecall: createMemoryRecallTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        memoryStore: createMemoryStoreTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        memoryForget: createMemoryForgetTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        projectList: createProjectListTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        projectGet: createProjectGetTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        projectCreate: createProjectCreateTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        todoList: createTodoListTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        todoCreate: createTodoCreateTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        todoComplete: createTodoCompleteTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        contactSearch: createContactSearchTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        contactGet: createContactGetTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
        contactCreate: createContactCreateTool({
            client: apiClient,
            logger,
            config,
            user_id,
        }),
    };
    // Create lifecycle hooks
    const hooks = {
        beforeAgentStart: createAutoRecallHook({
            client: apiClient,
            logger,
            config,
            getAgentId: () => user_id,
        }),
        agentEnd: createAutoCaptureHook({
            client: apiClient,
            logger,
            config,
            getAgentId: () => user_id,
        }),
    };
    // Create health check
    const healthCheck = createHealthCheck({ client: apiClient, logger });
    // Create CLI commands
    const cli = createCliCommands({
        client: apiClient,
        logger,
        config,
        user_id,
    });
    logger.info('Plugin registered', {
        agentId: context.agent.agentId,
        sessionId: context.session.sessionId,
        user_id,
        config: redactConfig(config),
    });
    return {
        id: 'openclaw-projects',
        name: 'OpenClaw Projects Plugin',
        kind: 'memory',
        config,
        apiClient,
        context,
        tools,
        hooks,
        cli,
        healthCheck,
    };
}
/**
 * Registers the plugin with OpenClaw.
 *
 * Validates the raw configuration and resolves direct secret values
 * to produce a fully initialized plugin instance.
 */
export function register(ctx) {
    const logger = ctx.logger ?? createLogger('openclaw-projects');
    // Validate as raw config first to check structure
    const rawConfig = validateRawConfig(ctx.config);
    // For sync registration, we require direct values - file/command refs won't work
    // Build a resolved config from direct values only
    const config = validateConfig({
        apiUrl: rawConfig.apiUrl,
        apiKey: rawConfig.apiKey ?? '',
        twilioAccountSid: rawConfig.twilioAccountSid,
        twilioAuthToken: rawConfig.twilioAuthToken,
        twilioPhoneNumber: rawConfig.twilioPhoneNumber,
        postmarkToken: rawConfig.postmarkToken,
        postmarkFromEmail: rawConfig.postmarkFromEmail,
        secretCommandTimeout: rawConfig.secretCommandTimeout,
        autoRecall: rawConfig.autoRecall,
        autoCapture: rawConfig.autoCapture,
        userScoping: rawConfig.userScoping,
        maxRecallMemories: rawConfig.maxRecallMemories,
        minRecallScore: rawConfig.minRecallScore,
        timeout: rawConfig.timeout,
        maxRetries: rawConfig.maxRetries,
        debug: rawConfig.debug,
    });
    return createPluginInstance(config, logger, ctx.runtime);
}
/** Plugin definition object for OpenClaw */
export const plugin = {
    id: 'openclaw-projects',
    name: 'OpenClaw Projects Plugin',
    kind: 'memory',
    register,
};
export { validateConfig, safeValidateConfig, validateRawConfig, safeValidateRawConfig, resolveConfigSecrets, resolveConfigSecretsSync, redactConfig, resolveNamespaceConfig, } from './config.js';
export { resolveSecret, resolveSecretSync, resolveSecrets, clearSecretCache, clearCachedSecret } from './secrets.js';
export { createLogger, redactSensitive } from './logger.js';
export { createApiClient } from './api-client.js';
export { extractContext, getUserScopeKey } from './context.js';
export { createMemoryRecallTool, createMemoryStoreTool, createMemoryForgetTool, createProjectListTool, createProjectGetTool, createProjectCreateTool, createTodoListTool, createTodoCreateTool, createTodoCompleteTool, createContactSearchTool, createContactGetTool, createContactCreateTool, MemoryRecallParamsSchema, MemoryStoreParamsSchema, MemoryForgetParamsSchema, MemoryCategory, ProjectListParamsSchema, ProjectGetParamsSchema, ProjectCreateParamsSchema, ProjectStatus, TodoListParamsSchema, TodoCreateParamsSchema, TodoCompleteParamsSchema, ContactSearchParamsSchema, ContactGetParamsSchema, ContactCreateParamsSchema, } from './tools/index.js';
export { createAutoRecallHook, createAutoCaptureHook, createHealthCheck, } from './hooks.js';
export { createCliCommands, createStatusCommand, createUsersCommand, createRecallCommand, } from './cli.js';
export { createGatewayMethods, registerGatewayRpcMethods, } from './gateway/rpc-methods.js';
