// @troykelly/openclaw-projects/dist/secrets.js
/**
 * Flexible secret resolution module.
 *
 * Supports three methods for loading secrets (in priority order):
 * 1. Command execution (e.g., 1Password CLI: `op read op://...`)
 * 2. File reference (e.g., ~/.secrets/api_key)
 * 3. Direct value (least secure, for development only)
 *
 * Security note: Command execution intentionally uses execSync with shell
 * because the command comes from trusted configuration, not user input.
 * This is required to support shell commands like `op read 'op://...'`
 * for secret managers.
 *
 * @module secrets
 */
import { execSync } from 'node:child_process';
import { readFileSync, statSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';
/** Default timeout for command execution in milliseconds */
const DEFAULT_COMMAND_TIMEOUT = 5000;
/**
 * Cache for resolved secrets.
 * Intentionally shared between async (resolveSecret) and sync (resolveSecretSync) paths.
 * This is safe because resolveSecretSync runs during synchronous plugin registration
 * (blocking the event loop), and resolveSecret runs afterward in async contexts.
 */
const secretCache = new Map();
/**
 * Expands ~ to the user's home directory.
 */
function expandTilde(filePath) {
    if (filePath.startsWith('~/')) {
        return join(homedir(), filePath.slice(2));
    }
    if (filePath === '~') {
        return homedir();
    }
    return filePath;
}
/**
 * Checks file permissions and warns if world-readable.
 */
function checkFilePermissions(filePath) {
    try {
        const stats = statSync(filePath);
        const mode = stats.mode & 0o777;
        if (mode & 0o004) {
            console.warn(`[Secrets] Warning: Secret file ${filePath} is world-readable (mode ${mode.toString(8)}). ` +
                'Consider restricting permissions with: chmod 600 ' +
                filePath);
        }
    }
    catch (error) {
        console.warn(`[Secrets] Could not check permissions for ${filePath}: ${error.message}`);
    }
}
/**
 * Resolves a secret from command execution.
 *
 * Security note: This uses execSync intentionally because:
 * 1. The command comes from plugin configuration set by system administrators
 * 2. It must support shell commands like `op read 'op://...'` for secret managers
 * 3. This is NOT user input - it's trusted configuration
 */
function resolveFromCommand(command, timeout) {
    try {
        // eslint-disable-next-line security/detect-child-process
        const result = execSync(command, {
            encoding: 'utf-8',
            timeout,
            stdio: ['ignore', 'pipe', 'pipe'],
        });
        return result.trim();
    }
    catch (error) {
        const err = error;
        if (err.killed && err.signal === 'SIGTERM') {
            throw new Error(`Secret command timed out after ${timeout}ms`);
        }
        if (err.killed) {
            throw new Error(`Secret command was killed (signal: ${err.signal ?? 'unknown'})`);
        }
        const message = error instanceof Error ? error.message : String(error);
        throw new Error(`Secret command failed: ${message}`);
    }
}
/**
 * Resolves a secret from a file.
 */
function resolveFromFile(filePath) {
    const expandedPath = expandTilde(filePath);
    checkFilePermissions(expandedPath);
    try {
        const content = readFileSync(expandedPath, 'utf-8');
        return content.trim();
    }
    catch (error) {
        const err = error instanceof Error ? error : new Error(String(error));
        if (error.code === 'ENOENT') {
            throw new Error(`Secret file does not exist: ${expandedPath}`);
        }
        throw new Error(`Failed to read secret file: ${err.message}`);
    }
}
/**
 * Resolves a secret using the configured method.
 *
 * Priority order (highest first):
 * 1. Command - Execute shell command and use output
 * 2. File - Read from file path
 * 3. Direct - Use direct value from config
 *
 * @param config - Secret configuration
 * @param cacheKey - Optional key to cache the resolved secret
 * @returns The resolved secret, or undefined if not configured
 */
export async function resolveSecret(config, cacheKey) {
    // Check cache first
    if (cacheKey && secretCache.has(cacheKey)) {
        return secretCache.get(cacheKey);
    }
    let resolved;
    // Priority 1: Command
    if (config.command?.trim()) {
        const timeout = config.commandTimeout ?? DEFAULT_COMMAND_TIMEOUT;
        resolved = resolveFromCommand(config.command, timeout);
    }
    // Priority 2: File
    else if (config.file?.trim()) {
        resolved = resolveFromFile(config.file);
    }
    // Priority 3: Direct
    else if (config.direct !== undefined) {
        const trimmed = config.direct.trim();
        resolved = trimmed.length > 0 ? trimmed : undefined;
    }
    // Cache if resolved and cacheKey provided
    if (resolved && cacheKey) {
        secretCache.set(cacheKey, resolved);
    }
    return resolved;
}
/**
 * Resolves a secret synchronously using the configured method.
 *
 * Same priority and behavior as resolveSecret, but fully synchronous.
 * Used during plugin registration where blocking I/O is acceptable
 * and the caller (OpenClaw loader) does not await the result.
 *
 * @param config - Secret configuration
 * @param cacheKey - Optional key to cache the resolved secret
 * @returns The resolved secret, or undefined if not configured
 */
export function resolveSecretSync(config, cacheKey) {
    // Check cache first
    if (cacheKey && secretCache.has(cacheKey)) {
        return secretCache.get(cacheKey);
    }
    let resolved;
    // Priority 1: Command
    if (config.command?.trim()) {
        const timeout = config.commandTimeout ?? DEFAULT_COMMAND_TIMEOUT;
        resolved = resolveFromCommand(config.command, timeout);
    }
    // Priority 2: File
    else if (config.file?.trim()) {
        resolved = resolveFromFile(config.file);
    }
    // Priority 3: Direct
    else if (config.direct !== undefined) {
        const trimmed = config.direct.trim();
        resolved = trimmed.length > 0 ? trimmed : undefined;
    }
    // Cache if resolved and cacheKey provided
    if (resolved && cacheKey) {
        secretCache.set(cacheKey, resolved);
    }
    return resolved;
}
/**
 * Resolves multiple secrets in parallel.
 *
 * @param configs - Map of secret names to configurations
 * @returns Map of secret names to resolved values
 */
export async function resolveSecrets(configs) {
    const entries = Object.entries(configs);
    const results = await Promise.all(entries.map(async ([key, config]) => {
        const value = await resolveSecret(config, key);
        return [key, value];
    }));
    return Object.fromEntries(results);
}
/**
 * Clears the secret cache.
 * Call this when configuration is reloaded or on SIGHUP.
 */
export function clearSecretCache() {
    secretCache.clear();
}
/**
 * Clears a specific secret from the cache.
 */
export function clearCachedSecret(key) {
    secretCache.delete(key);
}
