// ocft/dist/protocol.js
/**
 * OCFT - OpenClaw File Transfer Protocol
 * Message Types and Interfaces
 */
export const OCFT_VERSION = '1.0';
// Create Message Helper
export function createMessage(type, transferId, from, to, payload) {
    return {
        version: OCFT_VERSION,
        type,
        transferId,
        from,
        to,
        timestamp: Date.now(),
        payload
    };
}
// Serialize/Deserialize
export function serialize(msg) {
    return JSON.stringify(msg);
}
export function deserialize(data) {
    try {
        const msg = JSON.parse(data);
        if (msg.version && msg.type && msg.transferId) {
            return msg;
        }
        return null;
    }
    catch {
        return null;
    }
}
// Magic prefix for OCFT messages (to identify in chat)
export const OCFT_PREFIX = '🔗OCFT:';
// Encode message for chat (with prefix)
export function encodeForChat(msg) {
    const json = serialize(msg);
    const b64 = Buffer.from(json).toString('base64');
    return `${OCFT_PREFIX}${b64}`;
}
// Decode message from chat
export function decodeFromChat(text) {
    if (!text.startsWith(OCFT_PREFIX))
        return null;
    try {
        const b64 = text.slice(OCFT_PREFIX.length);
        const json = Buffer.from(b64, 'base64').toString('utf-8');
        return deserialize(json);
    }
    catch {
        return null;
    }
}
