// @troykelly/openclaw-projects/dist/utils/inbound-gate.js
/**
 * Inbound message gate — unified pre-processing gate for inbound messages.
 *
 * Orchestrates spam filter, rate limiter, and token budget as a single
 * evaluation point before any message processing (embedding, linking, etc.).
 *
 * NOTE: All state (rate limiter windows, token budget counters) is held
 * in-memory and does not persist across process restarts or span multiple
 * instances. This is acceptable for single-instance deployments. For
 * multi-instance deployments, the rate limiter and token budget would need
 * to be backed by skill_store or an external store (e.g. Redis) to share
 * state across processes.
 *
 * Part of Issue #1225 — rate limiting and spam protection.
 */
import { isSpam } from './spam-filter.js';
import { createRateLimiter } from './rate-limiter.js';
import { createTokenBudget } from './token-budget.js';
/**
 * Create an inbound message gate.
 *
 * Checks are applied in this order:
 * 1. Spam filter (reject bulk/marketing/known spam)
 * 2. Rate limiter (per-sender and per-recipient caps)
 * 3. Token budget (cost protection)
 * 4. Deferred processing (unknown senders, if configured)
 *
 * @param config - Gate configuration
 * @param logger - Logger for recording decisions
 * @returns InboundGate instance
 */
export function createInboundGate(config, logger) {
    const rateLimiter = createRateLimiter(config.rateLimiter);
    const tokenBudget = createTokenBudget(config.tokenBudget);
    let totalEvaluated = 0;
    let allowed = 0;
    let rejected = 0;
    let rateLimited = 0;
    let budgetExceeded = 0;
    let deferred = 0;
    return {
        evaluate(message, trust, estimatedTokens = 0) {
            totalEvaluated++;
            // Step 1: Spam filter
            const spamResult = isSpam(message, config.spamFilter);
            if (spamResult.isSpam) {
                rejected++;
                logger.info('inbound gate: spam rejected', {
                    sender: message.sender,
                    channel: message.channel,
                    reason: spamResult.reason,
                });
                return {
                    action: 'reject',
                    reason: `spam: ${spamResult.reason}`,
                    skipEmbedding: true,
                };
            }
            // Step 2: Rate limiter (pass channel for sender normalization)
            const rateResult = rateLimiter.check(message.sender, message.recipient, trust, message.channel);
            if (!rateResult.allowed) {
                rateLimited++;
                logger.warn('inbound gate: rate limited', {
                    sender: message.sender,
                    channel: message.channel,
                    trust,
                    reason: rateResult.reason,
                    remaining: rateResult.remaining,
                    retryAfterMs: rateResult.retryAfterMs,
                });
                return {
                    action: 'rate_limited',
                    reason: `rate limited: ${rateResult.reason}`,
                    skipEmbedding: true,
                };
            }
            // Step 3: Token budget (uses tryConsume for atomic check-and-record)
            if (estimatedTokens > 0) {
                const budgetResult = tokenBudget.tryConsume(estimatedTokens);
                if (!budgetResult.allowed) {
                    budgetExceeded++;
                    logger.warn('inbound gate: token budget exceeded', {
                        sender: message.sender,
                        channel: message.channel,
                        estimatedTokens,
                        reason: budgetResult.reason,
                    });
                    return {
                        action: 'budget_exceeded',
                        reason: `budget exceeded: ${budgetResult.reason}`,
                        skipEmbedding: true,
                    };
                }
            }
            // Step 4: Deferred processing for unknown senders
            if (config.deferUnknownSenders && trust === 'unknown') {
                deferred++;
                logger.debug('inbound gate: deferring unknown sender', {
                    sender: message.sender,
                    channel: message.channel,
                });
                return {
                    action: 'defer',
                    reason: 'unknown sender deferred for batch processing',
                    skipEmbedding: true,
                };
            }
            // Message is allowed
            allowed++;
            logger.debug('inbound gate: allowed', {
                sender: message.sender,
                channel: message.channel,
                trust,
            });
            return {
                action: 'allow',
                reason: null,
                skipEmbedding: false,
            };
        },
        getStats() {
            return {
                totalEvaluated,
                allowed,
                rejected,
                rateLimited,
                budgetExceeded,
                deferred,
            };
        },
    };
}
