// @troykelly/openclaw-projects/dist/utils/injection-log-rate-limiter.js
/**
 * Rate limiter for injection detection logging.
 *
 * Prevents log flooding when an attacker sends many injection-containing
 * messages. Logs the first N detections per key (sender/user_id) per window,
 * then samples at a configurable rate. Emits summary entries for suppressed
 * counts when a new window begins.
 *
 * Memory-bounded via maxEntries with LRU eviction of oldest keys.
 *
 * Issue #1257
 */
/** Default configuration */
export const DEFAULT_INJECTION_LOG_RATE_LIMITER_CONFIG = {
    windowMs: 60_000,
    maxLogsPerWindow: 10,
    sampleRate: 0.1,
    maxEntries: 10_000,
};
/**
 * Create a new injection log rate limiter.
 *
 * @param config - Configuration overrides (defaults to DEFAULT_INJECTION_LOG_RATE_LIMITER_CONFIG)
 * @param randomFn - Optional random function for testing (defaults to Math.random)
 */
export function createInjectionLogRateLimiter(config = DEFAULT_INJECTION_LOG_RATE_LIMITER_CONFIG, randomFn = Math.random) {
    const entries = new Map();
    function now() {
        return Date.now();
    }
    function evictIfNeeded() {
        if (entries.size <= config.maxEntries)
            return;
        // Sort by lastSeen ascending (oldest first), evict until at maxEntries
        const sorted = [...entries.entries()].sort((a, b) => a[1].lastSeen - b[1].lastSeen);
        const toEvict = sorted.length - config.maxEntries;
        for (let i = 0; i < toEvict; i++) {
            entries.delete(sorted[i][0]);
        }
    }
    function shouldLog(key) {
        const currentTime = now();
        const existing = entries.get(key);
        // Check if we need to start a new window
        if (existing && currentTime - existing.windowStart >= config.windowMs) {
            // Window expired — check if there were suppressions to report
            const previousSuppressed = existing.suppressedCount;
            // Reset to new window
            existing.windowStart = currentTime;
            existing.logCount = 1;
            existing.suppressedCount = 0;
            existing.lastSeen = currentTime;
            if (previousSuppressed > 0) {
                // Emit a summary log entry reporting previous window suppressions
                return { log: true, suppressed: previousSuppressed, summary: true };
            }
            // Normal first log in new window, no summary needed
            return { log: true, suppressed: 0, summary: false };
        }
        if (!existing) {
            // First detection for this key
            entries.set(key, {
                windowStart: currentTime,
                logCount: 1,
                suppressedCount: 0,
                lastSeen: currentTime,
            });
            evictIfNeeded();
            return { log: true, suppressed: 0, summary: false };
        }
        // Within the same window
        existing.lastSeen = currentTime;
        if (existing.logCount < config.maxLogsPerWindow) {
            // Still under the limit
            existing.logCount++;
            return { log: true, suppressed: 0, summary: false };
        }
        // Over the limit — apply sampling
        existing.suppressedCount++;
        if (config.sampleRate > 0 && randomFn() < config.sampleRate) {
            // Sampled in — log it
            existing.logCount++;
            return { log: true, suppressed: existing.suppressedCount, summary: false };
        }
        // Suppressed
        return { log: false, suppressed: existing.suppressedCount, summary: false };
    }
    function getStats() {
        let totalSuppressed = 0;
        for (const state of entries.values()) {
            totalSuppressed += state.suppressedCount;
        }
        return {
            activeKeys: entries.size,
            totalSuppressed,
        };
    }
    return { shouldLog, getStats };
}
/**
 * Shared singleton rate limiter for injection detection logging.
 * All callsites in the plugin share this instance so that rate limiting
 * is coordinated across tools and hooks within the same process.
 */
export const injectionLogLimiter = createInjectionLogRateLimiter();
