// @troykelly/openclaw-projects/dist/utils/token-budget.js
/**
 * Token budget tracking utility for inbound message processing.
 *
 * Provides configurable daily and monthly caps on token expenditure
 * to protect against cost overruns from bulk inbound messages.
 *
 * Part of Issue #1225 — rate limiting and spam protection.
 */
/** Default configuration (budget disabled) */
export const DEFAULT_TOKEN_BUDGET_CONFIG = {
    enabled: false,
};
/**
 * Create a token budget tracker.
 *
 * Tracks daily and monthly token usage with automatic reset at
 * UTC midnight (daily) and UTC month boundary (monthly).
 *
 * @param config - Token budget configuration
 * @returns TokenBudget instance
 */
export function createTokenBudget(config = DEFAULT_TOKEN_BUDGET_CONFIG) {
    let dailyUsed = 0;
    let monthlyUsed = 0;
    let currentDay = getCurrentDay();
    let currentMonth = getCurrentMonth();
    /**
     * Get the current UTC day as YYYY-MM-DD string.
     */
    function getCurrentDay() {
        return new Date().toISOString().slice(0, 10);
    }
    /**
     * Get the current UTC month as YYYY-MM string.
     */
    function getCurrentMonth() {
        return new Date().toISOString().slice(0, 7);
    }
    /**
     * Reset counters if the day or month has changed.
     */
    function maybeReset() {
        const today = getCurrentDay();
        const thisMonth = getCurrentMonth();
        if (today !== currentDay) {
            dailyUsed = 0;
            currentDay = today;
        }
        if (thisMonth !== currentMonth) {
            monthlyUsed = 0;
            currentMonth = thisMonth;
        }
    }
    return {
        check(tokens) {
            if (!config.enabled) {
                return { allowed: true, reason: null };
            }
            maybeReset();
            // Check daily limit
            if (config.dailyTokenLimit !== undefined) {
                if (dailyUsed + tokens > config.dailyTokenLimit) {
                    return {
                        allowed: false,
                        reason: `daily token budget exceeded (${dailyUsed}/${config.dailyTokenLimit} used, requested ${tokens})`,
                        remainingDaily: Math.max(0, config.dailyTokenLimit - dailyUsed),
                        remainingMonthly: config.monthlyTokenLimit !== undefined ? Math.max(0, config.monthlyTokenLimit - monthlyUsed) : undefined,
                    };
                }
            }
            // Check monthly limit
            if (config.monthlyTokenLimit !== undefined) {
                if (monthlyUsed + tokens > config.monthlyTokenLimit) {
                    return {
                        allowed: false,
                        reason: `monthly token budget exceeded (${monthlyUsed}/${config.monthlyTokenLimit} used, requested ${tokens})`,
                        remainingDaily: config.dailyTokenLimit !== undefined ? Math.max(0, config.dailyTokenLimit - dailyUsed) : undefined,
                        remainingMonthly: Math.max(0, config.monthlyTokenLimit - monthlyUsed),
                    };
                }
            }
            return {
                allowed: true,
                reason: null,
                remainingDaily: config.dailyTokenLimit !== undefined ? config.dailyTokenLimit - dailyUsed : undefined,
                remainingMonthly: config.monthlyTokenLimit !== undefined ? config.monthlyTokenLimit - monthlyUsed : undefined,
            };
        },
        record(tokens) {
            maybeReset();
            dailyUsed += tokens;
            monthlyUsed += tokens;
        },
        tryConsume(tokens) {
            if (!config.enabled) {
                return { allowed: true, reason: null };
            }
            maybeReset();
            // Check daily limit
            if (config.dailyTokenLimit !== undefined) {
                if (dailyUsed + tokens > config.dailyTokenLimit) {
                    return {
                        allowed: false,
                        reason: `daily token budget exceeded (${dailyUsed}/${config.dailyTokenLimit} used, requested ${tokens})`,
                        remainingDaily: Math.max(0, config.dailyTokenLimit - dailyUsed),
                        remainingMonthly: config.monthlyTokenLimit !== undefined ? Math.max(0, config.monthlyTokenLimit - monthlyUsed) : undefined,
                    };
                }
            }
            // Check monthly limit
            if (config.monthlyTokenLimit !== undefined) {
                if (monthlyUsed + tokens > config.monthlyTokenLimit) {
                    return {
                        allowed: false,
                        reason: `monthly token budget exceeded (${monthlyUsed}/${config.monthlyTokenLimit} used, requested ${tokens})`,
                        remainingDaily: config.dailyTokenLimit !== undefined ? Math.max(0, config.dailyTokenLimit - dailyUsed) : undefined,
                        remainingMonthly: Math.max(0, config.monthlyTokenLimit - monthlyUsed),
                    };
                }
            }
            // Atomically record usage since check passed
            dailyUsed += tokens;
            monthlyUsed += tokens;
            return {
                allowed: true,
                reason: null,
                remainingDaily: config.dailyTokenLimit !== undefined ? config.dailyTokenLimit - dailyUsed : undefined,
                remainingMonthly: config.monthlyTokenLimit !== undefined ? config.monthlyTokenLimit - monthlyUsed : undefined,
            };
        },
        getStats() {
            maybeReset();
            return {
                dailyUsed,
                monthlyUsed,
                dailyLimit: config.dailyTokenLimit,
                monthlyLimit: config.monthlyTokenLimit,
            };
        },
    };
}
