// @troykelly/openclaw-projects/dist/utils/zod-to-json-schema.js
/**
 * Zod to JSON Schema conversion utilities.
 *
 * Converts Zod schemas to JSON Schema format for OpenClaw tool registration.
 */
import { z } from 'zod';
/**
 * Convert a Zod schema to JSON Schema.
 *
 * Note: This is a simplified converter that handles common cases.
 * For complex schemas, consider using zodToJsonSchema library.
 */
export function zodToJsonSchema(schema) {
    return convertZodType(schema);
}
function convertZodType(schema) {
    // Unwrap optional/nullable
    if (schema instanceof z.ZodOptional || schema instanceof z.ZodNullable) {
        return convertZodType(schema.unwrap());
    }
    // Unwrap default
    if (schema instanceof z.ZodDefault) {
        const inner = convertZodType(schema._def.innerType);
        return { ...inner, default: schema._def.defaultValue() };
    }
    // String
    if (schema instanceof z.ZodString) {
        const result = { type: 'string' };
        const checks = schema._def.checks ?? [];
        for (const check of checks) {
            if (check.kind === 'min')
                result.minLength = check.value;
            if (check.kind === 'max')
                result.maxLength = check.value;
            if (check.kind === 'regex')
                result.pattern = check.regex.source;
            if (check.kind === 'email')
                result.format = 'email';
            if (check.kind === 'uuid')
                result.format = 'uuid';
            if (check.kind === 'url')
                result.format = 'uri';
        }
        if (schema.description)
            result.description = schema.description;
        return result;
    }
    // Number
    if (schema instanceof z.ZodNumber) {
        const checks = schema._def.checks ?? [];
        const isInt = checks.some((c) => c.kind === 'int');
        const result = { type: isInt ? 'integer' : 'number' };
        for (const check of checks) {
            if (check.kind === 'min')
                result.minimum = check.value;
            if (check.kind === 'max')
                result.maximum = check.value;
        }
        if (schema.description)
            result.description = schema.description;
        return result;
    }
    // Boolean
    if (schema instanceof z.ZodBoolean) {
        const result = { type: 'boolean' };
        if (schema.description)
            result.description = schema.description;
        return result;
    }
    // Enum
    if (schema instanceof z.ZodEnum) {
        const result = {
            type: 'string',
            enum: schema._def.values,
        };
        if (schema.description)
            result.description = schema.description;
        return result;
    }
    // Array
    if (schema instanceof z.ZodArray) {
        const result = {
            type: 'array',
            items: convertZodType(schema._def.type),
        };
        if (schema.description)
            result.description = schema.description;
        return result;
    }
    // Object
    if (schema instanceof z.ZodObject) {
        const shape = schema._def.shape();
        const properties = {};
        const required = [];
        for (const [key, value] of Object.entries(shape)) {
            properties[key] = convertZodType(value);
            // Check if field is required (not optional)
            if (!(value instanceof z.ZodOptional) && !(value instanceof z.ZodDefault)) {
                required.push(key);
            }
        }
        const result = {
            type: 'object',
            properties,
        };
        if (required.length > 0) {
            result.required = required;
        }
        if (schema.description) {
            result.description = schema.description;
        }
        return result;
    }
    // Fallback for unknown types
    return { type: 'object' };
}
/**
 * Add descriptions to JSON Schema properties from a map.
 */
export function addDescriptions(schema, descriptions) {
    const result = { ...schema };
    if (result.properties) {
        result.properties = { ...result.properties };
        for (const [key, desc] of Object.entries(descriptions)) {
            if (result.properties[key]) {
                result.properties[key] = { ...result.properties[key], description: desc };
            }
        }
    }
    return result;
}
