// @troykelly/openclaw-projects/dist/services/api-source-service.js
/**
 * Plugin API source service client.
 * HTTP client for calling api-sources API endpoints from the plugin.
 * Part of API Onboarding feature (#1783).
 */
// ── Service ─────────────────────────────────────────────────────────────────
/**
 * HTTP client wrapper for the api-sources API.
 * Follows the same pattern as notification-service.ts.
 */
export class ApiSourceService {
    client;
    constructor(client) {
        this.client = client;
    }
    /**
     * Onboard a new API source from an OpenAPI spec URL or inline content.
     */
    async onboard(params, options) {
        return this.client.post('/api-sources', params, options);
    }
    /**
     * Search API memories using hybrid search.
     */
    async search(params, options) {
        const searchParams = new URLSearchParams();
        searchParams.set('q', params.q);
        if (params.limit !== undefined)
            searchParams.set('limit', String(params.limit));
        if (params.offset !== undefined)
            searchParams.set('offset', String(params.offset));
        if (params.memory_kind)
            searchParams.set('memory_kind', params.memory_kind);
        if (params.api_source_id)
            searchParams.set('api_source_id', params.api_source_id);
        if (params.tags && params.tags.length > 0)
            searchParams.set('tags', params.tags.join(','));
        return this.client.get(`/api-memories/search?${searchParams}`, options);
    }
    /**
     * Get a single API source by ID.
     */
    async get(id, options) {
        return this.client.get(`/api-sources/${encodeURIComponent(id)}`, options);
    }
    /**
     * List API sources.
     */
    async list(params, options) {
        const searchParams = new URLSearchParams();
        if (params?.limit !== undefined)
            searchParams.set('limit', String(params.limit));
        if (params?.offset !== undefined)
            searchParams.set('offset', String(params.offset));
        if (params?.status)
            searchParams.set('status', params.status);
        const qs = searchParams.toString();
        const path = qs ? `/api-sources?${qs}` : '/api-sources';
        return this.client.get(path, options);
    }
    /**
     * Update an API source.
     */
    async update(id, params, options) {
        return this.client.patch(`/api-sources/${encodeURIComponent(id)}`, params, options);
    }
    /**
     * Manage credentials for an API source (add/update/remove).
     */
    async manageCredential(sourceId, params, options) {
        const basePath = `/api-sources/${encodeURIComponent(sourceId)}/credentials`;
        if (params.action === 'add') {
            return this.client.post(basePath, {
                header_name: params.header_name,
                header_prefix: params.header_prefix,
                resolve_strategy: params.resolve_strategy,
                resolve_reference: params.resolve_reference,
                purpose: params.purpose,
            }, options);
        }
        if (params.action === 'update' && params.credential_id) {
            return this.client.patch(`${basePath}/${encodeURIComponent(params.credential_id)}`, {
                header_name: params.header_name,
                header_prefix: params.header_prefix,
                resolve_strategy: params.resolve_strategy,
                resolve_reference: params.resolve_reference,
                purpose: params.purpose,
            }, options);
        }
        if (params.action === 'remove' && params.credential_id) {
            return this.client.delete(`${basePath}/${encodeURIComponent(params.credential_id)}`, options);
        }
        return {
            success: false,
            error: {
                status: 400,
                message: 'Invalid credential management action or missing credential_id',
            },
        };
    }
    /**
     * Refresh an API source from its spec URL.
     */
    async refresh(id, options) {
        return this.client.post(`/api-sources/${encodeURIComponent(id)}/refresh`, undefined, options);
    }
    /**
     * Soft-delete an API source.
     */
    async remove(id, options) {
        return this.client.delete(`/api-sources/${encodeURIComponent(id)}`, options);
    }
    /**
     * Restore a soft-deleted API source.
     */
    async restore(id, options) {
        return this.client.post(`/api-sources/${encodeURIComponent(id)}/restore`, undefined, options);
    }
}
