// browserclaw/dist/index.js
import os from 'os';
import path, { normalize, resolve, dirname, sep } from 'path';
import fs from 'fs';
import net from 'net';
import { spawn, execFileSync } from 'child_process';
import { devices, chromium } from 'playwright-core';
import { lookup } from 'dns/promises';
import { lstat, realpath } from 'fs/promises';

// src/chrome-launcher.ts
var CHROMIUM_BUNDLE_IDS = /* @__PURE__ */ new Set([
  "com.google.Chrome",
  "com.google.Chrome.beta",
  "com.google.Chrome.canary",
  "com.google.Chrome.dev",
  "com.brave.Browser",
  "com.brave.Browser.beta",
  "com.brave.Browser.nightly",
  "com.microsoft.Edge",
  "com.microsoft.EdgeBeta",
  "com.microsoft.EdgeDev",
  "com.microsoft.EdgeCanary",
  "org.chromium.Chromium",
  "com.vivaldi.Vivaldi",
  "com.operasoftware.Opera",
  "com.operasoftware.OperaGX",
  "com.yandex.desktop.yandex-browser",
  "company.thebrowser.Browser"
]);
var CHROMIUM_DESKTOP_IDS = /* @__PURE__ */ new Set([
  "google-chrome.desktop",
  "google-chrome-beta.desktop",
  "google-chrome-unstable.desktop",
  "brave-browser.desktop",
  "microsoft-edge.desktop",
  "microsoft-edge-beta.desktop",
  "microsoft-edge-dev.desktop",
  "microsoft-edge-canary.desktop",
  "chromium.desktop",
  "chromium-browser.desktop",
  "vivaldi.desktop",
  "vivaldi-stable.desktop",
  "opera.desktop",
  "opera-gx.desktop",
  "yandex-browser.desktop",
  "org.chromium.Chromium.desktop"
]);
var CHROMIUM_EXE_NAMES = /* @__PURE__ */ new Set([
  "chrome.exe",
  "msedge.exe",
  "brave.exe",
  "brave-browser.exe",
  "chromium.exe",
  "vivaldi.exe",
  "opera.exe",
  "launcher.exe",
  "yandex.exe",
  "yandexbrowser.exe",
  "google chrome",
  "google chrome canary",
  "brave browser",
  "microsoft edge",
  "chromium",
  "chrome",
  "brave",
  "msedge",
  "brave-browser",
  "google-chrome",
  "google-chrome-stable",
  "google-chrome-beta",
  "google-chrome-unstable",
  "microsoft-edge",
  "microsoft-edge-beta",
  "microsoft-edge-dev",
  "microsoft-edge-canary",
  "chromium-browser",
  "vivaldi",
  "vivaldi-stable",
  "opera",
  "opera-stable",
  "opera-gx",
  "yandex-browser"
]);
function fileExists(filePath) {
  try {
    return fs.existsSync(filePath);
  } catch {
    return false;
  }
}
function execText(command, args, timeoutMs = 1200) {
  try {
    const output = execFileSync(command, args, {
      timeout: timeoutMs,
      encoding: "utf8",
      maxBuffer: 1024 * 1024
    });
    return String(output ?? "").trim() || null;
  } catch {
    return null;
  }
}
function inferKindFromIdentifier(identifier) {
  const id = identifier.toLowerCase();
  if (id.includes("brave")) return "brave";
  if (id.includes("edge")) return "edge";
  if (id.includes("chromium")) return "chromium";
  if (id.includes("canary")) return "canary";
  if (id.includes("opera") || id.includes("vivaldi") || id.includes("yandex") || id.includes("thebrowser")) return "chromium";
  return "chrome";
}
function inferKindFromExeName(name) {
  const lower = name.toLowerCase();
  if (lower.includes("brave")) return "brave";
  if (lower.includes("edge") || lower.includes("msedge")) return "edge";
  if (lower.includes("chromium")) return "chromium";
  if (lower.includes("canary") || lower.includes("sxs")) return "canary";
  if (lower.includes("opera") || lower.includes("vivaldi") || lower.includes("yandex")) return "chromium";
  return "chrome";
}
function findFirstExe(candidates) {
  for (const c of candidates) if (fileExists(c.path)) return c;
  return null;
}
function detectDefaultBrowserBundleIdMac() {
  const plistPath = path.join(os.homedir(), "Library/Preferences/com.apple.LaunchServices/com.apple.launchservices.secure.plist");
  if (!fileExists(plistPath)) return null;
  const handlersRaw = execText("/usr/bin/plutil", ["-extract", "LSHandlers", "json", "-o", "-", "--", plistPath], 2e3);
  if (!handlersRaw) return null;
  let handlers;
  try {
    handlers = JSON.parse(handlersRaw);
  } catch {
    return null;
  }
  if (!Array.isArray(handlers)) return null;
  const resolveScheme = (scheme) => {
    let candidate = null;
    for (const entry of handlers) {
      if (!entry || typeof entry !== "object") continue;
      if (entry.LSHandlerURLScheme !== scheme) continue;
      const role = typeof entry.LSHandlerRoleAll === "string" && entry.LSHandlerRoleAll || typeof entry.LSHandlerRoleViewer === "string" && entry.LSHandlerRoleViewer || null;
      if (role) candidate = role;
    }
    return candidate;
  };
  return resolveScheme("http") ?? resolveScheme("https");
}
function detectDefaultChromiumMac() {
  const bundleId = detectDefaultBrowserBundleIdMac();
  if (!bundleId || !CHROMIUM_BUNDLE_IDS.has(bundleId)) return null;
  const appPathRaw = execText("/usr/bin/osascript", ["-e", `POSIX path of (path to application id "${bundleId}")`]);
  if (!appPathRaw) return null;
  const appPath = appPathRaw.trim().replace(/\/$/, "");
  const exeName = execText("/usr/bin/defaults", ["read", path.join(appPath, "Contents", "Info"), "CFBundleExecutable"]);
  if (!exeName) return null;
  const exePath = path.join(appPath, "Contents", "MacOS", exeName.trim());
  if (!fileExists(exePath)) return null;
  return { kind: inferKindFromIdentifier(bundleId), path: exePath };
}
function findChromeMac() {
  return findFirstExe([
    { kind: "chrome", path: "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome" },
    { kind: "chrome", path: path.join(os.homedir(), "Applications/Google Chrome.app/Contents/MacOS/Google Chrome") },
    { kind: "brave", path: "/Applications/Brave Browser.app/Contents/MacOS/Brave Browser" },
    { kind: "brave", path: path.join(os.homedir(), "Applications/Brave Browser.app/Contents/MacOS/Brave Browser") },
    { kind: "edge", path: "/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge" },
    { kind: "edge", path: path.join(os.homedir(), "Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge") },
    { kind: "chromium", path: "/Applications/Chromium.app/Contents/MacOS/Chromium" },
    { kind: "chromium", path: path.join(os.homedir(), "Applications/Chromium.app/Contents/MacOS/Chromium") },
    { kind: "canary", path: "/Applications/Google Chrome Canary.app/Contents/MacOS/Google Chrome Canary" },
    { kind: "canary", path: path.join(os.homedir(), "Applications/Google Chrome Canary.app/Contents/MacOS/Google Chrome Canary") }
  ]);
}
function detectDefaultChromiumLinux() {
  const desktopId = execText("xdg-settings", ["get", "default-web-browser"]) || execText("xdg-mime", ["query", "default", "x-scheme-handler/http"]);
  if (!desktopId) return null;
  const trimmed = desktopId.trim();
  if (!CHROMIUM_DESKTOP_IDS.has(trimmed)) return null;
  const searchDirs = [
    path.join(os.homedir(), ".local", "share", "applications"),
    "/usr/local/share/applications",
    "/usr/share/applications",
    "/var/lib/snapd/desktop/applications"
  ];
  let desktopPath = null;
  for (const dir of searchDirs) {
    const candidate = path.join(dir, trimmed);
    if (fileExists(candidate)) {
      desktopPath = candidate;
      break;
    }
  }
  if (!desktopPath) return null;
  let execLine = null;
  try {
    const lines = fs.readFileSync(desktopPath, "utf8").split(/\r?\n/);
    for (const line of lines) if (line.startsWith("Exec=")) {
      execLine = line.slice(5).trim();
      break;
    }
  } catch {
  }
  if (!execLine) return null;
  const tokens = execLine.split(/\s+/);
  let command = null;
  for (const token of tokens) {
    if (!token || token === "env" || token.includes("=") && !token.startsWith("/")) continue;
    command = token.replace(/^["']|["']$/g, "");
    break;
  }
  if (!command) return null;
  const resolved = command.startsWith("/") ? command : execText("which", [command], 800)?.trim() ?? null;
  if (!resolved) return null;
  const exeName = path.posix.basename(resolved).toLowerCase();
  if (!CHROMIUM_EXE_NAMES.has(exeName)) return null;
  return { kind: inferKindFromExeName(exeName), path: resolved };
}
function findChromeLinux() {
  return findFirstExe([
    { kind: "chrome", path: "/usr/bin/google-chrome" },
    { kind: "chrome", path: "/usr/bin/google-chrome-stable" },
    { kind: "chrome", path: "/usr/bin/chrome" },
    { kind: "brave", path: "/usr/bin/brave-browser" },
    { kind: "brave", path: "/usr/bin/brave-browser-stable" },
    { kind: "brave", path: "/usr/bin/brave" },
    { kind: "brave", path: "/snap/bin/brave" },
    { kind: "edge", path: "/usr/bin/microsoft-edge" },
    { kind: "edge", path: "/usr/bin/microsoft-edge-stable" },
    { kind: "chromium", path: "/usr/bin/chromium" },
    { kind: "chromium", path: "/usr/bin/chromium-browser" },
    { kind: "chromium", path: "/snap/bin/chromium" }
  ]);
}
function findChromeWindows() {
  const localAppData = process.env.LOCALAPPDATA ?? "";
  const programFiles = process.env.ProgramFiles ?? "C:\\Program Files";
  const programFilesX86 = process.env["ProgramFiles(x86)"] ?? "C:\\Program Files (x86)";
  const j = path.win32.join;
  const candidates = [];
  if (localAppData) {
    candidates.push({ kind: "chrome", path: j(localAppData, "Google", "Chrome", "Application", "chrome.exe") });
    candidates.push({ kind: "brave", path: j(localAppData, "BraveSoftware", "Brave-Browser", "Application", "brave.exe") });
    candidates.push({ kind: "edge", path: j(localAppData, "Microsoft", "Edge", "Application", "msedge.exe") });
    candidates.push({ kind: "chromium", path: j(localAppData, "Chromium", "Application", "chrome.exe") });
    candidates.push({ kind: "canary", path: j(localAppData, "Google", "Chrome SxS", "Application", "chrome.exe") });
  }
  candidates.push({ kind: "chrome", path: j(programFiles, "Google", "Chrome", "Application", "chrome.exe") });
  candidates.push({ kind: "chrome", path: j(programFilesX86, "Google", "Chrome", "Application", "chrome.exe") });
  candidates.push({ kind: "brave", path: j(programFiles, "BraveSoftware", "Brave-Browser", "Application", "brave.exe") });
  candidates.push({ kind: "brave", path: j(programFilesX86, "BraveSoftware", "Brave-Browser", "Application", "brave.exe") });
  candidates.push({ kind: "edge", path: j(programFiles, "Microsoft", "Edge", "Application", "msedge.exe") });
  candidates.push({ kind: "edge", path: j(programFilesX86, "Microsoft", "Edge", "Application", "msedge.exe") });
  return findFirstExe(candidates);
}
function resolveBrowserExecutable(opts) {
  if (opts?.executablePath) {
    if (!fileExists(opts.executablePath)) throw new Error(`executablePath not found: ${opts.executablePath}`);
    return { kind: "custom", path: opts.executablePath };
  }
  const platform = process.platform;
  if (platform === "darwin") return detectDefaultChromiumMac() ?? findChromeMac();
  if (platform === "linux") return detectDefaultChromiumLinux() ?? findChromeLinux();
  if (platform === "win32") return findChromeWindows();
  return null;
}
async function ensurePortAvailable(port) {
  await new Promise((resolve2, reject) => {
    const tester = net.createServer().once("error", (err) => {
      if (err.code === "EADDRINUSE") reject(new Error(`Port ${port} is already in use`));
      else reject(err);
    }).once("listening", () => {
      tester.close(() => resolve2());
    }).listen(port);
  });
}
function safeReadJson(filePath) {
  try {
    if (!fs.existsSync(filePath)) return null;
    const parsed = JSON.parse(fs.readFileSync(filePath, "utf-8"));
    if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) return null;
    return parsed;
  } catch {
    return null;
  }
}
function safeWriteJson(filePath, data) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  fs.writeFileSync(filePath, JSON.stringify(data, null, 2));
}
function setDeep(obj, keys, value) {
  let node = obj;
  for (const key of keys.slice(0, -1)) {
    const next = node[key];
    if (typeof next !== "object" || next === null || Array.isArray(next)) node[key] = {};
    node = node[key];
  }
  node[keys[keys.length - 1]] = value;
}
function parseHexRgbToSignedArgbInt(hex) {
  const cleaned = hex.trim().replace(/^#/, "");
  if (!/^[0-9a-fA-F]{6}$/.test(cleaned)) return null;
  const argbUnsigned = 255 << 24 | Number.parseInt(cleaned, 16);
  return argbUnsigned > 2147483647 ? argbUnsigned - 4294967296 : argbUnsigned;
}
function decorateProfile(userDataDir, name, color) {
  const colorInt = parseHexRgbToSignedArgbInt(color);
  const localStatePath = path.join(userDataDir, "Local State");
  const preferencesPath = path.join(userDataDir, "Default", "Preferences");
  const localState = safeReadJson(localStatePath) ?? {};
  setDeep(localState, ["profile", "info_cache", "Default", "name"], name);
  setDeep(localState, ["profile", "info_cache", "Default", "shortcut_name"], name);
  setDeep(localState, ["profile", "info_cache", "Default", "user_name"], name);
  setDeep(localState, ["profile", "info_cache", "Default", "profile_color"], color);
  if (colorInt != null) {
    setDeep(localState, ["profile", "info_cache", "Default", "profile_color_seed"], colorInt);
    setDeep(localState, ["profile", "info_cache", "Default", "profile_highlight_color"], colorInt);
  }
  safeWriteJson(localStatePath, localState);
  const prefs = safeReadJson(preferencesPath) ?? {};
  setDeep(prefs, ["profile", "name"], name);
  setDeep(prefs, ["profile", "profile_color"], color);
  if (colorInt != null) {
    setDeep(prefs, ["autogenerated", "theme", "color"], colorInt);
    setDeep(prefs, ["browser", "theme", "user_color2"], colorInt);
  }
  safeWriteJson(preferencesPath, prefs);
}
function ensureCleanExit(userDataDir) {
  const preferencesPath = path.join(userDataDir, "Default", "Preferences");
  const prefs = safeReadJson(preferencesPath) ?? {};
  setDeep(prefs, ["exit_type"], "Normal");
  setDeep(prefs, ["exited_cleanly"], true);
  safeWriteJson(preferencesPath, prefs);
}
var DEFAULT_CDP_PORT = 9222;
var DEFAULT_PROFILE_NAME = "browserclaw";
var DEFAULT_PROFILE_COLOR = "#FF4500";
function resolveUserDataDir(profileName) {
  const configDir = process.env.XDG_CONFIG_HOME ?? path.join(os.homedir(), ".config");
  return path.join(configDir, "browserclaw", "profiles", profileName, "user-data");
}
async function isChromeReachable(cdpUrl, timeoutMs = 500, authToken) {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const headers = {};
    if (authToken) headers["Authorization"] = `Bearer ${authToken}`;
    const res = await fetch(`${cdpUrl.replace(/\/+$/, "")}/json/version`, { signal: ctrl.signal, headers });
    if (!res.ok) return false;
    const data = await res.json();
    return data != null && typeof data === "object";
  } catch {
    return false;
  } finally {
    clearTimeout(t);
  }
}
async function getChromeWebSocketUrl(cdpUrl, timeoutMs = 500, authToken) {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const headers = {};
    if (authToken) headers["Authorization"] = `Bearer ${authToken}`;
    const res = await fetch(`${cdpUrl.replace(/\/+$/, "")}/json/version`, { signal: ctrl.signal, headers });
    if (!res.ok) return null;
    const data = await res.json();
    if (!data || typeof data !== "object") return null;
    return String(data?.webSocketDebuggerUrl ?? "").trim() || null;
  } catch {
    return null;
  } finally {
    clearTimeout(t);
  }
}
async function isChromeCdpReady(cdpUrl, timeoutMs = 500, handshakeTimeoutMs = 800) {
  const wsUrl = await getChromeWebSocketUrl(cdpUrl, timeoutMs);
  if (!wsUrl) return false;
  return await canRunCdpHealthCommand(wsUrl, handshakeTimeoutMs);
}
async function canRunCdpHealthCommand(wsUrl, timeoutMs = 800) {
  return new Promise((resolve2) => {
    let settled = false;
    const finish = (value) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      try {
        ws.close();
      } catch {
      }
      resolve2(value);
    };
    const timer = setTimeout(() => finish(false), Math.max(50, timeoutMs + 25));
    let ws;
    try {
      ws = new WebSocket(wsUrl);
    } catch {
      finish(false);
      return;
    }
    ws.onopen = () => {
      try {
        ws.send(JSON.stringify({ id: 1, method: "Browser.getVersion" }));
      } catch {
        finish(false);
      }
    };
    ws.onmessage = (event) => {
      try {
        const parsed = JSON.parse(String(event.data));
        if (parsed?.id !== 1) return;
        finish(Boolean(parsed.result && typeof parsed.result === "object"));
      } catch {
      }
    };
    ws.onerror = () => finish(false);
    ws.onclose = () => finish(false);
  });
}
async function launchChrome(opts = {}) {
  const cdpPort = opts.cdpPort ?? DEFAULT_CDP_PORT;
  await ensurePortAvailable(cdpPort);
  const exe = resolveBrowserExecutable({ executablePath: opts.executablePath });
  if (!exe) throw new Error("No supported browser found (Chrome/Brave/Edge/Chromium). Install one or provide executablePath.");
  const profileName = opts.profileName ?? DEFAULT_PROFILE_NAME;
  const userDataDir = opts.userDataDir ?? resolveUserDataDir(profileName);
  fs.mkdirSync(userDataDir, { recursive: true });
  const spawnChrome = () => {
    const args = [
      `--remote-debugging-port=${cdpPort}`,
      `--user-data-dir=${userDataDir}`,
      "--no-first-run",
      "--no-default-browser-check",
      "--disable-sync",
      "--disable-background-networking",
      "--disable-component-update",
      "--disable-features=Translate,MediaRouter",
      "--disable-blink-features=AutomationControlled",
      "--disable-session-crashed-bubble",
      "--hide-crash-restore-bubble",
      "--password-store=basic"
    ];
    if (opts.headless) {
      args.push("--headless=new", "--disable-gpu");
    }
    if (opts.noSandbox) {
      args.push("--no-sandbox", "--disable-setuid-sandbox");
    }
    if (process.platform === "linux") args.push("--disable-dev-shm-usage");
    const extraArgs = Array.isArray(opts.chromeArgs) ? opts.chromeArgs.filter((a) => typeof a === "string" && a.trim().length > 0) : [];
    if (extraArgs.length) args.push(...extraArgs);
    args.push("about:blank");
    return spawn(exe.path, args, {
      stdio: "pipe",
      env: { ...process.env, HOME: os.homedir() }
    });
  };
  const startedAt = Date.now();
  const localStatePath = path.join(userDataDir, "Local State");
  const preferencesPath = path.join(userDataDir, "Default", "Preferences");
  if (!fileExists(localStatePath) || !fileExists(preferencesPath)) {
    const bootstrap = spawnChrome();
    const deadline = Date.now() + 1e4;
    while (Date.now() < deadline) {
      if (fileExists(localStatePath) && fileExists(preferencesPath)) break;
      await new Promise((r) => setTimeout(r, 100));
    }
    try {
      bootstrap.kill("SIGTERM");
    } catch {
    }
    const exitDeadline = Date.now() + 5e3;
    while (Date.now() < exitDeadline) {
      if (bootstrap.exitCode != null) break;
      await new Promise((r) => setTimeout(r, 50));
    }
    if (bootstrap.exitCode == null) {
      try {
        bootstrap.kill("SIGKILL");
      } catch {
      }
    }
  }
  try {
    decorateProfile(userDataDir, profileName, opts.profileColor ?? DEFAULT_PROFILE_COLOR);
  } catch {
  }
  try {
    ensureCleanExit(userDataDir);
  } catch {
  }
  const proc = spawnChrome();
  const cdpUrl = `http://127.0.0.1:${cdpPort}`;
  const stderrChunks = [];
  const onStderr = (chunk) => {
    stderrChunks.push(chunk);
  };
  proc.stderr?.on("data", onStderr);
  const readyDeadline = Date.now() + 15e3;
  while (Date.now() < readyDeadline) {
    if (await isChromeReachable(cdpUrl, 500)) break;
    await new Promise((r) => setTimeout(r, 200));
  }
  if (!await isChromeReachable(cdpUrl, 500)) {
    const stderrOutput = Buffer.concat(stderrChunks).toString("utf8").trim();
    const stderrHint = stderrOutput ? `
Chrome stderr:
${stderrOutput.slice(0, 2e3)}` : "";
    const sandboxHint = process.platform === "linux" && !opts.noSandbox ? "\nHint: If running in a container or as root, try setting noSandbox: true." : "";
    try {
      proc.kill("SIGKILL");
    } catch {
    }
    throw new Error(`Failed to start Chrome CDP on port ${cdpPort}.${sandboxHint}${stderrHint}`);
  }
  proc.stderr?.off("data", onStderr);
  stderrChunks.length = 0;
  return {
    pid: proc.pid ?? -1,
    exe,
    userDataDir,
    cdpPort,
    startedAt,
    proc
  };
}
async function stopChrome(running, timeoutMs = 2500) {
  const proc = running.proc;
  if (proc.exitCode != null) return;
  try {
    proc.kill("SIGTERM");
  } catch {
  }
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    if (proc.exitCode != null) return;
    await new Promise((r) => setTimeout(r, 100));
  }
  try {
    proc.kill("SIGKILL");
  } catch {
  }
}
var cached = null;
var connectingByUrl = /* @__PURE__ */ new Map();
var pageStates = /* @__PURE__ */ new WeakMap();
var observedContexts = /* @__PURE__ */ new WeakSet();
var observedPages = /* @__PURE__ */ new WeakSet();
var roleRefsByTarget = /* @__PURE__ */ new Map();
var MAX_ROLE_REFS_CACHE = 50;
var MAX_CONSOLE_MESSAGES = 500;
var MAX_PAGE_ERRORS = 200;
var MAX_NETWORK_REQUESTS = 500;
function normalizeCdpUrl(raw) {
  return raw.replace(/\/$/, "");
}
function roleRefsKey(cdpUrl, targetId) {
  return `${normalizeCdpUrl(cdpUrl)}::${targetId}`;
}
function ensurePageState(page) {
  const existing = pageStates.get(page);
  if (existing) return existing;
  const state = {
    console: [],
    errors: [],
    requests: [],
    requestIds: /* @__PURE__ */ new WeakMap(),
    nextRequestId: 0
  };
  pageStates.set(page, state);
  if (!observedPages.has(page)) {
    observedPages.add(page);
    page.on("console", (msg) => {
      state.console.push({
        type: msg.type(),
        text: msg.text(),
        timestamp: (/* @__PURE__ */ new Date()).toISOString(),
        location: msg.location()
      });
      if (state.console.length > MAX_CONSOLE_MESSAGES) state.console.shift();
    });
    page.on("pageerror", (err) => {
      state.errors.push({
        message: err?.message ? String(err.message) : String(err),
        name: err?.name ? String(err.name) : void 0,
        stack: err?.stack ? String(err.stack) : void 0,
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      });
      if (state.errors.length > MAX_PAGE_ERRORS) state.errors.shift();
    });
    page.on("request", (req) => {
      state.nextRequestId += 1;
      const id = `r${state.nextRequestId}`;
      state.requestIds.set(req, id);
      state.requests.push({
        id,
        timestamp: (/* @__PURE__ */ new Date()).toISOString(),
        method: req.method(),
        url: req.url(),
        resourceType: req.resourceType()
      });
      if (state.requests.length > MAX_NETWORK_REQUESTS) state.requests.shift();
    });
    page.on("response", (resp) => {
      const req = resp.request();
      const id = state.requestIds.get(req);
      if (!id) return;
      for (let i = state.requests.length - 1; i >= 0; i--) {
        const rec = state.requests[i];
        if (rec && rec.id === id) {
          rec.status = resp.status();
          rec.ok = resp.ok();
          break;
        }
      }
    });
    page.on("requestfailed", (req) => {
      const id = state.requestIds.get(req);
      if (!id) return;
      for (let i = state.requests.length - 1; i >= 0; i--) {
        const rec = state.requests[i];
        if (rec && rec.id === id) {
          rec.failureText = req.failure()?.errorText;
          rec.ok = false;
          break;
        }
      }
    });
    page.on("close", () => {
      pageStates.delete(page);
      observedPages.delete(page);
    });
  }
  return state;
}
var STEALTH_SCRIPT = `Object.defineProperty(navigator, 'webdriver', { get: () => undefined })`;
function applyStealthToPage(page) {
  page.evaluate(STEALTH_SCRIPT).catch((e) => {
    if (process.env.DEBUG) console.warn("[browserclaw] stealth evaluate failed:", e.message);
  });
}
function observeContext(context) {
  if (observedContexts.has(context)) return;
  observedContexts.add(context);
  context.addInitScript(STEALTH_SCRIPT).catch((e) => {
    if (process.env.DEBUG) console.warn("[browserclaw] stealth initScript failed:", e.message);
  });
  for (const page of context.pages()) {
    ensurePageState(page);
    applyStealthToPage(page);
  }
  context.on("page", (page) => {
    ensurePageState(page);
    applyStealthToPage(page);
  });
}
function observeBrowser(browser) {
  for (const context of browser.contexts()) observeContext(context);
}
function storeRoleRefsForTarget(opts) {
  const state = ensurePageState(opts.page);
  state.roleRefs = opts.refs;
  state.roleRefsFrameSelector = opts.frameSelector;
  state.roleRefsMode = opts.mode;
  const targetId = opts.targetId?.trim();
  if (!targetId) return;
  roleRefsByTarget.set(roleRefsKey(opts.cdpUrl, targetId), {
    refs: opts.refs,
    ...opts.frameSelector ? { frameSelector: opts.frameSelector } : {},
    ...opts.mode ? { mode: opts.mode } : {}
  });
  while (roleRefsByTarget.size > MAX_ROLE_REFS_CACHE) {
    const first = roleRefsByTarget.keys().next();
    if (first.done) break;
    roleRefsByTarget.delete(first.value);
  }
}
function restoreRoleRefsForTarget(opts) {
  const targetId = opts.targetId?.trim() || "";
  if (!targetId) return;
  const entry = roleRefsByTarget.get(roleRefsKey(opts.cdpUrl, targetId));
  if (!entry) return;
  const state = ensurePageState(opts.page);
  if (state.roleRefs) return;
  state.roleRefs = entry.refs;
  state.roleRefsFrameSelector = entry.frameSelector;
  state.roleRefsMode = entry.mode;
}
async function connectBrowser(cdpUrl, authToken) {
  const normalized = normalizeCdpUrl(cdpUrl);
  if (cached?.cdpUrl === normalized) return cached;
  const existing = connectingByUrl.get(normalized);
  if (existing) return await existing;
  const connectWithRetry = async () => {
    let lastErr;
    for (let attempt = 0; attempt < 3; attempt++) {
      try {
        const timeout = 5e3 + attempt * 2e3;
        const endpoint = await getChromeWebSocketUrl(normalized, timeout, authToken).catch(() => null) ?? normalized;
        const headers = {};
        if (authToken) headers["Authorization"] = `Bearer ${authToken}`;
        const browser = await chromium.connectOverCDP(endpoint, { timeout, headers });
        const connected = { browser, cdpUrl: normalized, authToken };
        cached = connected;
        observeBrowser(browser);
        browser.on("disconnected", () => {
          if (cached?.browser === browser) cached = null;
          for (const key of roleRefsByTarget.keys()) {
            if (key.startsWith(normalized + "::")) roleRefsByTarget.delete(key);
          }
        });
        return connected;
      } catch (err) {
        lastErr = err;
        await new Promise((r) => setTimeout(r, 250 + attempt * 250));
      }
    }
    throw lastErr instanceof Error ? lastErr : new Error("CDP connect failed");
  };
  const promise = connectWithRetry().finally(() => {
    connectingByUrl.delete(normalized);
  });
  connectingByUrl.set(normalized, promise);
  return await promise;
}
async function disconnectBrowser() {
  if (connectingByUrl.size) {
    for (const p of connectingByUrl.values()) {
      try {
        await p;
      } catch {
      }
    }
  }
  const cur = cached;
  cached = null;
  if (cur) await cur.browser.close().catch(() => {
  });
}
async function getAllPages(browser) {
  return browser.contexts().flatMap((c) => c.pages());
}
async function pageTargetId(page) {
  const session = await page.context().newCDPSession(page);
  try {
    const info = await session.send("Target.getTargetInfo");
    const targetInfo = info.targetInfo;
    return String(targetInfo?.targetId ?? "").trim() || null;
  } finally {
    await session.detach().catch(() => {
    });
  }
}
async function findPageByTargetId(browser, targetId, cdpUrl) {
  const pages = await getAllPages(browser);
  let resolvedViaCdp = false;
  for (const page of pages) {
    let tid = null;
    try {
      tid = await pageTargetId(page);
      resolvedViaCdp = true;
    } catch {
      tid = null;
    }
    if (tid && tid === targetId) return page;
  }
  if (!resolvedViaCdp && pages.length === 1) {
    return pages[0];
  }
  if (cdpUrl) {
    try {
      const listUrl = `${cdpUrl.replace(/\/+$/, "").replace(/^ws:/, "http:").replace(/\/cdp$/, "")}/json/list`;
      const headers = {};
      if (cached?.authToken) headers["Authorization"] = `Bearer ${cached.authToken}`;
      const response = await fetch(listUrl, { headers });
      if (response.ok) {
        const targets = await response.json();
        const target = targets.find((t) => t.id === targetId);
        if (target) {
          const urlMatch = pages.filter((p) => p.url() === target.url);
          if (urlMatch.length === 1) return urlMatch[0];
          if (urlMatch.length > 1) {
            const sameUrlTargets = targets.filter((t) => t.url === target.url);
            if (sameUrlTargets.length === urlMatch.length) {
              const idx = sameUrlTargets.findIndex((t) => t.id === targetId);
              if (idx >= 0 && idx < urlMatch.length) return urlMatch[idx];
            }
          }
        }
      }
    } catch {
    }
  }
  return null;
}
async function getPageForTargetId(opts) {
  const { browser } = await connectBrowser(opts.cdpUrl);
  const pages = await getAllPages(browser);
  if (!pages.length) throw new Error("No pages available in the connected browser.");
  const first = pages[0];
  if (!opts.targetId) return first;
  const found = await findPageByTargetId(browser, opts.targetId, opts.cdpUrl);
  if (!found) {
    if (pages.length === 1) return first;
    throw new Error(`Tab not found (targetId: ${opts.targetId}). Call browser.tabs() to list open tabs.`);
  }
  return found;
}
function refLocator(page, ref) {
  const normalized = ref.startsWith("@") ? ref.slice(1) : ref.startsWith("ref=") ? ref.slice(4) : ref;
  if (!normalized.trim()) throw new Error("ref is required");
  if (/^e\d+$/.test(normalized)) {
    const state = pageStates.get(page);
    if (state?.roleRefsMode === "aria") {
      return (state.roleRefsFrameSelector ? page.frameLocator(state.roleRefsFrameSelector) : page).locator(`aria-ref=${normalized}`);
    }
    const info = state?.roleRefs?.[normalized];
    if (!info) throw new Error(`Unknown ref "${normalized}". Run a new snapshot and use a ref from that snapshot.`);
    const locAny = state?.roleRefsFrameSelector ? page.frameLocator(state.roleRefsFrameSelector) : page;
    const role = info.role;
    const locator = info.name ? locAny.getByRole(role, { name: info.name, exact: true }) : locAny.getByRole(role);
    return info.nth !== void 0 ? locator.nth(info.nth) : locator;
  }
  return page.locator(`aria-ref=${normalized}`);
}
function toAIFriendlyError(error, selector) {
  const message = error instanceof Error ? error.message : String(error);
  if (message.includes("strict mode violation")) {
    const countMatch = message.match(/resolved to (\d+) elements/);
    const count = countMatch ? countMatch[1] : "multiple";
    return new Error(`Selector "${selector}" matched ${count} elements. Run a new snapshot to get updated refs, or use a different ref.`);
  }
  if ((message.includes("Timeout") || message.includes("waiting for")) && (message.includes("to be visible") || message.includes("not visible"))) {
    return new Error(`Element "${selector}" not found or not visible. Run a new snapshot to see current page elements.`);
  }
  if (message.includes("intercepts pointer events") || message.includes("not visible") || message.includes("not receive pointer events")) {
    return new Error(`Element "${selector}" is not interactable (hidden or covered). Try scrolling it into view, closing overlays, or re-snapshotting.`);
  }
  return error instanceof Error ? error : new Error(message);
}
function normalizeTimeoutMs(timeoutMs, fallback, maxMs = 12e4) {
  return Math.max(500, Math.min(maxMs, timeoutMs ?? fallback));
}

// src/snapshot/ref-map.ts
var INTERACTIVE_ROLES = /* @__PURE__ */ new Set([
  "button",
  "link",
  "textbox",
  "checkbox",
  "radio",
  "combobox",
  "listbox",
  "menuitem",
  "menuitemcheckbox",
  "menuitemradio",
  "option",
  "searchbox",
  "slider",
  "spinbutton",
  "switch",
  "tab",
  "treeitem"
]);
var CONTENT_ROLES = /* @__PURE__ */ new Set([
  "heading",
  "cell",
  "gridcell",
  "columnheader",
  "rowheader",
  "listitem",
  "article",
  "region",
  "main",
  "navigation"
]);
var STRUCTURAL_ROLES = /* @__PURE__ */ new Set([
  "generic",
  "group",
  "list",
  "table",
  "row",
  "rowgroup",
  "grid",
  "treegrid",
  "menu",
  "menubar",
  "toolbar",
  "tablist",
  "tree",
  "directory",
  "document",
  "application",
  "presentation",
  "none"
]);
function getIndentLevel(line) {
  const match = line.match(/^(\s*)/);
  return match ? Math.floor(match[1].length / 2) : 0;
}
function matchInteractiveSnapshotLine(line, options) {
  const depth = getIndentLevel(line);
  if (options.maxDepth !== void 0 && depth > options.maxDepth) {
    return null;
  }
  const match = line.match(/^(\s*-\s*)(\w+)(?:\s+"([^"]*)")?(.*)$/);
  if (!match) {
    return null;
  }
  const [, , roleRaw, name, suffix] = match;
  if (roleRaw.startsWith("/")) {
    return null;
  }
  const role = roleRaw.toLowerCase();
  return {
    roleRaw,
    role,
    ...name ? { name } : {},
    suffix
  };
}
function createRoleNameTracker() {
  const counts = /* @__PURE__ */ new Map();
  const refsByKey = /* @__PURE__ */ new Map();
  return {
    counts,
    refsByKey,
    getKey(role, name) {
      return `${role}:${name ?? ""}`;
    },
    getNextIndex(role, name) {
      const key = this.getKey(role, name);
      const current = counts.get(key) ?? 0;
      counts.set(key, current + 1);
      return current;
    },
    trackRef(role, name, ref) {
      const key = this.getKey(role, name);
      const list = refsByKey.get(key) ?? [];
      list.push(ref);
      refsByKey.set(key, list);
    },
    getDuplicateKeys() {
      const out = /* @__PURE__ */ new Set();
      for (const [key, refs] of refsByKey) if (refs.length > 1) out.add(key);
      return out;
    }
  };
}
function removeNthFromNonDuplicates(refs, tracker) {
  const duplicates = tracker.getDuplicateKeys();
  for (const [ref, data] of Object.entries(refs)) {
    const key = tracker.getKey(data.role, data.name);
    if (!duplicates.has(key)) delete refs[ref]?.nth;
  }
}
function compactTree(tree) {
  const lines = tree.split("\n");
  const result = [];
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (line.includes("[ref=")) {
      result.push(line);
      continue;
    }
    if (line.includes(":") && !line.trimEnd().endsWith(":")) {
      result.push(line);
      continue;
    }
    const currentIndent = getIndentLevel(line);
    let hasRelevantChildren = false;
    for (let j = i + 1; j < lines.length; j++) {
      if (getIndentLevel(lines[j]) <= currentIndent) break;
      if (lines[j]?.includes("[ref=")) {
        hasRelevantChildren = true;
        break;
      }
    }
    if (hasRelevantChildren) result.push(line);
  }
  return result.join("\n");
}
function buildRoleSnapshotFromAriaSnapshot(ariaSnapshot, options = {}) {
  const lines = ariaSnapshot.split("\n");
  const refs = {};
  const tracker = createRoleNameTracker();
  let counter = 0;
  const nextRef = () => {
    counter++;
    return `e${counter}`;
  };
  if (options.interactive) {
    const result2 = [];
    for (const line of lines) {
      const parsed = matchInteractiveSnapshotLine(line, options);
      if (!parsed) continue;
      const { roleRaw, role, name, suffix } = parsed;
      if (!INTERACTIVE_ROLES.has(role)) continue;
      const prefix = line.match(/^(\s*-\s*)/)?.[1] ?? "";
      const ref = nextRef();
      const nth = tracker.getNextIndex(role, name);
      tracker.trackRef(role, name, ref);
      refs[ref] = { role, name, nth };
      let enhanced = `${prefix}${roleRaw}`;
      if (name) enhanced += ` "${name}"`;
      enhanced += ` [ref=${ref}]`;
      if (nth > 0) enhanced += ` [nth=${nth}]`;
      if (suffix.includes("[")) enhanced += suffix;
      result2.push(enhanced);
    }
    removeNthFromNonDuplicates(refs, tracker);
    return { snapshot: result2.join("\n") || "(no interactive elements)", refs };
  }
  const result = [];
  for (const line of lines) {
    const depth = getIndentLevel(line);
    if (options.maxDepth !== void 0 && depth > options.maxDepth) continue;
    const match = line.match(/^(\s*-\s*)(\w+)(?:\s+"([^"]*)")?(.*)$/);
    if (!match) {
      result.push(line);
      continue;
    }
    const [, prefix, roleRaw, name, suffix] = match;
    if (roleRaw.startsWith("/")) {
      result.push(line);
      continue;
    }
    const role = roleRaw.toLowerCase();
    const isInteractive = INTERACTIVE_ROLES.has(role);
    const isContent = CONTENT_ROLES.has(role);
    const isStructural = STRUCTURAL_ROLES.has(role);
    if (options.compact && isStructural && !name) continue;
    if (!(isInteractive || isContent && name)) {
      result.push(line);
      continue;
    }
    const ref = nextRef();
    const nth = tracker.getNextIndex(role, name);
    tracker.trackRef(role, name, ref);
    refs[ref] = { role, name, nth };
    let enhanced = `${prefix}${roleRaw}`;
    if (name) enhanced += ` "${name}"`;
    enhanced += ` [ref=${ref}]`;
    if (nth > 0) enhanced += ` [nth=${nth}]`;
    if (suffix) enhanced += suffix;
    result.push(enhanced);
  }
  removeNthFromNonDuplicates(refs, tracker);
  const tree = result.join("\n") || "(empty)";
  return { snapshot: options.compact ? compactTree(tree) : tree, refs };
}
function buildRoleSnapshotFromAiSnapshot(aiSnapshot, options = {}) {
  const lines = String(aiSnapshot ?? "").split("\n");
  const refs = {};
  function parseAiSnapshotRef(suffix) {
    const match = suffix.match(/\[ref=(e\d+)\]/i);
    return match ? match[1] : null;
  }
  if (options.interactive) {
    const out2 = [];
    for (const line of lines) {
      const parsed = matchInteractiveSnapshotLine(line, options);
      if (!parsed) continue;
      const { roleRaw, role, name, suffix } = parsed;
      if (!INTERACTIVE_ROLES.has(role)) continue;
      const ref = parseAiSnapshotRef(suffix);
      if (!ref) continue;
      const prefix = line.match(/^(\s*-\s*)/)?.[1] ?? "";
      refs[ref] = { role, ...name ? { name } : {} };
      out2.push(`${prefix}${roleRaw}${name ? ` "${name}"` : ""}${suffix}`);
    }
    return { snapshot: out2.join("\n") || "(no interactive elements)", refs };
  }
  const out = [];
  for (const line of lines) {
    const depth = getIndentLevel(line);
    if (options.maxDepth !== void 0 && depth > options.maxDepth) continue;
    const match = line.match(/^(\s*-\s*)(\w+)(?:\s+"([^"]*)")?(.*)$/);
    if (!match) {
      out.push(line);
      continue;
    }
    const [, , roleRaw, name, suffix] = match;
    if (roleRaw.startsWith("/")) {
      out.push(line);
      continue;
    }
    const role = roleRaw.toLowerCase();
    const isStructural = STRUCTURAL_ROLES.has(role);
    if (options.compact && isStructural && !name) continue;
    const ref = parseAiSnapshotRef(suffix);
    if (ref) refs[ref] = { role, ...name ? { name } : {} };
    out.push(line);
  }
  const tree = out.join("\n") || "(empty)";
  return { snapshot: options.compact ? compactTree(tree) : tree, refs };
}
function getRoleSnapshotStats(snapshot, refs) {
  const interactive = Object.values(refs).filter((r) => INTERACTIVE_ROLES.has(r.role)).length;
  return {
    lines: snapshot.split("\n").length,
    chars: snapshot.length,
    refs: Object.keys(refs).length,
    interactive
  };
}

// src/snapshot/ai-snapshot.ts
async function snapshotAi(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const maybe = page;
  if (!maybe._snapshotForAI) {
    throw new Error("Playwright _snapshotForAI is not available. Upgrade playwright-core to >= 1.50.");
  }
  const sourceUrl = page.url();
  const result = await maybe._snapshotForAI({
    timeout: normalizeTimeoutMs(opts.timeoutMs, 5e3, 6e4),
    track: "response"
  });
  let snapshot = String(result?.full ?? "");
  const maxChars = opts.maxChars;
  const limit = typeof maxChars === "number" && Number.isFinite(maxChars) && maxChars > 0 ? Math.floor(maxChars) : void 0;
  if (limit && snapshot.length > limit) {
    const lastNewline = snapshot.lastIndexOf("\n", limit);
    const cutoff = lastNewline > 0 ? lastNewline : limit;
    snapshot = `${snapshot.slice(0, cutoff)}

[...TRUNCATED - page too large]`;
  }
  const built = buildRoleSnapshotFromAiSnapshot(snapshot, opts.options);
  storeRoleRefsForTarget({
    page,
    cdpUrl: opts.cdpUrl,
    targetId: opts.targetId,
    refs: built.refs,
    mode: "aria"
  });
  return {
    snapshot: built.snapshot,
    refs: built.refs,
    stats: getRoleSnapshotStats(built.snapshot, built.refs),
    untrusted: true,
    contentMeta: {
      sourceUrl,
      contentType: "browser-snapshot",
      capturedAt: (/* @__PURE__ */ new Date()).toISOString()
    }
  };
}

// src/snapshot/aria-snapshot.ts
async function snapshotRole(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const sourceUrl = page.url();
  const frameSelector = opts.frameSelector?.trim() || "";
  const selector = opts.selector?.trim() || "";
  const locator = frameSelector ? selector ? page.frameLocator(frameSelector).locator(selector) : page.frameLocator(frameSelector).locator(":root") : selector ? page.locator(selector) : page.locator(":root");
  const ariaSnapshot = await locator.ariaSnapshot({ timeout: normalizeTimeoutMs(opts.timeoutMs, 5e3) });
  const built = buildRoleSnapshotFromAriaSnapshot(String(ariaSnapshot ?? ""), opts.options);
  storeRoleRefsForTarget({
    page,
    cdpUrl: opts.cdpUrl,
    targetId: opts.targetId,
    refs: built.refs,
    frameSelector: frameSelector || void 0,
    mode: opts.refsMode ?? "role"
  });
  return {
    snapshot: built.snapshot,
    refs: built.refs,
    stats: getRoleSnapshotStats(built.snapshot, built.refs),
    untrusted: true,
    contentMeta: {
      sourceUrl,
      contentType: "browser-snapshot",
      capturedAt: (/* @__PURE__ */ new Date()).toISOString()
    }
  };
}
async function snapshotAria(opts) {
  const limit = Math.max(1, Math.min(2e3, Math.floor(opts.limit ?? 500)));
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const sourceUrl = page.url();
  const session = await page.context().newCDPSession(page);
  try {
    await session.send("Accessibility.enable").catch(() => {
    });
    const res = await session.send("Accessibility.getFullAXTree");
    return {
      nodes: formatAriaNodes(Array.isArray(res?.nodes) ? res.nodes : [], limit),
      untrusted: true,
      contentMeta: {
        sourceUrl,
        contentType: "browser-aria-tree",
        capturedAt: (/* @__PURE__ */ new Date()).toISOString()
      }
    };
  } finally {
    await session.detach().catch(() => {
    });
  }
}
function axValue(v) {
  if (!v || typeof v !== "object") return "";
  const value = v.value;
  if (typeof value === "string") return value;
  if (typeof value === "number" || typeof value === "boolean") return String(value);
  return "";
}
function formatAriaNodes(nodes, limit) {
  const byId = /* @__PURE__ */ new Map();
  for (const n of nodes) if (n.nodeId) byId.set(n.nodeId, n);
  const referenced = /* @__PURE__ */ new Set();
  for (const n of nodes) for (const c of n.childIds ?? []) referenced.add(c);
  const root = nodes.find((n) => n.nodeId && !referenced.has(n.nodeId)) ?? nodes[0];
  if (!root?.nodeId) return [];
  const out = [];
  const stack = [{ id: root.nodeId, depth: 0 }];
  while (stack.length && out.length < limit) {
    const popped = stack.pop();
    if (!popped) break;
    const { id, depth } = popped;
    const n = byId.get(id);
    if (!n) continue;
    const role = axValue(n.role);
    const name = axValue(n.name);
    const value = axValue(n.value);
    const description = axValue(n.description);
    const ref = `ax${out.length + 1}`;
    out.push({
      ref,
      role: role || "unknown",
      name: name || "",
      ...value ? { value } : {},
      ...description ? { description } : {},
      ...typeof n.backendDOMNodeId === "number" ? { backendDOMNodeId: n.backendDOMNodeId } : {},
      depth
    });
    const children = (n.childIds ?? []).filter((c) => byId.has(c));
    for (let i = children.length - 1; i >= 0; i--) {
      if (children[i]) stack.push({ id: children[i], depth: depth + 1 });
    }
  }
  return out;
}
var InvalidBrowserNavigationUrlError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "InvalidBrowserNavigationUrlError";
  }
};
function withBrowserNavigationPolicy(ssrfPolicy) {
  return { ssrfPolicy };
}
var NETWORK_NAVIGATION_PROTOCOLS = /* @__PURE__ */ new Set(["http:", "https:"]);
var SAFE_NON_NETWORK_URLS = /* @__PURE__ */ new Set(["about:blank"]);
async function assertBrowserNavigationAllowed(opts) {
  const rawUrl = String(opts.url ?? "").trim();
  let parsed;
  try {
    parsed = new URL(rawUrl);
  } catch {
    throw new InvalidBrowserNavigationUrlError(`Invalid URL: "${rawUrl}"`);
  }
  if (!NETWORK_NAVIGATION_PROTOCOLS.has(parsed.protocol)) {
    if (SAFE_NON_NETWORK_URLS.has(parsed.href)) return;
    throw new InvalidBrowserNavigationUrlError(`Navigation blocked: unsupported protocol "${parsed.protocol}"`);
  }
  const policy = opts.ssrfPolicy;
  if (policy?.dangerouslyAllowPrivateNetwork ?? policy?.allowPrivateNetwork ?? true) return;
  const allowedHostnames = [
    ...policy?.allowedHostnames ?? [],
    ...policy?.hostnameAllowlist ?? []
  ];
  if (allowedHostnames.length) {
    const hostname = parsed.hostname.toLowerCase();
    if (allowedHostnames.some((h) => h.toLowerCase() === hostname)) return;
  }
  const ipOpts = { allowRfc2544BenchmarkRange: policy?.allowRfc2544BenchmarkRange };
  if (await isInternalUrlResolved(rawUrl, opts.lookupFn, ipOpts)) {
    throw new InvalidBrowserNavigationUrlError(
      `Navigation to internal/loopback address blocked: "${rawUrl}". ssrfPolicy.dangerouslyAllowPrivateNetwork is false (strict mode).`
    );
  }
}
async function assertSafeOutputPath(path2, allowedRoots) {
  if (!path2 || typeof path2 !== "string") {
    throw new Error("Output path is required.");
  }
  const normalized = normalize(path2);
  if (normalized.includes("..")) {
    throw new Error(`Unsafe output path: directory traversal detected in "${path2}".`);
  }
  if (allowedRoots?.length) {
    const resolved = resolve(normalized);
    let parentReal;
    try {
      parentReal = await realpath(dirname(resolved));
    } catch {
      throw new Error(`Unsafe output path: parent directory is inaccessible for "${path2}".`);
    }
    try {
      const targetStat = await lstat(resolved);
      if (targetStat.isSymbolicLink()) {
        throw new Error(`Unsafe output path: "${path2}" is a symbolic link.`);
      }
    } catch (e) {
      if (e.code !== "ENOENT") throw e;
    }
    const results = await Promise.all(
      allowedRoots.map(async (root) => {
        try {
          const rootStat = await lstat(resolve(root));
          if (!rootStat.isDirectory() || rootStat.isSymbolicLink()) return false;
          const rootReal = await realpath(resolve(root));
          return parentReal === rootReal || parentReal.startsWith(rootReal + sep);
        } catch {
          return false;
        }
      })
    );
    if (!results.some(Boolean)) {
      throw new Error(`Unsafe output path: "${path2}" is outside allowed directories.`);
    }
  }
}
function expandIPv6(ip) {
  let normalized = ip;
  const v4Match = normalized.match(/^(.+:)(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})$/);
  if (v4Match) {
    const octets = v4Match[2].split(".").map(Number);
    if (octets.some((o) => o > 255)) return null;
    const hexHi = (octets[0] << 8 | octets[1]).toString(16).padStart(4, "0");
    const hexLo = (octets[2] << 8 | octets[3]).toString(16).padStart(4, "0");
    normalized = v4Match[1] + hexHi + ":" + hexLo;
  }
  const halves = normalized.split("::");
  if (halves.length > 2) return null;
  if (halves.length === 2) {
    const left = halves[0] !== "" ? halves[0].split(":") : [];
    const right = halves[1] !== "" ? halves[1].split(":") : [];
    const needed = 8 - left.length - right.length;
    if (needed < 0) return null;
    const groups2 = [...left, ...Array(needed).fill("0"), ...right];
    if (groups2.length !== 8) return null;
    return groups2.map((g) => g.padStart(4, "0")).join(":");
  }
  const groups = normalized.split(":");
  if (groups.length !== 8) return null;
  return groups.map((g) => g.padStart(4, "0")).join(":");
}
function hexToIPv4(hiHex, loHex) {
  const hi = parseInt(hiHex, 16);
  const lo = parseInt(loHex, 16);
  return `${hi >> 8 & 255}.${hi & 255}.${lo >> 8 & 255}.${lo & 255}`;
}
function extractEmbeddedIPv4(lower) {
  if (lower.startsWith("::ffff:")) {
    return lower.slice(7);
  }
  const expanded = expandIPv6(lower);
  if (expanded === null) return "";
  const groups = expanded.split(":");
  if (groups.length !== 8) return "";
  if (groups[0] === "0064" && groups[1] === "ff9b" && groups[2] === "0000" && groups[3] === "0000" && groups[4] === "0000" && groups[5] === "0000") {
    return hexToIPv4(groups[6], groups[7]);
  }
  if (groups[0] === "0064" && groups[1] === "ff9b" && groups[2] === "0001") {
    return hexToIPv4(groups[6], groups[7]);
  }
  if (groups[0] === "2002") {
    return hexToIPv4(groups[1], groups[2]);
  }
  if (groups[0] === "2001" && groups[1] === "0000") {
    const hiXored = (parseInt(groups[6], 16) ^ 65535).toString(16).padStart(4, "0");
    const loXored = (parseInt(groups[7], 16) ^ 65535).toString(16).padStart(4, "0");
    return hexToIPv4(hiXored, loXored);
  }
  return null;
}
function isStrictDecimalOctet(part) {
  if (!/^[0-9]+$/.test(part)) return false;
  const n = parseInt(part, 10);
  if (n < 0 || n > 255) return false;
  if (String(n) !== part) return false;
  return true;
}
function isUnsupportedIPv4Literal(ip) {
  if (/^[0-9]+$/.test(ip)) return true;
  const parts = ip.split(".");
  if (parts.length !== 4) return true;
  if (!parts.every(isStrictDecimalOctet)) return true;
  return false;
}
function isInternalIP(ip, opts) {
  if (!ip.includes(":") && isUnsupportedIPv4Literal(ip)) return true;
  if (/^127\./.test(ip)) return true;
  if (/^10\./.test(ip)) return true;
  if (/^172\.(1[6-9]|2\d|3[01])\./.test(ip)) return true;
  if (/^192\.168\./.test(ip)) return true;
  if (/^169\.254\./.test(ip)) return true;
  if (/^100\.(6[4-9]|[7-9]\d|1[01]\d|12[0-7])\./.test(ip)) return true;
  if (ip === "0.0.0.0") return true;
  if (!opts?.allowRfc2544BenchmarkRange && /^198\.1[89]\./.test(ip)) return true;
  const lower = ip.toLowerCase();
  if (lower === "::1") return true;
  if (lower.startsWith("fe80:")) return true;
  if (lower.startsWith("fc") || lower.startsWith("fd")) return true;
  if (lower.startsWith("ff")) return true;
  const embedded = extractEmbeddedIPv4(lower);
  if (embedded !== null) {
    if (embedded === "") return true;
    return isInternalIP(embedded, opts);
  }
  return false;
}
function isInternalUrl(url, opts) {
  let parsed;
  try {
    parsed = new URL(url);
  } catch {
    return true;
  }
  const hostname = parsed.hostname.toLowerCase();
  if (hostname === "localhost") return true;
  if (isInternalIP(hostname, opts)) return true;
  if (hostname.endsWith(".local") || hostname.endsWith(".internal") || hostname.endsWith(".localhost")) {
    return true;
  }
  return false;
}
async function assertSafeUploadPaths(paths) {
  for (const filePath of paths) {
    let stat;
    try {
      stat = await lstat(filePath);
    } catch {
      throw new Error(`Upload path does not exist or is inaccessible: "${filePath}".`);
    }
    if (stat.isSymbolicLink()) {
      throw new Error(`Upload path is a symbolic link: "${filePath}".`);
    }
    if (!stat.isFile()) {
      throw new Error(`Upload path is not a regular file: "${filePath}".`);
    }
  }
}
async function isInternalUrlResolved(url, lookupFn = lookup, opts) {
  if (isInternalUrl(url, opts)) return true;
  let parsed;
  try {
    parsed = new URL(url);
  } catch {
    return true;
  }
  try {
    const { address } = await lookupFn(parsed.hostname);
    if (isInternalIP(address, opts)) return true;
  } catch {
    return true;
  }
  return false;
}
async function assertBrowserNavigationResultAllowed(opts) {
  const rawUrl = String(opts.url ?? "").trim();
  if (!rawUrl) return;
  let parsed;
  try {
    parsed = new URL(rawUrl);
  } catch {
    return;
  }
  if (NETWORK_NAVIGATION_PROTOCOLS.has(parsed.protocol) || SAFE_NON_NETWORK_URLS.has(parsed.href)) {
    await assertBrowserNavigationAllowed(opts);
  }
}

// src/actions/interaction.ts
async function clickViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  const locator = refLocator(page, opts.ref);
  const timeout = normalizeTimeoutMs(opts.timeoutMs, 8e3, 6e4);
  try {
    if (opts.doubleClick) {
      await locator.dblclick({ timeout, button: opts.button, modifiers: opts.modifiers });
    } else {
      await locator.click({ timeout, button: opts.button, modifiers: opts.modifiers });
    }
  } catch (err) {
    throw toAIFriendlyError(err, opts.ref);
  }
}
async function hoverViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  try {
    await refLocator(page, opts.ref).hover({
      timeout: normalizeTimeoutMs(opts.timeoutMs, 8e3, 6e4)
    });
  } catch (err) {
    throw toAIFriendlyError(err, opts.ref);
  }
}
async function typeViaPlaywright(opts) {
  const text = String(opts.text ?? "");
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  const locator = refLocator(page, opts.ref);
  const timeout = normalizeTimeoutMs(opts.timeoutMs, 8e3, 6e4);
  try {
    if (opts.slowly) {
      await locator.click({ timeout });
      await locator.pressSequentially(text, { timeout, delay: 75 });
    } else {
      await locator.fill(text, { timeout });
    }
    if (opts.submit) await locator.press("Enter", { timeout });
  } catch (err) {
    throw toAIFriendlyError(err, opts.ref);
  }
}
async function selectOptionViaPlaywright(opts) {
  if (!opts.values?.length) throw new Error("values are required");
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  try {
    await refLocator(page, opts.ref).selectOption(opts.values, {
      timeout: normalizeTimeoutMs(opts.timeoutMs, 8e3, 6e4)
    });
  } catch (err) {
    throw toAIFriendlyError(err, opts.ref);
  }
}
async function dragViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  try {
    await refLocator(page, opts.startRef).dragTo(refLocator(page, opts.endRef), {
      timeout: normalizeTimeoutMs(opts.timeoutMs, 8e3, 6e4)
    });
  } catch (err) {
    throw toAIFriendlyError(err, `${opts.startRef} -> ${opts.endRef}`);
  }
}
async function fillFormViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  const timeout = normalizeTimeoutMs(opts.timeoutMs, 8e3, 6e4);
  for (let i = 0; i < opts.fields.length; i++) {
    const field = opts.fields[i];
    const ref = field.ref.trim();
    const type = (typeof field.type === "string" ? field.type.trim() : "") || "text";
    const rawValue = field.value;
    const value = rawValue == null ? "" : String(rawValue);
    if (!ref) throw new Error(`fill(): field at index ${i} has empty ref`);
    const locator = refLocator(page, ref);
    if (type === "checkbox" || type === "radio") {
      const checked = rawValue === true || rawValue === 1 || rawValue === "1" || rawValue === "true";
      try {
        await locator.setChecked(checked, { timeout });
      } catch (err) {
        throw toAIFriendlyError(err, ref);
      }
      continue;
    }
    try {
      await locator.fill(value, { timeout });
    } catch (err) {
      throw toAIFriendlyError(err, ref);
    }
  }
}
async function scrollIntoViewViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  try {
    await refLocator(page, opts.ref).scrollIntoViewIfNeeded({
      timeout: normalizeTimeoutMs(opts.timeoutMs, 2e4)
    });
  } catch (err) {
    throw toAIFriendlyError(err, opts.ref);
  }
}
async function highlightViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  try {
    await refLocator(page, opts.ref).highlight();
  } catch (err) {
    throw toAIFriendlyError(err, opts.ref);
  }
}
async function setInputFilesViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  const locator = opts.ref ? refLocator(page, opts.ref) : opts.element ? page.locator(opts.element).first() : null;
  if (!locator) throw new Error("Either ref or element is required for setInputFiles");
  await assertSafeUploadPaths(opts.paths);
  try {
    await locator.setInputFiles(opts.paths);
  } catch (err) {
    throw toAIFriendlyError(err, opts.ref ?? opts.element ?? "unknown");
  }
}
async function armDialogViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const timeout = normalizeTimeoutMs(opts.timeoutMs, 3e4, 12e4);
  return new Promise((resolve2, reject) => {
    const timer = setTimeout(() => {
      page.removeListener("dialog", handler);
      reject(new Error(`No dialog appeared within ${timeout}ms`));
    }, timeout);
    const handler = async (dialog) => {
      clearTimeout(timer);
      try {
        if (opts.accept) {
          await dialog.accept(opts.promptText);
        } else {
          await dialog.dismiss();
        }
        resolve2();
      } catch (err) {
        reject(err);
      }
    };
    page.once("dialog", handler);
  });
}
async function armFileUploadViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const timeout = normalizeTimeoutMs(opts.timeoutMs, 3e4, 12e4);
  return new Promise((resolve2, reject) => {
    const timer = setTimeout(() => {
      page.removeListener("filechooser", handler);
      reject(new Error(`No file chooser appeared within ${timeout}ms`));
    }, timeout);
    const handler = async (fc) => {
      clearTimeout(timer);
      try {
        const paths = opts.paths ?? [];
        if (paths.length > 0) await assertSafeUploadPaths(paths);
        await fc.setFiles(paths);
        resolve2();
      } catch (err) {
        reject(err);
      }
    };
    page.once("filechooser", handler);
  });
}

// src/actions/keyboard.ts
async function pressKeyViaPlaywright(opts) {
  const key = String(opts.key ?? "").trim();
  if (!key) throw new Error("key is required");
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  await page.keyboard.press(key, { delay: Math.max(0, Math.floor(opts.delayMs ?? 0)) });
}

// src/actions/navigation.ts
async function navigateViaPlaywright(opts) {
  const url = String(opts.url ?? "").trim();
  if (!url) throw new Error("url is required");
  const policy = opts.allowInternal ? { ...opts.ssrfPolicy, dangerouslyAllowPrivateNetwork: true } : opts.ssrfPolicy;
  await assertBrowserNavigationAllowed({ url, ssrfPolicy: policy });
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  await page.goto(url, { timeout: normalizeTimeoutMs(opts.timeoutMs, 2e4) });
  const finalUrl = page.url();
  await assertBrowserNavigationResultAllowed({ url: finalUrl, ssrfPolicy: policy });
  return { url: finalUrl };
}
async function listPagesViaPlaywright(opts) {
  const { browser } = await connectBrowser(opts.cdpUrl);
  const pages = await getAllPages(browser);
  const results = [];
  for (const page of pages) {
    const tid = await pageTargetId(page).catch(() => null);
    if (tid) results.push({
      targetId: tid,
      title: await page.title().catch(() => ""),
      url: page.url(),
      type: "page"
    });
  }
  return results;
}
async function createPageViaPlaywright(opts) {
  const targetUrl = (opts.url ?? "").trim() || "about:blank";
  const policy = opts.allowInternal ? { ...opts.ssrfPolicy, dangerouslyAllowPrivateNetwork: true } : opts.ssrfPolicy;
  if (targetUrl !== "about:blank") {
    await assertBrowserNavigationAllowed({ url: targetUrl, ssrfPolicy: policy });
  }
  const { browser } = await connectBrowser(opts.cdpUrl);
  const context = browser.contexts()[0] ?? await browser.newContext();
  const page = await context.newPage();
  ensurePageState(page);
  if (targetUrl !== "about:blank") {
    await page.goto(targetUrl, { timeout: normalizeTimeoutMs(void 0, 2e4) });
    await assertBrowserNavigationResultAllowed({ url: page.url(), ssrfPolicy: policy });
  }
  const tid = await pageTargetId(page).catch(() => null);
  if (!tid) throw new Error("Failed to get targetId for new page");
  return {
    targetId: tid,
    title: await page.title().catch(() => ""),
    url: page.url(),
    type: "page"
  };
}
async function closePageByTargetIdViaPlaywright(opts) {
  const { browser } = await connectBrowser(opts.cdpUrl);
  const page = await findPageByTargetId(browser, opts.targetId, opts.cdpUrl);
  if (!page) throw new Error(`Tab not found (targetId: ${opts.targetId}). Use browser.tabs() to list open tabs.`);
  await page.close();
}
async function focusPageByTargetIdViaPlaywright(opts) {
  const { browser } = await connectBrowser(opts.cdpUrl);
  const page = await findPageByTargetId(browser, opts.targetId, opts.cdpUrl);
  if (!page) throw new Error(`Tab not found (targetId: ${opts.targetId}). Use browser.tabs() to list open tabs.`);
  try {
    await page.bringToFront();
  } catch (err) {
    const session = await page.context().newCDPSession(page);
    try {
      await session.send("Page.bringToFront");
    } catch {
      throw err;
    } finally {
      await session.detach().catch(() => {
      });
    }
  }
}
async function resizeViewportViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  await page.setViewportSize({
    width: Math.max(1, Math.floor(opts.width)),
    height: Math.max(1, Math.floor(opts.height))
  });
}

// src/actions/wait.ts
async function waitForViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const timeout = normalizeTimeoutMs(opts.timeoutMs, 2e4);
  if (typeof opts.timeMs === "number" && Number.isFinite(opts.timeMs)) {
    await page.waitForTimeout(Math.max(0, opts.timeMs));
  }
  if (opts.text) {
    await page.getByText(opts.text).first().waitFor({ state: "visible", timeout });
  }
  if (opts.textGone) {
    await page.getByText(opts.textGone).first().waitFor({ state: "hidden", timeout });
  }
  if (opts.selector) {
    const selector = String(opts.selector).trim();
    if (selector) await page.locator(selector).first().waitFor({ state: "visible", timeout });
  }
  if (opts.url) {
    const url = String(opts.url).trim();
    if (url) await page.waitForURL(url, { timeout });
  }
  if (opts.loadState) {
    await page.waitForLoadState(opts.loadState, { timeout });
  }
  if (opts.fn) {
    const fn = String(opts.fn).trim();
    if (fn) await page.waitForFunction(fn, void 0, { timeout });
  }
}

// src/actions/evaluate.ts
async function evaluateInAllFramesViaPlaywright(opts) {
  const fnText = String(opts.fn ?? "").trim();
  if (!fnText) throw new Error("function is required");
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  const frames = page.frames();
  const results = [];
  for (const frame of frames) {
    try {
      const result = await frame.evaluate(
        // eslint-disable-next-line no-eval
        (fnBody) => {
          "use strict";
          try {
            const candidate = (0, eval)("(" + fnBody + ")");
            return typeof candidate === "function" ? candidate() : candidate;
          } catch (err) {
            throw new Error("Invalid evaluate function: " + (err instanceof Error ? err.message : String(err)));
          }
        },
        fnText
      );
      results.push({
        frameUrl: frame.url(),
        frameName: frame.name(),
        result
      });
    } catch {
    }
  }
  return results;
}
async function evaluateViaPlaywright(opts) {
  const fnText = String(opts.fn ?? "").trim();
  if (!fnText) throw new Error("function is required");
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  const timeout = opts.timeoutMs != null ? opts.timeoutMs : void 0;
  if (opts.ref) {
    const locator = refLocator(page, opts.ref);
    return await locator.evaluate(
      // eslint-disable-next-line no-eval
      (el, fnBody) => {
        try {
          const candidate = (0, eval)("(" + fnBody + ")");
          return typeof candidate === "function" ? candidate(el) : candidate;
        } catch (err) {
          throw new Error("Invalid evaluate function: " + (err instanceof Error ? err.message : String(err)));
        }
      },
      fnText,
      { timeout }
    );
  }
  const evalPromise = page.evaluate(
    // eslint-disable-next-line no-eval
    (fnBody) => {
      try {
        const candidate = (0, eval)("(" + fnBody + ")");
        return typeof candidate === "function" ? candidate() : candidate;
      } catch (err) {
        throw new Error("Invalid evaluate function: " + (err instanceof Error ? err.message : String(err)));
      }
    },
    fnText
  );
  if (!opts.signal) return evalPromise;
  return Promise.race([
    evalPromise,
    new Promise((_, reject) => {
      opts.signal.addEventListener("abort", () => reject(new Error("Evaluate aborted")), { once: true });
    })
  ]);
}

// src/actions/download.ts
async function downloadViaPlaywright(opts) {
  await assertSafeOutputPath(opts.path, opts.allowedOutputRoots);
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  const timeout = normalizeTimeoutMs(opts.timeoutMs, 3e4, 12e4);
  const locator = refLocator(page, opts.ref);
  try {
    const [download] = await Promise.all([
      page.waitForEvent("download", { timeout }),
      locator.click({ timeout })
    ]);
    await download.saveAs(opts.path);
    return {
      url: download.url(),
      suggestedFilename: download.suggestedFilename(),
      path: opts.path
    };
  } catch (err) {
    throw toAIFriendlyError(err, opts.ref);
  }
}
async function waitForDownloadViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const timeout = normalizeTimeoutMs(opts.timeoutMs, 3e4, 12e4);
  const download = await page.waitForEvent("download", { timeout });
  const savePath = opts.path ?? download.suggestedFilename();
  await assertSafeOutputPath(savePath, opts.allowedOutputRoots);
  await download.saveAs(savePath);
  return {
    url: download.url(),
    suggestedFilename: download.suggestedFilename(),
    path: savePath
  };
}
async function emulateMediaViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  await page.emulateMedia({ colorScheme: opts.colorScheme });
}
async function setDeviceViaPlaywright(opts) {
  const device = devices[opts.name];
  if (!device) {
    const available = Object.keys(devices).slice(0, 10).join(", ");
    throw new Error(`Unknown device "${opts.name}". Some available devices: ${available}...`);
  }
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  if (device.viewport) {
    await page.setViewportSize(device.viewport);
  }
  if (device.userAgent) {
    const context = page.context();
    const session = await context.newCDPSession(page);
    try {
      await session.send("Emulation.setUserAgentOverride", {
        userAgent: device.userAgent
      });
    } finally {
      await session.detach().catch(() => {
      });
    }
  }
}
async function setExtraHTTPHeadersViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  await page.setExtraHTTPHeaders(opts.headers);
}
async function setGeolocationViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const context = page.context();
  if (opts.clear) {
    await context.setGeolocation(null);
    await context.clearPermissions();
    return;
  }
  if (opts.latitude === void 0 || opts.longitude === void 0) {
    throw new Error("latitude and longitude are required (or set clear=true)");
  }
  await context.grantPermissions(["geolocation"], opts.origin ? { origin: opts.origin } : void 0);
  await context.setGeolocation({
    latitude: opts.latitude,
    longitude: opts.longitude,
    accuracy: opts.accuracy
  });
}
async function setHttpCredentialsViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const context = page.context();
  if (opts.clear) {
    await context.setHTTPCredentials({ username: "", password: "" });
    return;
  }
  await context.setHTTPCredentials({
    username: opts.username ?? "",
    password: opts.password ?? ""
  });
}
async function setLocaleViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const session = await page.context().newCDPSession(page);
  try {
    await session.send("Emulation.setLocaleOverride", { locale: opts.locale });
  } finally {
    await session.detach().catch(() => {
    });
  }
}
async function setOfflineViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  await page.context().setOffline(opts.offline);
}
async function setTimezoneViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const session = await page.context().newCDPSession(page);
  try {
    await session.send("Emulation.setTimezoneOverride", { timezoneId: opts.timezoneId });
  } finally {
    await session.detach().catch(() => {
    });
  }
}

// src/capture/screenshot.ts
async function takeScreenshotViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  const type = opts.type ?? "png";
  if (opts.ref) {
    if (opts.fullPage) throw new Error("fullPage is not supported for element screenshots");
    return { buffer: await refLocator(page, opts.ref).screenshot({ type }) };
  }
  if (opts.element) {
    if (opts.fullPage) throw new Error("fullPage is not supported for element screenshots");
    return { buffer: await page.locator(opts.element).first().screenshot({ type }) };
  }
  return { buffer: await page.screenshot({ type, fullPage: Boolean(opts.fullPage) }) };
}
async function screenshotWithLabelsViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  restoreRoleRefsForTarget({ cdpUrl: opts.cdpUrl, targetId: opts.targetId, page });
  const maxLabels = opts.maxLabels ?? 50;
  const type = opts.type ?? "png";
  const refs = opts.refs.slice(0, maxLabels);
  const skipped = opts.refs.slice(maxLabels);
  const labels = [];
  for (let i = 0; i < refs.length; i++) {
    const ref = refs[i];
    try {
      const locator = refLocator(page, ref);
      const box = await locator.boundingBox({ timeout: 2e3 });
      if (box) {
        labels.push({ ref, index: i + 1, box });
      } else {
        skipped.push(ref);
      }
    } catch {
      skipped.push(ref);
    }
  }
  await page.evaluate((labelData) => {
    const container = document.createElement("div");
    container.id = "__browserclaw_labels__";
    container.style.cssText = "position:fixed;top:0;left:0;width:100%;height:100%;pointer-events:none;z-index:2147483647;";
    for (const { index, box } of labelData) {
      const border = document.createElement("div");
      border.style.cssText = `position:absolute;left:${box.x}px;top:${box.y}px;width:${box.width}px;height:${box.height}px;border:2px solid #FF4500;box-sizing:border-box;`;
      container.appendChild(border);
      const badge = document.createElement("div");
      badge.textContent = String(index);
      badge.style.cssText = `position:absolute;left:${box.x}px;top:${Math.max(0, box.y - 18)}px;background:#FF4500;color:#fff;font:bold 12px/16px monospace;padding:0 4px;border-radius:2px;`;
      container.appendChild(badge);
    }
    document.body.appendChild(container);
  }, labels.map((l) => ({ index: l.index, box: l.box })));
  const buffer = await page.screenshot({ type });
  await page.evaluate(() => {
    const el = document.getElementById("__browserclaw_labels__");
    if (el) el.remove();
  }).catch(() => {
  });
  return { buffer, labels, skipped };
}

// src/capture/pdf.ts
async function pdfViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  return { buffer: await page.pdf({ printBackground: true }) };
}

// src/capture/trace.ts
async function traceStartViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const context = page.context();
  await context.tracing.start({
    screenshots: opts.screenshots ?? true,
    snapshots: opts.snapshots ?? true,
    sources: opts.sources
  });
}
async function traceStopViaPlaywright(opts) {
  await assertSafeOutputPath(opts.path, opts.allowedOutputRoots);
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const context = page.context();
  await context.tracing.stop({ path: opts.path });
}

// src/capture/response.ts
async function responseBodyViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const timeout = normalizeTimeoutMs(opts.timeoutMs, 3e4, 12e4);
  const response = await page.waitForResponse(opts.url, { timeout });
  let body = await response.text();
  let truncated = false;
  const maxChars = typeof opts.maxChars === "number" && Number.isFinite(opts.maxChars) ? Math.max(1, Math.min(5e6, Math.floor(opts.maxChars))) : void 0;
  if (maxChars !== void 0 && body.length > maxChars) {
    body = body.slice(0, maxChars);
    truncated = true;
  }
  const headers = {};
  const allHeaders = response.headers();
  for (const [key, value] of Object.entries(allHeaders)) {
    headers[key] = value;
  }
  return {
    url: response.url(),
    status: response.status(),
    headers,
    body,
    truncated
  };
}

// src/capture/activity.ts
function consolePriority(level) {
  switch (level) {
    case "error":
      return 3;
    case "warning":
    case "warn":
      return 2;
    case "info":
    case "log":
      return 1;
    case "debug":
      return 0;
    default:
      return 1;
  }
}
async function getConsoleMessagesViaPlaywright(opts) {
  const state = ensurePageState(await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId }));
  const messages = opts.level ? state.console.filter((msg) => consolePriority(msg.type) >= consolePriority(opts.level)) : [...state.console];
  if (opts.clear) state.console = [];
  return messages;
}
async function getPageErrorsViaPlaywright(opts) {
  const state = ensurePageState(await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId }));
  const errors = [...state.errors];
  if (opts.clear) state.errors = [];
  return { errors };
}
async function getNetworkRequestsViaPlaywright(opts) {
  const state = ensurePageState(await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId }));
  const raw = [...state.requests];
  const filter = typeof opts.filter === "string" ? opts.filter.trim() : "";
  const requests = filter ? raw.filter((r) => r.url.includes(filter)) : raw;
  if (opts.clear) {
    state.requests = [];
    state.requestIds = /* @__PURE__ */ new WeakMap();
  }
  return { requests };
}

// src/storage/index.ts
async function cookiesGetViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  return { cookies: await page.context().cookies() };
}
async function cookiesSetViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  const cookie = opts.cookie;
  if (!cookie.name || cookie.value === void 0) throw new Error("cookie name and value are required");
  const hasUrl = typeof cookie.url === "string" && cookie.url.trim();
  const hasDomain = typeof cookie.domain === "string" && cookie.domain.trim();
  if (!hasUrl && !hasDomain) throw new Error("cookie requires url or domain");
  await page.context().addCookies([cookie]);
}
async function cookiesClearViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  await page.context().clearCookies();
}
async function storageGetViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  return {
    values: await page.evaluate(
      ({ kind, key }) => {
        const store = kind === "session" ? window.sessionStorage : window.localStorage;
        if (key) {
          const value = store.getItem(key);
          return value === null ? {} : { [key]: value };
        }
        const out = {};
        for (let i = 0; i < store.length; i++) {
          const k = store.key(i);
          if (!k) continue;
          const v = store.getItem(k);
          if (v !== null) out[k] = v;
        }
        return out;
      },
      { kind: opts.kind, key: opts.key }
    ) ?? {}
  };
}
async function storageSetViaPlaywright(opts) {
  const key = String(opts.key ?? "");
  if (!key) throw new Error("key is required");
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  await page.evaluate(
    ({ kind, key: k, value }) => {
      (kind === "session" ? window.sessionStorage : window.localStorage).setItem(k, value);
    },
    { kind: opts.kind, key, value: String(opts.value ?? "") }
  );
}
async function storageClearViaPlaywright(opts) {
  const page = await getPageForTargetId({ cdpUrl: opts.cdpUrl, targetId: opts.targetId });
  ensurePageState(page);
  await page.evaluate(
    ({ kind }) => {
      (kind === "session" ? window.sessionStorage : window.localStorage).clear();
    },
    { kind: opts.kind }
  );
}

// src/browser.ts
var CrawlPage = class {
  cdpUrl;
  targetId;
  ssrfPolicy;
  /** @internal */
  constructor(cdpUrl, targetId, ssrfPolicy) {
    this.cdpUrl = cdpUrl;
    this.targetId = targetId;
    this.ssrfPolicy = ssrfPolicy;
  }
  /** The CDP target ID for this page. Use this to identify the page in multi-tab scenarios. */
  get id() {
    return this.targetId;
  }
  // ── Snapshot ──────────────────────────────────────────────────
  /**
   * Take an AI-readable snapshot of the page.
   *
   * Returns a text tree with numbered refs (`e1`, `e2`, ...) that map to
   * interactive elements. Use these refs with actions like `click()` and `type()`.
   *
   * @param opts - Snapshot options (mode, filtering, depth limits)
   * @returns Snapshot text, ref map, and statistics
   *
   * @example
   * ```ts
   * // Default snapshot (aria mode)
   * const { snapshot, refs } = await page.snapshot();
   *
   * // Interactive elements only, compact
   * const result = await page.snapshot({ interactive: true, compact: true });
   *
   * // Role-based mode (uses getByRole resolution)
   * const result = await page.snapshot({ mode: 'role' });
   * ```
   */
  async snapshot(opts) {
    if (opts?.mode === "role") {
      return snapshotRole({
        cdpUrl: this.cdpUrl,
        targetId: this.targetId,
        selector: opts.selector,
        frameSelector: opts.frameSelector,
        refsMode: opts.refsMode,
        timeoutMs: opts.timeoutMs,
        options: {
          interactive: opts.interactive,
          compact: opts.compact,
          maxDepth: opts.maxDepth
        }
      });
    }
    if (opts?.selector || opts?.frameSelector) {
      throw new Error('selector and frameSelector are only supported in role mode. Use { mode: "role" } or omit these options.');
    }
    return snapshotAi({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      maxChars: opts?.maxChars,
      options: {
        interactive: opts?.interactive,
        compact: opts?.compact,
        maxDepth: opts?.maxDepth
      }
    });
  }
  /**
   * Take a raw ARIA accessibility tree snapshot via CDP.
   *
   * Unlike `snapshot()`, this returns structured node data rather than
   * an AI-readable text tree. Useful for programmatic accessibility analysis.
   *
   * @param opts - Options (limit: max nodes to return, default 500)
   * @returns Array of accessibility tree nodes
   */
  async ariaSnapshot(opts) {
    return snapshotAria({ cdpUrl: this.cdpUrl, targetId: this.targetId, limit: opts?.limit });
  }
  // ── Interactions ─────────────────────────────────────────────
  /**
   * Click an element by ref.
   *
   * @param ref - Ref ID from a snapshot (e.g. `'e1'`)
   * @param opts - Click options (double-click, button, modifiers)
   *
   * @example
   * ```ts
   * await page.click('e1');
   * await page.click('e2', { doubleClick: true });
   * await page.click('e3', { button: 'right' });
   * await page.click('e4', { modifiers: ['Control'] });
   * ```
   */
  async click(ref, opts) {
    return clickViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      ref,
      doubleClick: opts?.doubleClick,
      button: opts?.button,
      modifiers: opts?.modifiers,
      timeoutMs: opts?.timeoutMs
    });
  }
  /**
   * Type text into an input element by ref.
   *
   * By default, uses Playwright's `fill()` for instant input. Use `slowly: true`
   * to simulate real keystroke typing with a 75ms delay per character.
   *
   * @param ref - Ref ID of the input element (e.g. `'e3'`)
   * @param text - Text to type
   * @param opts - Type options (submit, slowly)
   *
   * @example
   * ```ts
   * await page.type('e3', 'hello world');
   * await page.type('e3', 'slow typing', { slowly: true });
   * await page.type('e3', 'search query', { submit: true }); // press Enter after
   * ```
   */
  async type(ref, text, opts) {
    return typeViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      ref,
      text,
      submit: opts?.submit,
      slowly: opts?.slowly,
      timeoutMs: opts?.timeoutMs
    });
  }
  /**
   * Hover over an element by ref.
   *
   * @param ref - Ref ID from a snapshot
   * @param opts - Timeout options
   */
  async hover(ref, opts) {
    return hoverViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      ref,
      timeoutMs: opts?.timeoutMs
    });
  }
  /**
   * Select option(s) in a `<select>` dropdown by ref.
   *
   * @param ref - Ref ID of the select element
   * @param values - One or more option labels/values to select
   *
   * @example
   * ```ts
   * await page.select('e5', 'Option A');
   * await page.select('e5', 'Option A', 'Option B'); // multi-select
   * ```
   */
  async select(ref, ...values) {
    return selectOptionViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      ref,
      values
    });
  }
  /**
   * Drag one element to another.
   *
   * @param startRef - Ref ID of the element to drag
   * @param endRef - Ref ID of the drop target
   * @param opts - Timeout options
   */
  async drag(startRef, endRef, opts) {
    return dragViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      startRef,
      endRef,
      timeoutMs: opts?.timeoutMs
    });
  }
  /**
   * Fill multiple form fields at once.
   *
   * Supports text inputs, checkboxes, and radio buttons.
   *
   * @param fields - Array of form fields to fill
   *
   * @example
   * ```ts
   * await page.fill([
   *   { ref: 'e2', type: 'text', value: 'Jane Doe' },
   *   { ref: 'e4', type: 'text', value: 'jane@example.com' },
   *   { ref: 'e6', type: 'checkbox', value: true },
   * ]);
   * ```
   */
  async fill(fields) {
    return fillFormViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      fields
    });
  }
  /**
   * Scroll an element into the visible viewport.
   *
   * @param ref - Ref ID of the element to scroll to
   * @param opts - Timeout options
   */
  async scrollIntoView(ref, opts) {
    return scrollIntoViewViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      ref,
      timeoutMs: opts?.timeoutMs
    });
  }
  /**
   * Highlight an element in the browser (Playwright built-in highlight).
   *
   * @param ref - Ref ID of the element to highlight
   */
  async highlight(ref) {
    return highlightViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      ref
    });
  }
  /**
   * Set files on an `<input type="file">` element.
   *
   * @param ref - Ref ID of the file input element
   * @param paths - Array of file paths to upload
   */
  async uploadFile(ref, paths) {
    return setInputFilesViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      ref,
      paths
    });
  }
  /**
   * Arm a one-shot dialog handler (alert, confirm, prompt).
   *
   * Returns a promise — store it (don't await), trigger the dialog, then await it.
   *
   * @param opts - Dialog options (accept/dismiss, prompt text, timeout)
   *
   * @example
   * ```ts
   * const dialogDone = page.armDialog({ accept: true }); // don't await here
   * await page.click('e5'); // triggers confirm()
   * await dialogDone;       // wait for dialog to be handled
   * ```
   */
  async armDialog(opts) {
    return armDialogViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      accept: opts.accept,
      promptText: opts.promptText,
      timeoutMs: opts.timeoutMs
    });
  }
  /**
   * Arm a one-shot file chooser handler.
   *
   * Returns a promise — store it (don't await), trigger the file picker, then await it.
   *
   * @param paths - File paths to set when the chooser appears (empty to clear)
   * @param opts - Timeout options
   *
   * @example
   * ```ts
   * const uploadDone = page.armFileUpload(['/path/to/file.pdf']); // don't await here
   * await page.click('e3'); // triggers file picker
   * await uploadDone;       // wait for files to be set
   * ```
   */
  async armFileUpload(paths, opts) {
    return armFileUploadViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      paths,
      timeoutMs: opts?.timeoutMs
    });
  }
  // ── Keyboard ─────────────────────────────────────────────────
  /**
   * Press a keyboard key or key combination.
   *
   * Uses Playwright's key names. Supports combinations with `+`.
   *
   * @param key - Key to press (e.g. `'Enter'`, `'Tab'`, `'Control+a'`, `'Meta+c'`)
   * @param opts - Options (delayMs: hold time between keydown and keyup)
   *
   * @example
   * ```ts
   * await page.press('Enter');
   * await page.press('Control+a');
   * await page.press('Meta+Shift+p');
   * ```
   */
  async press(key, opts) {
    return pressKeyViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      key,
      delayMs: opts?.delayMs
    });
  }
  // ── Navigation ───────────────────────────────────────────────
  /**
   * Get the current URL of the page.
   */
  async url() {
    const page = await getPageForTargetId({ cdpUrl: this.cdpUrl, targetId: this.targetId });
    return page.url();
  }
  /**
   * Get the page title.
   */
  async title() {
    const page = await getPageForTargetId({ cdpUrl: this.cdpUrl, targetId: this.targetId });
    return page.title();
  }
  /**
   * Navigate to a URL.
   *
   * @param url - The URL to navigate to
   * @param opts - Timeout options
   * @returns The final URL after navigation (may differ due to redirects)
   */
  async goto(url, opts) {
    return navigateViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      url,
      timeoutMs: opts?.timeoutMs,
      ssrfPolicy: this.ssrfPolicy
    });
  }
  /**
   * Reload the current page.
   *
   * @param opts - Timeout options
   */
  async reload(opts) {
    const page = await getPageForTargetId({ cdpUrl: this.cdpUrl, targetId: this.targetId });
    ensurePageState(page);
    await page.reload({ timeout: normalizeTimeoutMs(opts?.timeoutMs, 2e4) });
  }
  /**
   * Navigate back in browser history.
   *
   * @param opts - Timeout options
   */
  async goBack(opts) {
    const page = await getPageForTargetId({ cdpUrl: this.cdpUrl, targetId: this.targetId });
    ensurePageState(page);
    await page.goBack({ timeout: normalizeTimeoutMs(opts?.timeoutMs, 2e4) });
  }
  /**
   * Navigate forward in browser history.
   *
   * @param opts - Timeout options
   */
  async goForward(opts) {
    const page = await getPageForTargetId({ cdpUrl: this.cdpUrl, targetId: this.targetId });
    ensurePageState(page);
    await page.goForward({ timeout: normalizeTimeoutMs(opts?.timeoutMs, 2e4) });
  }
  // ── Wait ─────────────────────────────────────────────────────
  /**
   * Wait for various conditions on the page.
   *
   * Multiple conditions can be specified — they are checked in order.
   *
   * @param opts - Wait conditions (text, URL, load state, selector, etc.)
   *
   * @example
   * ```ts
   * await page.waitFor({ loadState: 'networkidle' });
   * await page.waitFor({ text: 'Welcome back' });
   * await page.waitFor({ url: '**\/dashboard' });
   * await page.waitFor({ timeMs: 1000 }); // sleep
   * ```
   */
  async waitFor(opts) {
    return waitForViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      ...opts
    });
  }
  // ── Evaluate ─────────────────────────────────────────────────
  /**
   * Run JavaScript in the browser page context.
   *
   * The function string is evaluated in the browser's sandbox, not in Node.js.
   * Pass a `ref` to receive the element as the first argument.
   *
   * @param fn - JavaScript function body as a string
   * @param opts - Options (ref: scope evaluation to a specific element)
   * @returns The return value of the evaluated function
   *
   * @example
   * ```ts
   * const title = await page.evaluate('() => document.title');
   * const text = await page.evaluate('(el) => el.textContent', { ref: 'e1' });
   * const count = await page.evaluate('() => document.querySelectorAll("img").length');
   * ```
   */
  async evaluate(fn, opts) {
    return evaluateViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      fn,
      ref: opts?.ref,
      timeoutMs: opts?.timeoutMs,
      signal: opts?.signal
    });
  }
  /**
   * Run JavaScript in ALL frames on the page (including cross-origin iframes).
   *
   * Playwright can access cross-origin frames via CDP, bypassing the same-origin policy.
   * This is essential for filling payment iframes (Stripe, etc.).
   *
   * @param fn - JavaScript function body as a string
   * @returns Array of results from each frame where evaluation succeeded
   *
   * @example
   * ```ts
   * const results = await page.evaluateInAllFrames(`() => {
   *   const el = document.querySelector('input[name="cardnumber"]');
   *   return el ? 'found' : null;
   * }`);
   * ```
   */
  async evaluateInAllFrames(fn) {
    return evaluateInAllFramesViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      fn
    });
  }
  // ── Capture ──────────────────────────────────────────────────
  /**
   * Take a screenshot of the page or a specific element.
   *
   * @param opts - Screenshot options (fullPage, ref, element, type)
   * @returns PNG or JPEG image as a Buffer
   *
   * @example
   * ```ts
   * const screenshot = await page.screenshot();
   * const fullPage = await page.screenshot({ fullPage: true });
   * const element = await page.screenshot({ ref: 'e1' });
   * ```
   */
  async screenshot(opts) {
    const result = await takeScreenshotViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      fullPage: opts?.fullPage,
      ref: opts?.ref,
      element: opts?.element,
      type: opts?.type
    });
    return result.buffer;
  }
  /**
   * Export the page as a PDF.
   *
   * Only works in headless mode.
   *
   * @returns PDF document as a Buffer
   */
  async pdf() {
    const result = await pdfViaPlaywright({ cdpUrl: this.cdpUrl, targetId: this.targetId });
    return result.buffer;
  }
  /**
   * Take a screenshot with numbered labels overlaid on referenced elements.
   *
   * Useful for visual debugging — each ref gets a numbered badge and border.
   *
   * @param refs - Array of ref IDs to label
   * @param opts - Options (maxLabels: limit, type: image format)
   * @returns Screenshot buffer, label positions, and any skipped refs
   *
   * @example
   * ```ts
   * const { buffer, labels, skipped } = await page.screenshotWithLabels(['e1', 'e2', 'e3']);
   * fs.writeFileSync('labeled.png', buffer);
   * ```
   */
  async screenshotWithLabels(refs, opts) {
    return screenshotWithLabelsViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      refs,
      maxLabels: opts?.maxLabels,
      type: opts?.type
    });
  }
  /**
   * Start recording a Playwright trace.
   *
   * Traces capture screenshots, DOM snapshots, and network activity.
   * Stop with `traceStop()` to save the trace file.
   *
   * @param opts - Trace options (screenshots, snapshots, sources)
   */
  async traceStart(opts) {
    return traceStartViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      screenshots: opts?.screenshots,
      snapshots: opts?.snapshots,
      sources: opts?.sources
    });
  }
  /**
   * Stop recording a trace and save it to a file.
   *
   * @param path - File path to save the trace (e.g. `'trace.zip'`)
   * @param opts - Options (allowedOutputRoots: constrain output to specific directories)
   */
  async traceStop(path2, opts) {
    return traceStopViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      path: path2,
      allowedOutputRoots: opts?.allowedOutputRoots
    });
  }
  /**
   * Wait for a network response matching a URL pattern and return its body.
   *
   * @param url - URL string or pattern to match
   * @param opts - Options (timeoutMs, maxChars)
   * @returns Response body, status, headers, and truncation info
   *
   * @example
   * ```ts
   * const resp = await page.responseBody('/api/data');
   * console.log(resp.status, resp.body);
   * ```
   */
  async responseBody(url, opts) {
    return responseBodyViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      url,
      timeoutMs: opts?.timeoutMs,
      maxChars: opts?.maxChars
    });
  }
  /**
   * Get console messages captured from the page.
   *
   * Messages are buffered automatically. Use `level` to filter by minimum severity.
   *
   * @param opts - Filter options (level, clear)
   * @returns Array of captured console messages
   */
  async consoleLogs(opts) {
    return getConsoleMessagesViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      level: opts?.level,
      clear: opts?.clear
    });
  }
  /**
   * Get uncaught errors from the page.
   *
   * @param opts - Options (clear: reset the error buffer after reading)
   * @returns Array of captured page errors
   */
  async pageErrors(opts) {
    const result = await getPageErrorsViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      clear: opts?.clear
    });
    return result.errors;
  }
  /**
   * Get network requests captured from the page.
   *
   * @param opts - Options (filter: URL substring match, clear: reset the buffer)
   * @returns Array of captured network requests
   *
   * @example
   * ```ts
   * const all = await page.networkRequests();
   * const apiCalls = await page.networkRequests({ filter: '/api/' });
   * const fresh = await page.networkRequests({ clear: true }); // read and clear
   * ```
   */
  async networkRequests(opts) {
    const result = await getNetworkRequestsViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      filter: opts?.filter,
      clear: opts?.clear
    });
    return result.requests;
  }
  // ── Viewport ─────────────────────────────────────────────────
  /**
   * Resize the browser viewport.
   *
   * @param width - Viewport width in pixels
   * @param height - Viewport height in pixels
   */
  async resize(width, height) {
    return resizeViewportViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      width,
      height
    });
  }
  // ── Storage ──────────────────────────────────────────────────
  /**
   * Get all cookies for the current browser context.
   *
   * @returns Array of cookie objects
   */
  async cookies() {
    const result = await cookiesGetViaPlaywright({ cdpUrl: this.cdpUrl, targetId: this.targetId });
    return result.cookies;
  }
  /**
   * Set a cookie in the browser context.
   *
   * @param cookie - Cookie data (must include `name`, `value`, and either `url` or `domain`+`path`)
   *
   * @example
   * ```ts
   * await page.setCookie({
   *   name: 'token',
   *   value: 'abc123',
   *   url: 'https://example.com',
   * });
   * ```
   */
  async setCookie(cookie) {
    return cookiesSetViaPlaywright({ cdpUrl: this.cdpUrl, targetId: this.targetId, cookie });
  }
  /** Clear all cookies in the browser context. */
  async clearCookies() {
    return cookiesClearViaPlaywright({ cdpUrl: this.cdpUrl, targetId: this.targetId });
  }
  /**
   * Get values from localStorage or sessionStorage.
   *
   * @param kind - `'local'` for localStorage, `'session'` for sessionStorage
   * @param key - Optional specific key to retrieve (returns all if omitted)
   * @returns Key-value map of storage entries
   */
  async storageGet(kind, key) {
    const result = await storageGetViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      kind,
      key
    });
    return result.values;
  }
  /**
   * Set a value in localStorage or sessionStorage.
   *
   * @param kind - `'local'` for localStorage, `'session'` for sessionStorage
   * @param key - Storage key
   * @param value - Storage value
   */
  async storageSet(kind, key, value) {
    return storageSetViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      kind,
      key,
      value
    });
  }
  /**
   * Clear all entries in localStorage or sessionStorage.
   *
   * @param kind - `'local'` for localStorage, `'session'` for sessionStorage
   */
  async storageClear(kind) {
    return storageClearViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      kind
    });
  }
  // ── Downloads ───────────────────────────────────────────────
  /**
   * Click a ref and save the resulting file download.
   *
   * @param ref - Ref ID of the element that triggers the download
   * @param path - Local file path to save the download to
   * @param opts - Timeout options
   * @returns Download result with URL, suggested filename, and saved path
   *
   * @example
   * ```ts
   * const result = await page.download('e7', '/tmp/report.pdf');
   * console.log(result.suggestedFilename); // 'report.pdf'
   * ```
   */
  async download(ref, path2, opts) {
    return downloadViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      ref,
      path: path2,
      timeoutMs: opts?.timeoutMs,
      allowedOutputRoots: opts?.allowedOutputRoots
    });
  }
  /**
   * Wait for the next download event (without clicking).
   *
   * Returns a promise — store it (don't await), trigger the download, then await it.
   *
   * @param opts - Options (path: save location, timeoutMs)
   * @returns Download result with URL, suggested filename, and saved path
   */
  async waitForDownload(opts) {
    return waitForDownloadViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      path: opts?.path,
      timeoutMs: opts?.timeoutMs,
      allowedOutputRoots: opts?.allowedOutputRoots
    });
  }
  // ── Emulation ───────────────────────────────────────────────
  /**
   * Set the browser to offline or online mode.
   *
   * @param offline - `true` to go offline, `false` to go online
   */
  async setOffline(offline) {
    return setOfflineViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      offline
    });
  }
  /**
   * Set extra HTTP headers for all requests.
   *
   * @param headers - Headers to add to every request
   *
   * @example
   * ```ts
   * await page.setExtraHeaders({ 'X-Custom': 'value' });
   * ```
   */
  async setExtraHeaders(headers) {
    return setExtraHTTPHeadersViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      headers
    });
  }
  /**
   * Set HTTP authentication credentials.
   *
   * @param opts - Credentials (username, password) or `{ clear: true }` to remove
   */
  async setHttpCredentials(opts) {
    return setHttpCredentialsViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      username: opts.username,
      password: opts.password,
      clear: opts.clear
    });
  }
  /**
   * Emulate a geolocation.
   *
   * @param opts - Geolocation coordinates or `{ clear: true }` to clear
   *
   * @example
   * ```ts
   * await page.setGeolocation({ latitude: 48.8566, longitude: 2.3522 }); // Paris
   * await page.setGeolocation({ clear: true }); // reset
   * ```
   */
  async setGeolocation(opts) {
    return setGeolocationViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      latitude: opts.latitude,
      longitude: opts.longitude,
      accuracy: opts.accuracy,
      origin: opts.origin,
      clear: opts.clear
    });
  }
  /**
   * Emulate a preferred color scheme.
   *
   * @param opts - Color scheme options
   *
   * @example
   * ```ts
   * await page.emulateMedia({ colorScheme: 'dark' });
   * ```
   */
  async emulateMedia(opts) {
    return emulateMediaViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      colorScheme: opts.colorScheme
    });
  }
  /**
   * Override the browser locale.
   *
   * @param locale - BCP-47 locale string (e.g. `'fr-FR'`, `'ja-JP'`)
   */
  async setLocale(locale) {
    return setLocaleViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      locale
    });
  }
  /**
   * Override the browser timezone.
   *
   * @param timezoneId - IANA timezone ID (e.g. `'America/New_York'`, `'Asia/Tokyo'`)
   */
  async setTimezone(timezoneId) {
    return setTimezoneViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      timezoneId
    });
  }
  /**
   * Emulate a specific device (viewport + user agent).
   *
   * @param name - Playwright device name (e.g. `'iPhone 13'`, `'Pixel 5'`)
   *
   * @example
   * ```ts
   * await page.setDevice('iPhone 13');
   * ```
   */
  async setDevice(name) {
    return setDeviceViaPlaywright({
      cdpUrl: this.cdpUrl,
      targetId: this.targetId,
      name
    });
  }
};
var BrowserClaw = class _BrowserClaw {
  cdpUrl;
  ssrfPolicy;
  chrome;
  constructor(cdpUrl, chrome, ssrfPolicy) {
    this.cdpUrl = cdpUrl;
    this.chrome = chrome;
    this.ssrfPolicy = ssrfPolicy;
  }
  /**
   * Launch a new Chrome instance and connect to it.
   *
   * Automatically detects Chrome, Brave, Edge, or Chromium on the system.
   * Creates a dedicated browser profile to avoid conflicts with your daily browser.
   *
   * @param opts - Launch options (headless, executablePath, cdpPort, etc.)
   * @returns A connected BrowserClaw instance
   *
   * @example
   * ```ts
   * // Default: visible Chrome window
   * const browser = await BrowserClaw.launch();
   *
   * // Headless mode
   * const browser = await BrowserClaw.launch({ headless: true });
   *
   * // Specific browser
   * const browser = await BrowserClaw.launch({
   *   executablePath: '/usr/bin/google-chrome',
   * });
   * ```
   */
  static async launch(opts = {}) {
    const chrome = await launchChrome(opts);
    const cdpUrl = `http://127.0.0.1:${chrome.cdpPort}`;
    const ssrfPolicy = opts.allowInternal ? { ...opts.ssrfPolicy, dangerouslyAllowPrivateNetwork: true } : opts.ssrfPolicy;
    return new _BrowserClaw(cdpUrl, chrome, ssrfPolicy);
  }
  /**
   * Connect to an already-running Chrome instance via its CDP endpoint.
   *
   * The Chrome instance must have been started with `--remote-debugging-port`.
   *
   * @param cdpUrl - CDP endpoint URL (e.g. `'http://localhost:9222'`)
   * @returns A connected BrowserClaw instance
   *
   * @example
   * ```ts
   * // Chrome started with: chrome --remote-debugging-port=9222
   * const browser = await BrowserClaw.connect('http://localhost:9222');
   * ```
   */
  static async connect(cdpUrl, opts) {
    if (!await isChromeReachable(cdpUrl, 3e3, opts?.authToken)) {
      throw new Error(`Cannot connect to Chrome at ${cdpUrl}. Is Chrome running with --remote-debugging-port?`);
    }
    await connectBrowser(cdpUrl, opts?.authToken);
    const ssrfPolicy = opts?.allowInternal ? { ...opts.ssrfPolicy, dangerouslyAllowPrivateNetwork: true } : opts?.ssrfPolicy;
    return new _BrowserClaw(cdpUrl, null, ssrfPolicy);
  }
  /**
   * Open a URL in a new tab and return the page handle.
   *
   * @param url - URL to navigate to
   * @returns A CrawlPage for the new tab
   *
   * @example
   * ```ts
   * const page = await browser.open('https://example.com');
   * const { snapshot, refs } = await page.snapshot();
   * ```
   */
  async open(url) {
    const tab = await createPageViaPlaywright({ cdpUrl: this.cdpUrl, url, ssrfPolicy: this.ssrfPolicy });
    return new CrawlPage(this.cdpUrl, tab.targetId, this.ssrfPolicy);
  }
  /**
   * Get a CrawlPage handle for the currently active tab.
   *
   * @returns CrawlPage for the first/active page
   */
  async currentPage() {
    const { browser } = await connectBrowser(this.cdpUrl);
    const pages = await getAllPages(browser);
    if (!pages.length) throw new Error("No pages available. Use browser.open(url) to create a tab.");
    const tid = await pageTargetId(pages[0]).catch(() => null);
    if (!tid) throw new Error("Failed to get targetId for the current page.");
    return new CrawlPage(this.cdpUrl, tid, this.ssrfPolicy);
  }
  /**
   * List all open tabs.
   *
   * @returns Array of tab information objects
   */
  async tabs() {
    return listPagesViaPlaywright({ cdpUrl: this.cdpUrl });
  }
  /**
   * Bring a tab to the foreground.
   *
   * @param targetId - CDP target ID of the tab (from `tabs()` or `page.id`)
   */
  async focus(targetId) {
    return focusPageByTargetIdViaPlaywright({ cdpUrl: this.cdpUrl, targetId });
  }
  /**
   * Close a tab.
   *
   * @param targetId - CDP target ID of the tab to close
   */
  async close(targetId) {
    return closePageByTargetIdViaPlaywright({ cdpUrl: this.cdpUrl, targetId });
  }
  /**
   * Get a CrawlPage handle for a specific tab by its target ID.
   *
   * Unlike `open()`, this doesn't create a new tab — it wraps an existing one.
   *
   * @param targetId - CDP target ID of the tab
   * @returns CrawlPage for the specified tab
   */
  page(targetId) {
    return new CrawlPage(this.cdpUrl, targetId, this.ssrfPolicy);
  }
  /** The CDP endpoint URL for this browser connection. */
  get url() {
    return this.cdpUrl;
  }
  /**
   * Stop the browser and clean up all resources.
   *
   * If the browser was launched by `BrowserClaw.launch()`, the Chrome process
   * will be terminated. If connected via `BrowserClaw.connect()`, only the
   * Playwright connection is closed.
   */
  async stop() {
    await disconnectBrowser();
    if (this.chrome) {
      await stopChrome(this.chrome);
      this.chrome = null;
    }
  }
};

export { BrowserClaw, CrawlPage, InvalidBrowserNavigationUrlError, assertBrowserNavigationAllowed, isChromeCdpReady, withBrowserNavigationPolicy };
