// mr-memory/index.ts
/**
 * MemoryRouter Plugin for OpenClaw
 *
 * Persistent AI memory via MemoryRouter (memoryrouter.ai).
 * Uses before_agent_start + agent_end hooks to inject/store memories
 * via the MemoryRouter relay API. No proxy interception needed.
 *
 * BYOK — provider API keys never touch MemoryRouter.
 */

import { readFile, readdir, stat } from "node:fs/promises";
import { join } from "node:path";
import { spawn } from "node:child_process";
import type { OpenClawPluginApi } from "openclaw/plugin-sdk";

const DEFAULT_ENDPOINT = "https://api.memoryrouter.ai";

/** Strip media-attached references and OpenClaw media instruction text so
 *  old screenshots/photos/audio don't get stored and re-injected forever. */
const MEDIA_ATTACHED_RE = /\[media attached[^\]]*\]/gi;
const MEDIA_TAGS_RE = /<media:(?:audio|image|video|document)>/gi;
const MEDIA_INSTRUCTION_RE = /To send an image back,[\s\S]*?Keep caption in the text body\./gi;
const MEDIA_INBOUND_PATH_RE = /\/Users\/[^\s]*\/media\/inbound\/[^\s)\]"]*/gi;
const IMAGE_DATA_RE = /\[image data removed[^\]]*\]/gi;
function stripMediaReferences(text: string): string {
  return text
    .replace(MEDIA_INSTRUCTION_RE, "")
    .replace(MEDIA_ATTACHED_RE, "[media reference removed]")
    .replace(MEDIA_TAGS_RE, "")
    .replace(MEDIA_INBOUND_PATH_RE, "[media reference removed]")
    .replace(IMAGE_DATA_RE, "")
    .replace(/\n{3,}/g, "\n\n");
}

/** Wrap raw memory context in XML tags with a strong instruction */
/** Wrap API response in extraction markers so we can strip it next turn. */
function wrapForInjection(context: string): string {
  return `<mr-memory>\n${stripMediaReferences(context)}\n</mr-memory>`;
}

/** Strip previous memory injections from message text to prevent stacking.
 *  prependContext persists in conversation history — without stripping,
 *  each turn accumulates another full injection (~20K tokens). */
const MEMORY_TAG_RE = /<mr-memory>[\s\S]*?<\/mr-memory>\s*/g;
/** Legacy tag pattern for backward compat (pre-2.7.0 injections still in history) */
const LEGACY_TAG_RE = /<memory_context>[\s\S]*?<\/memory_context>\s*(?:The above are retrieved memories|IMPORTANT: The above block contains retrieved memories)[^\n]*\n*/g;
function stripOldMemory(text: string): string {
  return text.replace(MEMORY_TAG_RE, "").replace(LEGACY_TAG_RE, "").trim();
}

// ── Ingest sanitization: strip system noise so only real conversations get stored
// Patterns that indicate system-generated messages (not user conversations)
const SYSTEM_NOISE_PATTERNS = [
  // Heartbeat prompts and responses
  /^Read HEARTBEAT\.md if it exists/,
  /^Read HEARTBEAT\.md/,
  /HEARTBEAT_OK\s*$/,
  // Pre-compaction flush prompts
  /^Pre-compaction memory flush/,
  // Post-compaction audit warnings
  /⚠️ Post-Compaction Audit:/,
  // Silent reply ack
  /^NO_REPLY\s*$/,
  // Agent abort notices
  /^Note: The previous agent run was aborted/,
  // Queued message headers
  /^\[Queued messages while agent was busy\]/,
  // Media reference placeholders
  /^\[media reference removed\]\s*$/,
  // System event timestamps (cron fires, exec failures, etc.)
  /^System: \[\d{4}-\d{2}-\d{2}/,
  /^\[\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} [A-Z]{2,4}\] ⚠️/,
  // Memory flush with date pattern
  /^Store durable memories now/,
  // Current time system lines (injected by OpenClaw into system events)
  /^Current time: .+\(America\//,
  // New session notice
  /^A new session was started via \/new or \/reset/,
  // Post-compaction recovery instructions
  /^Please read them now using the Read tool/,
  /^GatewayRestart:\s*\{/,
  /^This ensures your operating protocols are restored/,
  // Cron job prompts — automated system instructions, not human conversation
  /^\[cron:[^\]]+\]/,
  // Sub-agent completion announcements (system-generated)
  /^A background task "[^"]*" just completed/,
  // Sub-agent stats lines
  /^Stats: runtime \d+/,
  // Summarize instruction for sub-agent completions
  /^Summarize this naturally for the user/,
  // Session reset greet prompts
  /^Greet the user in your configured persona/,
];

/** Strip OpenClaw envelope metadata from user messages */
const ENVELOPE_METADATA_RE = /Conversation info \(untrusted metadata\):\s*```json\s*\{[^}]*\}\s*```\s*/g;
const SENDER_METADATA_RE = /Sender \(untrusted metadata\):\s*```json\s*\{[^}]*\}\s*```\s*/g;
const REPLY_CONTEXT_RE = /Replied message \(untrusted, for context\):\s*```json\s*\{[^}]*\}\s*```\s*/g;
const MEMORY_INSTRUCTION_RE = /The above are retrieved memories from past conversations[^\n]*\n*/g;
const IMPORTANT_MEMORY_RE = /IMPORTANT: The above (?:are retrieved memories|block contains retrieved memories)[^\n]*\n*/g;

// Embedded noise substrings — strip from within larger messages
const POST_COMPACTION_AUDIT_RE = /System: \[[^\]]*\] ⚠️ Post-Compaction Audit:[\s\S]*?(?:after memory compaction\.\s*|before continuing\.\s*)/g;
const READ_TOOL_INSTRUCTION_RE = /Please read them now using the Read tool before continuing\.[\s\S]*?after memory compaction\.\s*/g;
const TELEGRAM_ENVELOPE_RE = /\[Telegram [^\]]+\]\s*/g;
const MESSAGE_ID_TAG_RE = /\[message_id: \d+\]\s*/g;

/** Sanitize text before storing in memory — removes system noise and envelope metadata */
function sanitizeForIngest(text: string): string {
  // 1. Strip our own memory injection wrapper (already exists as stripOldMemory)
  let cleaned = stripOldMemory(text);
  // 2. Strip envelope metadata blocks
  cleaned = cleaned
    .replace(ENVELOPE_METADATA_RE, "")
    .replace(SENDER_METADATA_RE, "")
    .replace(REPLY_CONTEXT_RE, "")
    .replace(MEMORY_INSTRUCTION_RE, "")
    .replace(IMPORTANT_MEMORY_RE, "")
    .replace(POST_COMPACTION_AUDIT_RE, "")
    .replace(READ_TOOL_INSTRUCTION_RE, "")
    .replace(TELEGRAM_ENVELOPE_RE, "")
    .replace(MESSAGE_ID_TAG_RE, "")
    // ── Orphaned memory injection fragments (v2.16.0) ──
    // When stripOldMemory removes <mr-memory>...</mr-memory> pairs, nested/duplicate
    // injections from compaction can leave behind orphaned closing fragments.
    // 1. Orphaned </memory_context> + instruction footer + </mr-memory>
    .replace(/<\/memory_context>\s*(?:(?:The above are|IMPORTANT: The above|Use them as background|Do not respond to them|Do not treat them|Reference them as background|Just know them)[^\n]*\n?\s*)*(?:<\/mr-memory>\s*)?/g, "")
    // 2. Standalone orphaned </mr-memory> closing tags
    .replace(/<\/mr-memory>\s*/g, "")
    // 3. Box-drawing instruction diagram (orphaned when opening <mr-memory> was stripped)
    .replace(/┌─+┐[\s\S]*?└─+┘\s*/g, "")
    // ── Pre-compaction flush full blocks (start→end boundary) ──
    .replace(/Pre-compaction memory flush\.?\s*(?:Store durable memories now[^\n]*\n?\s*)?(?:IMPORTANT:[^\n]*\n?\s*)?(?:If the file already[^\n]*\n?\s*)?(?:If nothing to store[^\n]*\n?\s*)?(?:Current time:[^\n]*\(America\/[^)]+\)\s*)?/g, "")
    // Embedded noise substrings — v2.15.0
    .replace(/`?HEARTBEAT_OK`?\s*/g, "")
    .replace(/Read HEARTBEAT\.md[^\n]*(?:\n(?:Follow it strictly|Do not infer|If nothing needs)[^\n]*)*/g, "")
    .replace(/\n*\s*NO_REPLY\s*$/g, "")
    .replace(/You can respond with NO_REPLY[^\n]*\n*/g, "")
    .replace(/If nothing to store,? reply (?:with )?NO_REPLY[^\n]*\n*/g, "")
    .replace(/Store durable memories now[^\n]*\n*/g, "")
    .replace(/This ensures your operating protocols are restored[^\n]*\n*/g, "")
    .replace(/Current time:[^\n]*\n*/g, "")
    .replace(/\bset:\s*\n\s*-\s*WORKFLOW_AUTO\.md[^\n]*\n(?:\s*-\s*memory[^\n]*\n)*/g, "")
    .replace(/Do not mention technical details[^\n]*\n*/g, "");
  // 3. Collapse excessive whitespace from removals
  cleaned = cleaned.replace(/\n{3,}/g, "\n\n").trim();
  return cleaned;
}

/** Check if a message is system noise that should not be stored */
function isSystemNoise(text: string): boolean {
  const trimmed = text.trim();
  if (!trimmed) return true;
  return SYSTEM_NOISE_PATTERNS.some(pattern => pattern.test(trimmed));
}

/**
 * Detect OpenClaw compaction summary blobs — large user messages containing
 * interleaved [USER] and [ASSISTANT] markers from conversation history dumps.
 * These are 100% redundant since individual messages are already uploaded.
 */
function isCompactionSummary(text: string, role: string): boolean {
  if (role !== "user") return false;
  if (text.length < 2000) return false;
  const markers = (text.match(/\[(?:USER|ASSISTANT)[\] ]/g) || []).length;
  return markers >= 3;
}

// Workspace files OpenClaw loads into the system prompt
const WORKSPACE_FILES = [
  "IDENTITY.md", "USER.md", "MEMORY.md", "HEARTBEAT.md",
  "TOOLS.md", "AGENTS.md", "SOUL.md", "BOOTSTRAP.md",
];

type MemoryRouterConfig = {
  key: string;
  endpoint?: string;
  density?: "low" | "high" | "xhigh";
  mode?: "relay" | "proxy";
  logging?: boolean;
  embeddings?: string;
};

// ──────────────────────────────────────────────────────
// Helpers
// ──────────────────────────────────────────────────────

function resolveOpenClawInvocation(): { command: string; args: string[] } {
  const entry = process.argv[1];
  if (entry) return { command: process.execPath, args: [entry] };
  return { command: "openclaw", args: [] };
}

async function runOpenClawConfigSet(path: string, value: string, json = false): Promise<void> {
  const base = resolveOpenClawInvocation();
  const args = [...base.args, "config", "set", path, value];
  if (json) args.push("--json");

  await new Promise<void>((resolve, reject) => {
    const child = spawn(base.command, args, {
      stdio: ["ignore", "ignore", "pipe"],
      env: process.env,
    });
    let stderr = "";
    child.stderr.on("data", (chunk) => { stderr += String(chunk); });
    child.on("error", reject);
    child.on("close", (code) => {
      if (code === 0) resolve();
      else reject(new Error(`openclaw config set failed (exit ${code}): ${stderr.trim()}`));
    });
  });
}

type CompatApi = OpenClawPluginApi & {
  updatePluginConfig?: (config: Record<string, unknown>) => Promise<void>;
  updatePluginEnabled?: (enabled: boolean) => Promise<void>;
};

async function setPluginConfig(api: OpenClawPluginApi, config: Record<string, unknown>): Promise<void> {
  const compat = api as CompatApi;
  if (typeof compat.updatePluginConfig === "function") {
    await compat.updatePluginConfig(config);
    return;
  }
  await runOpenClawConfigSet(`plugins.entries.${api.id}.config`, JSON.stringify(config), true);
}

async function setPluginEnabled(api: OpenClawPluginApi, enabled: boolean): Promise<void> {
  const compat = api as CompatApi;
  if (typeof compat.updatePluginEnabled === "function") {
    await compat.updatePluginEnabled(enabled);
    return;
  }
  await runOpenClawConfigSet(`plugins.entries.${api.id}.enabled`, enabled ? "true" : "false", true);
}

/**
 * Read all workspace files and return as a single text blob for token counting.
 */
async function readWorkspaceFiles(workspaceDir: string): Promise<string> {
  const parts: string[] = [];
  for (const file of WORKSPACE_FILES) {
    try {
      const content = await readFile(join(workspaceDir, file), "utf8");
      if (content.trim()) parts.push(`## ${file}\n${content}`);
    } catch { /* file doesn't exist — skip */ }
  }
  return parts.join("\n\n");
}

/**
 * Build a text representation of tools config for token counting.
 */
function serializeToolsConfig(config: Record<string, unknown>): string {
  const tools = config.tools;
  if (!tools) return "";
  try {
    return `## Tools Config\n${JSON.stringify(tools, null, 2)}`;
  } catch { return ""; }
}

/**
 * Build a text representation of skills for token counting.
 */
function serializeSkillsConfig(config: Record<string, unknown>): string {
  const skills = (config as any).skills?.entries;
  if (!skills || typeof skills !== "object") return "";
  try {
    const names = Object.keys(skills);
    return `## Skills (${names.length})\n${names.join(", ")}`;
  } catch { return ""; }
}

// ──────────────────────────────────────────────────────
// Plugin Definition
// ──────────────────────────────────────────────────────

const memoryRouterPlugin = {
  id: "mr-memory",
  name: "MemoryRouter",
  description: "Persistent AI memory powered by MemoryRouter",

  register(api: OpenClawPluginApi) {
    const cfg = api.pluginConfig as MemoryRouterConfig | undefined;
    const endpoint = cfg?.endpoint?.replace(/\/v1\/?$/, "") || DEFAULT_ENDPOINT;
    const memoryKey = cfg?.key;
    const density = cfg?.density || "high";
    const mode = cfg?.mode || "relay";
    const logging = cfg?.logging ?? false;
    const embeddings = cfg?.embeddings; // undefined = bge (default), "qwen" = qwen3-8b
    const log = (msg: string) => { if (logging) api.logger.info?.(msg); };

    if (memoryKey) {
      const modelLabel = embeddings ? `, embeddings: ${embeddings}` : "";
      api.logger.info?.(`memoryrouter: active (key: ${memoryKey.slice(0, 6)}..., mode: ${mode}${modelLabel})`);
    } else {
      api.logger.info?.("memoryrouter: no key configured — run: openclaw mr <key>");
      api.logger.info?.("memoryrouter: get your free API key at https://memoryrouter.ai");
    }

    // ==================================================================
    // Core: before_agent_start — search memories, inject context
    // ==================================================================

    if (memoryKey) {
      // Track whether we've already fired for this prompt (dedup double-fire)
      let lastPreparedPrompt = "";
      // Track whether before_prompt_build already handled the first call in this run
      let promptBuildFiredThisRun = false;

      // ── llm_input: fires on EVERY LLM call (tool iterations, cron, sub-agents)
      // On stock OpenClaw, the return value is ignored (fire-and-forget).
      // When PR #24122 merges, OpenClaw will use the returned prependContext.
      // This gives forward compatibility — no plugin update needed.
      api.on("llm_input", async (event, ctx) => {
        log(`memoryrouter: llm_input fired (sessionKey=${ctx.sessionKey}, promptBuildFired=${promptBuildFiredThisRun})`);
        // Skip the first call — before_prompt_build already handled it
        // (before_prompt_build includes workspace+tools+skills for accurate billing)
        if (promptBuildFiredThisRun) {
          promptBuildFiredThisRun = false; // reset so subsequent calls go through
          return;
        }

        try {
          const prompt = event.prompt;
          if (prompt === lastPreparedPrompt && lastPreparedPrompt !== "") return;
          lastPreparedPrompt = prompt;

          // Build lightweight context (no workspace/tools — just history + prompt)
          const contextPayload: Array<{ role: string; content: string }> = [];
          if (event.historyMessages && Array.isArray(event.historyMessages)) {
            for (const msg of event.historyMessages) {
              const m = msg as { role?: string; content?: unknown };
              if (!m.role) continue;
              // Skip tool result/use messages — contains file reads, exec output, etc.
              if (m.role !== "user" && m.role !== "assistant" && m.role !== "system") continue;
              let text = "";
              if (typeof m.content === "string") text = m.content;
              else if (Array.isArray(m.content)) {
                text = (m.content as Array<{ type?: string; text?: string }>)
                  .filter(b => b.type === "text" && b.text)
                  .map(b => b.text!)
                  .join("\n");
              }
              if (!text) continue;
              // Sanitize ALL roles: strip memory injections, envelope metadata, system noise
              const cleaned = sanitizeForIngest(text);
              if (!cleaned || isSystemNoise(cleaned)) continue;
              if (isCompactionSummary(cleaned, m.role || "")) continue;
              contextPayload.push({ role: m.role, content: cleaned });
            }
          }
          const cleanedPrompt = sanitizeForIngest(prompt);
          if (cleanedPrompt && !isSystemNoise(cleanedPrompt)) {
            contextPayload.push({ role: "user", content: cleanedPrompt });
          }


          const res = await fetch(`${endpoint}/v1/memory/prepare`, {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
              Authorization: `Bearer ${memoryKey}`,
            },
            body: JSON.stringify({
              messages: contextPayload,
              session_id: ctx.sessionKey,
              density,
              ...(embeddings && { embeddings }),
            }),
          });

          if (!res.ok) {
            const errBody = await res.json().catch(() => null) as any;
            if (errBody?.error?.code === "payment_required") {
              api.logger.warn?.(`⚠️ MemoryRouter: Out of credits. Memory is NOT being injected. Add credits at https://app.memoryrouter.ai/settings/billing`);
            } else {
              log(`memoryrouter: prepare failed (${res.status})`);
            }
            return;
          }

          const data = (await res.json()) as {
            context?: string;
            memories_found?: number;
            tokens_billed?: number;
            memory_tokens?: number;
          };

          if (data.context) {
            log(
              `memoryrouter: injected ${data.memories_found || 0} memories (${data.memory_tokens || 0} tokens)`,
            );
            const wrapped = wrapForInjection(data.context);
            return { appendSystemPrompt: wrapped };
          }
        } catch {
          // Silent fail on tool iterations — don't block the agent
        }
      });

      // ── before_prompt_build: fires once per run (primary, includes full billing context)
      api.on("before_prompt_build", async (event, ctx) => {
        promptBuildFiredThisRun = true;
        log(`memoryrouter: before_prompt_build fired (sessionKey=${ctx.sessionKey}, promptLen=${event.prompt?.length})`);
        try {
          const prompt = event.prompt;

          // Deduplicate — if we already prepared this exact prompt, skip
          if (prompt === lastPreparedPrompt && lastPreparedPrompt !== "") {
            return;
          }
          lastPreparedPrompt = prompt;

          // 1. Read workspace files for full token count
          const workspaceDir = ctx.workspaceDir || "";
          let workspaceText = "";
          if (workspaceDir) {
            workspaceText = await readWorkspaceFiles(workspaceDir);
          }

          // 2. Serialize tools + skills from config
          const toolsText = serializeToolsConfig(api.config as unknown as Record<string, unknown>);
          const skillsText = serializeSkillsConfig(api.config as unknown as Record<string, unknown>);

          // 3. Build full context payload (messages + workspace + tools + skills)
          const contextPayload: Array<{ role: string; content: string }> = [];

          // Add workspace context as a system-level entry
          const fullContext = [workspaceText, toolsText, skillsText].filter(Boolean).join("\n\n");
          if (fullContext) {
            contextPayload.push({ role: "system", content: fullContext });
          }

          // Add conversation history (sanitized — strip noise, metadata, old memory injections)
          if (event.messages && Array.isArray(event.messages)) {
            for (const msg of event.messages) {
              const m = msg as { role?: string; content?: unknown };
              if (!m.role) continue;
              // Skip tool result/use messages — contains file reads, exec output, etc.
              if (m.role !== "user" && m.role !== "assistant" && m.role !== "system") continue;
              
              let text = "";
              if (typeof m.content === "string") {
                text = m.content;
              } else if (Array.isArray(m.content)) {
                // Handle Anthropic-style content blocks [{type:"text", text:"..."}, ...]
                text = (m.content as Array<{ type?: string; text?: string }>)
                  .filter(b => b.type === "text" && b.text)
                  .map(b => b.text!)
                  .join("\n");
              }
              
              if (!text) continue;
              // Sanitize ALL roles: strip memory injections, envelope metadata, system noise
              const cleaned = sanitizeForIngest(text);
              if (!cleaned || isSystemNoise(cleaned)) continue;
              if (isCompactionSummary(cleaned, m.role || "")) continue;
              contextPayload.push({ role: m.role, content: cleaned });
            }
          }

          // Add current user prompt (sanitized)
          const cleanedPrompt = sanitizeForIngest(prompt);
          if (cleanedPrompt && !isSystemNoise(cleanedPrompt)) {
            contextPayload.push({ role: "user", content: cleanedPrompt });
          }

          // 4. Call /v1/memory/prepare

          const res = await fetch(`${endpoint}/v1/memory/prepare`, {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
              Authorization: `Bearer ${memoryKey}`,
            },
            body: JSON.stringify({
              messages: contextPayload,
              session_id: ctx.sessionKey,
              density,
              ...(embeddings && { embeddings }),
            }),
          });

          if (!res.ok) {
            const errBody = await res.json().catch(() => null) as any;
            if (errBody?.error?.code === "payment_required") {
              api.logger.warn?.(`⚠️ MemoryRouter: Out of credits. Memory is NOT being injected. Add credits at https://app.memoryrouter.ai/settings/billing`);
            } else {
              log(`memoryrouter: prepare failed (${res.status})`);
            }
            return;
          }

          const data = (await res.json()) as {
            context?: string;
            memories_found?: number;
            tokens_billed?: number;
            memory_tokens?: number;
          };

          log(`memoryrouter: prepare response — memories_found=${data.memories_found || 0}, context_length=${data.context?.length || 0}, tokens=${data.memory_tokens || 0}, embeddings=${embeddings || 'bge'}`);

          if (data.context) {
            log(
              `memoryrouter: injected ${data.memories_found || 0} memories (${data.memory_tokens || 0} tokens)`,
            );
            const wrapped = wrapForInjection(data.context);
            return { prependContext: wrapped };
          } else {
            log(`memoryrouter: prepare returned no context (memories_found=${data.memories_found || 0})`);
          }
        } catch (err) {
          log(
            `memoryrouter: prepare error — ${err instanceof Error ? err.message : String(err)}`,
          );
        }
      });

      // ==================================================================
      // Core: agent_end — store this turn's conversation
      // ==================================================================

      api.on("agent_end", async (event, ctx) => {
        try {
          const msgs = event.messages;
          if (!msgs || !Array.isArray(msgs) || msgs.length === 0) return;

          // Extract text from a message (handles string + content block arrays)
          function extractText(content: unknown): string {
            if (typeof content === "string") return content;
            if (Array.isArray(content)) {
              return (content as Array<{ type?: string; text?: string }>)
                .filter(b => b.type === "text" && b.text)
                .map(b => b.text!)
                .join("\n");
            }
            return "";
          }

          // Find the last user message, then collect ALL assistant messages after it
          // This captures the full response even if she sent multiple messages
          let lastUserIdx = -1;
          for (let i = msgs.length - 1; i >= 0; i--) {
            const msg = msgs[i] as { role?: string; content?: unknown };
            const text = extractText(msg.content);
            if (msg.role === "user" && text) {
              lastUserIdx = i;
              break;
            }
          }

          const toStore: Array<{ role: string; content: string }> = [];

          // Add the user message (sanitized)
          if (lastUserIdx >= 0) {
            const userMsg = msgs[lastUserIdx] as { content?: unknown };
            let userText = extractText(userMsg.content);
            if (userText) {
              // Sanitize: strip memory injections, envelope metadata, and system noise
              userText = sanitizeForIngest(stripMediaReferences(userText));
              if (userText && !isSystemNoise(userText)) {
                toStore.push({ role: "user", content: userText });
              }
            }
          }

          // Collect ALL assistant messages after the last user message
          const assistantParts: string[] = [];
          for (let i = (lastUserIdx >= 0 ? lastUserIdx + 1 : 0); i < msgs.length; i++) {
            const msg = msgs[i] as { role?: string; content?: unknown };
            if (msg.role === "assistant") {
              const text = extractText(msg.content);
              if (text) assistantParts.push(text);
            }
          }
          if (assistantParts.length > 0) {
            // Strip media refs so image paths don't get stored and re-injected
            let assistantText = stripMediaReferences(assistantParts.join("\n\n"));
            // Skip storing system noise responses (HEARTBEAT_OK, NO_REPLY, etc.)
            if (!isSystemNoise(assistantText)) {
              toStore.push({ role: "assistant", content: assistantText });
            }
          }

          // Skip if nothing meaningful to store (both messages were system noise)
          if (toStore.length === 0) {
            log("memoryrouter: skipped ingest — system noise filtered out");
            return;
          }

          // Await the fetch so OpenClaw's runVoidHook keeps the event loop alive.
          try {
            const res = await fetch(`${endpoint}/v1/memory/ingest`, {
              method: "POST",
              headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${memoryKey}`,
              },
              body: JSON.stringify({
                messages: toStore,
                session_id: ctx.sessionKey,
                model: "unknown",
                ...(embeddings && { embeddings }),
              }),
            });
            if (!res.ok) {
              const details = await res.text().catch(() => "");
              if (details.includes("payment_required")) {
                api.logger.warn?.(`⚠️ MemoryRouter: Out of credits. Conversations are NOT being stored. Add credits at https://app.memoryrouter.ai/settings/billing`);
              } else {
                const suffix = details ? ` — ${details.slice(0, 200)}` : "";
                log(`memoryrouter: ingest failed (${res.status})${suffix}`);
              }
            } else {
              api.logger.debug?.(`memoryrouter: ingest accepted (${toStore.length} messages)`);
            }
          } catch (err) {
            log(
              `memoryrouter: ingest error — ${err instanceof Error ? err.message : String(err)}`,
            );
          }
        } catch (err) {
          log(
            `memoryrouter: agent_end error — ${err instanceof Error ? err.message : String(err)}`,
          );
        }
      });
    } // end if (memoryKey)

    // ==================================================================
    // CLI Commands
    // ==================================================================

    // ── Helper: Optimize OpenClaw's built-in memorySearch config
    // When MR is active, disable session transcript scanning (MR handles it)
    // but keep workspace file search (local, fast, no conflict)
    async function optimizeMemorySearchConfig(): Promise<void> {
      try {
        // Set sources to only "memory" (workspace files), remove "sessions"
        await runOpenClawConfigSet("agents.defaults.memorySearch.sources", JSON.stringify(["memory"]), true);
        // Disable experimental session memory
        await runOpenClawConfigSet("agents.defaults.memorySearch.experimental.sessionMemory", "false", true);
      } catch {
        // Non-fatal — config optimization is best-effort
      }
    }

    // ── Helper: Restore OpenClaw's built-in memorySearch config when MR is disabled
    async function restoreMemorySearchConfig(): Promise<void> {
      try {
        await runOpenClawConfigSet("agents.defaults.memorySearch.sources", JSON.stringify(["memory", "sessions"]), true);
        await runOpenClawConfigSet("agents.defaults.memorySearch.experimental.sessionMemory", "true", true);
      } catch {
        // Non-fatal
      }
    }

    api.registerCli(
      ({ program }) => {
        const applyKey = async (key: string) => {
          if (!key.startsWith("mk")) {
            console.error("Invalid key format. Keys start with 'mk' (e.g. mk_xxx)");
            return;
          }
          try {
            await setPluginConfig(api, { key });
            await setPluginEnabled(api, true);
            // Optimize OpenClaw's built-in memorySearch: keep workspace file search
            // but disable session transcript scanning (MR handles conversational memory)
            await optimizeMemorySearchConfig();
            console.log(`✓ MemoryRouter enabled. Key: ${key.slice(0, 6)}...${key.slice(-3)}`);
            console.log(`  • Session memory scanning disabled (MR handles this)`);
            console.log(`  • Workspace file search kept active`);
            console.log(`\nRun: openclaw mr upload    to upload your memories`);
          } catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            console.error(`Failed to enable MemoryRouter: ${message}`);
          }
        };

        const mr = program
          .command("mr")
          .description("MemoryRouter memory commands")
          .argument("[key]", "Your MemoryRouter memory key (mk_xxx)")
          .action(async (key: string | undefined) => {
            if (!key) { mr.help(); return; }
            await applyKey(key);
          });

        mr.command("enable")
          .description("Enable MemoryRouter with a memory key (alias)")
          .argument("<key>", "Your MemoryRouter memory key (mk_xxx)")
          .action(async (key: string) => { await applyKey(key); });

        mr.command("off")
          .description("Disable MemoryRouter (removes key, keeps CLI available)")
          .action(async () => {
            try {
              // Soft disable: clear key but keep plugin enabled so CLI remains available
              // This avoids the catch-22 where disabling the plugin removes the `mr` command
              await setPluginConfig(api, {});
              // Note: we intentionally do NOT call setPluginEnabled(api, false)
              // Restore OpenClaw's built-in session memory scanning
              await restoreMemorySearchConfig();
              console.log("✓ MemoryRouter disabled (key cleared).");
              console.log("  • Session memory scanning restored");
              console.log("  • CLI still available — run `openclaw mr <key>` to re-enable");
            } catch (err) {
              console.error(`Failed to disable: ${err instanceof Error ? err.message : String(err)}`);
            }
          });

        // Density commands
        for (const [name, desc] of [
          ["xhigh", "Set density to xhigh (160 results, ~50k tokens)"],
          ["high", "Set density to high (80 results, ~24k tokens) [default]"],
          ["low", "Set density to low (40 results, ~12k tokens)"],
        ] as const) {
          mr.command(name)
            .description(desc)
            .action(async () => {
              if (!memoryKey) { console.error("Not configured. Run: openclaw mr <key>"); return; }
              try {
                await setPluginConfig(api, { key: memoryKey, endpoint: cfg?.endpoint, density: name });
                console.log(`✓ Memory density set to ${name}`);
              } catch (err) {
                console.error(`Failed: ${err instanceof Error ? err.message : String(err)}`);
              }
            });
        }

        // Logging toggle
        mr.command("logging")
          .description("Toggle debug logging on/off")
          .action(async () => {
            if (!memoryKey) {
              console.error("Not configured. Run: openclaw mr <key>");
              return;
            }
            const newLogging = !logging;
            try {
              await setPluginConfig(api, {
                key: memoryKey,
                endpoint: cfg?.endpoint,
                density,
                mode,
                logging: newLogging,
              });
              console.log(`✓ Logging ${newLogging ? "ON" : "OFF"} (restart gateway to apply)`);
            } catch (err) {
              console.error(`Failed: ${err instanceof Error ? err.message : String(err)}`);
            }
          });

        // Mode commands
        for (const [modeName, modeDesc] of [
          ["relay", "Relay mode — hooks only, works on stock OpenClaw [default]"],
          ["proxy", "Proxy mode — memory on every LLM call (requires registerStreamFnWrapper)"],
        ] as const) {
          mr.command(modeName)
            .description(modeDesc)
            .action(async () => {
              if (!memoryKey) { console.error("Not configured. Run: openclaw mr <key>"); return; }
              try {
                await setPluginConfig(api, { key: memoryKey, endpoint: cfg?.endpoint, density, mode: modeName });
                console.log(`✓ Mode set to ${modeName}`);
              } catch (err) {
                console.error(`Failed: ${err instanceof Error ? err.message : String(err)}`);
              }
            });
        }

        // Embedding model command
        mr.command("embeddings")
          .description("Get or set embedding model (bge, qwen)")
          .argument("[model]", "Embedding model: bge (default, 1024d) or qwen (4096d)")
          .action(async (model: string | undefined) => {
            if (!model) {
              console.log(`Current embedding model: ${embeddings || "bge (default)"}`);
              console.log(`\nAvailable models:`);
              console.log(`  bge    BGE-M3 1024 dims  (Cloudflare Workers AI, ~18ms, free)`);
              console.log(`  qwen   Qwen3-8B 4096 dims (HuggingFace, ~69ms, $0.80/hr)`);
              console.log(`\nUsage: openclaw mr embeddings <model>`);
              return;
            }
            const normalized = model.toLowerCase().trim();
            const valid = ["bge", "bge-m3", "qwen", "qwen3", "qwen3-8b", "default"];
            if (!valid.includes(normalized)) {
              console.error(`Unknown model: "${model}". Valid options: bge, qwen`);
              return;
            }
            if (!memoryKey) { console.error("Not configured. Run: openclaw mr <key>"); return; }
            // "bge" / "default" → clear embeddings (use server default)
            const newEmbeddings = (normalized === "bge" || normalized === "bge-m3" || normalized === "default")
              ? undefined
              : normalized.startsWith("qwen") ? "qwen" : normalized;
            try {
              await setPluginConfig(api, {
                key: memoryKey,
                endpoint: cfg?.endpoint,
                density,
                mode,
                logging,
                ...(newEmbeddings && { embeddings: newEmbeddings }),
              });
              const label = newEmbeddings || "bge (default)";
              console.log(`✓ Embedding model set to ${label}`);
              console.log(`  All future prepare/ingest/search will use ${label}`);
              console.log(`\n  Restart gateway to apply: openclaw gateway restart`);
            } catch (err) {
              console.error(`Failed: ${err instanceof Error ? err.message : String(err)}`);
            }
          });

        mr.command("status")
          .description("Show MemoryRouter vault stats")
          .option("--json", "JSON output")
          .action(async (opts) => {
            if (!memoryKey) {
              if (opts.json) {
                console.log(JSON.stringify({ enabled: false, key: null }, null, 2));
              } else {
                console.log("MemoryRouter Status");
                console.log("───────────────────────────");
                console.log(`Enabled:     ✗ No (no key configured)`);
                console.log(`\nGet your free API key: https://memoryrouter.ai`);
                console.log(`Then run: openclaw mr <key>`);
              }
              return;
            }
            try {
              const statsUrl = embeddings
                ? `${endpoint}/v1/memory/stats?embeddings=${encodeURIComponent(embeddings)}`
                : `${endpoint}/v1/memory/stats`;
              const res = await fetch(statsUrl, {
                headers: { Authorization: `Bearer ${memoryKey}` },
              });
              const data = (await res.json()) as { totalVectors?: number; totalTokens?: number };
              if (opts.json) {
                console.log(JSON.stringify({ enabled: true, key: memoryKey, density, mode, embeddings: embeddings || "bge", stats: data }, null, 2));
              } else {
                console.log("MemoryRouter Status");
                console.log("───────────────────────────");
                console.log(`Enabled:     ✓ Yes`);
                console.log(`Key:         ${memoryKey.slice(0, 6)}...${memoryKey.slice(-3)}`);
                console.log(`Mode:        ${mode}`);
                console.log(`Density:     ${density}`);
                console.log(`Embeddings:  ${embeddings || "bge (default)"}`);
                console.log(`Endpoint:    ${endpoint}`);
                console.log(`Memories:    ${data.totalVectors ?? 0}`);
                console.log(`Tokens:      ${data.totalTokens ?? 0}`);
              }
            } catch (err) {
              console.error(`Failed: ${err instanceof Error ? err.message : String(err)}`);
            }
          });

        mr.command("upload")
          .description("Upload workspace + session history to vault")
          .argument("[path]", "Specific file or directory to upload")
          .option("--workspace <dir>", "Workspace directory")
          .option("--brain <dir>", "State directory with sessions")
          .action(async (targetPath: string | undefined, opts: { workspace?: string; brain?: string }) => {
            if (!memoryKey) { console.error("Not configured. Run: openclaw mr <key>"); return; }
            const os = await import("node:os");
            const path = await import("node:path");
            const stateDir = opts.brain ? path.resolve(opts.brain) : path.join(os.homedir(), ".openclaw");
            const configWorkspace = (api.config as any).workspace || (api.config as any).agents?.defaults?.workspace;
            const workspacePath = opts.workspace
              ? path.resolve(opts.workspace)
              : configWorkspace
                ? path.resolve(configWorkspace.replace(/^~/, os.homedir()))
                : path.join(os.homedir(), ".openclaw", "workspace");
            const { runUpload } = await import("./upload.js");
            await runUpload({ memoryKey, endpoint, targetPath, stateDir, workspacePath, hasWorkspaceFlag: !!opts.workspace, hasBrainFlag: !!opts.brain, embeddings });
          });

        mr.command("search")
          .description("Search memories in vault")
          .argument("<query>", "Search query")
          .option("-n, --limit <number>", "Number of results", "5")
          .option("--json", "Output raw JSON response")
          .action(async (query: string, opts: { limit: string; json?: boolean }) => {
            if (!memoryKey) { console.error("Not configured. Run: openclaw mr <key>"); return; }
            const limit = parseInt(opts.limit, 10) || 5;
            try {
              const res = await fetch(`${endpoint}/v1/memory/search`, {
                method: "POST",
                headers: {
                  Authorization: `Bearer ${memoryKey}`,
                  "Content-Type": "application/json",
                  ...(embeddings && { "X-Embedding-Model": embeddings }),
                },
                body: JSON.stringify({ query, limit }),
              });
              if (!res.ok) {
                const err = await res.json() as { error?: string };
                console.error(`Search failed: ${err.error || `HTTP ${res.status}`}`);
                return;
              }
              const data = await res.json() as {
                query: string;
                sessionId: string | null;
                memoryKey: string;
                totalMemories: number;
                tokenCount: number;
                windowBreakdown: Record<string, number>;
                metrics: Record<string, number>;
                memories: Array<{
                  id: number;
                  content: string;
                  score: number;
                  window: string;
                  timestampHuman: string;
                  timestamp: number;
                  role: string;
                  source: string;
                }>;
              };

              // Raw JSON output
              if (opts.json) {
                console.log(JSON.stringify(data, null, 2));
                return;
              }

              if (data.totalMemories === 0) {
                console.log("No results found.");
                return;
              }
              const modelLabel = embeddings ? ` (${embeddings})` : " (bge)";
              console.log(`\n🔍 ${data.totalMemories} results for "${query}"${modelLabel}\n`);
              for (let i = 0; i < data.memories.length; i++) {
                const m = data.memories[i];
                const score = (m.score * 100).toFixed(1);
                const date = new Date(m.timestampHuman).toLocaleDateString("en-US", {
                  month: "short", day: "numeric", year: "numeric",
                });
                const windowIcon = m.window === "immediate" ? "⚡" : m.window === "short" ? "🔥" : m.window === "long" ? "🧠" : "📚";
                console.log(`  ${i + 1}. ${windowIcon} ${score}% | ${m.role} | ${date} | id:${m.id}`);
                console.log(`     ${m.content}\n`);
              }
              console.log(`  Tokens: ${data.tokenCount} | Windows: ${JSON.stringify(data.windowBreakdown)}`);
              if (data.metrics) {
                const ms = data.metrics;
                console.log(`  Latency: ${ms.raceMs ?? ms.totalMs ?? '-'}ms`);
              }
            } catch (err) {
              console.error(`Failed: ${err instanceof Error ? err.message : String(err)}`);
            }
          });

        mr.command("delete")
          .description("Clear all memories from vault")
          .action(async () => {
            if (!memoryKey) { console.error("Not configured. Run: openclaw mr <key>"); return; }
            try {
              const deleteUrl = embeddings
                ? `${endpoint}/v1/memory?embeddings=${encodeURIComponent(embeddings)}`
                : `${endpoint}/v1/memory`;
              const res = await fetch(deleteUrl, {
                method: "DELETE",
                headers: { Authorization: `Bearer ${memoryKey}` },
              });
              const data = (await res.json()) as { message?: string };
              const modelLabel = embeddings ? ` (${embeddings})` : "";
              console.log(`✓ ${data.message || `Vault cleared${modelLabel}`}`);
            } catch (err) {
              console.error(`Failed: ${err instanceof Error ? err.message : String(err)}`);
            }
          });
      },
      { commands: ["mr"] },
    );

    // ==================================================================
    // Service
    // ==================================================================

    api.registerService({
      id: "mr-memory",
      start: () => {
        if (memoryKey) api.logger.info?.(`memoryrouter: active (key: ${memoryKey.slice(0, 6)}...)`);
      },
      stop: () => {
        api.logger.info?.("memoryrouter: stopped");
      },
    });
  },
};

export default memoryRouterPlugin;
