// @troykelly/openclaw-projects/dist/tools/api-restore.js
/**
 * api_restore tool implementation.
 * Restore a soft-deleted API source.
 * Part of API Onboarding feature (#1785).
 */
import { z } from 'zod';
import { ApiSourceService } from '../services/api-source-service.js';
import { sanitizeErrorMessage } from '../utils/sanitize.js';
/** Parameters for api_restore tool */
export const ApiRestoreParamsSchema = z.object({
    id: z.string().uuid('id must be a valid UUID'),
});
export function createApiRestoreTool(options) {
    const { client, logger, user_id } = options;
    const service = new ApiSourceService(client);
    return {
        name: 'api_restore',
        description: 'Restore a previously soft-deleted API source.',
        parameters: ApiRestoreParamsSchema,
        async execute(params) {
            const parseResult = ApiRestoreParamsSchema.safeParse(params);
            if (!parseResult.success) {
                return { success: false, error: parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ') };
            }
            try {
                const response = await service.restore(parseResult.data.id, { user_id });
                if (!response.success) {
                    return { success: false, error: response.error.message || 'Failed to restore API source' };
                }
                const source = response.data.data;
                const content = `Restored API source: ${source.name} (${source.status})`;
                return { success: true, data: { content, details: source } };
            }
            catch (error) {
                logger.error('api_restore failed', { user_id, error: error instanceof Error ? error.message : String(error) });
                return { success: false, error: sanitizeErrorMessage(error) };
            }
        },
    };
}
