// @troykelly/openclaw-projects/dist/tools/api-get.js
/**
 * api_get tool implementation.
 * Retrieve details about a specific onboarded API source.
 * Part of API Onboarding feature (#1785).
 */
import { z } from 'zod';
import { ApiSourceService } from '../services/api-source-service.js';
import { sanitizeErrorMessage } from '../utils/sanitize.js';
/** Parameters for api_get tool */
export const ApiGetParamsSchema = z.object({
    id: z.string().uuid('id must be a valid UUID'),
});
export function createApiGetTool(options) {
    const { client, logger, user_id } = options;
    const service = new ApiSourceService(client);
    return {
        name: 'api_get',
        description: 'Get details about a specific onboarded API source including its status, spec version, and tags.',
        parameters: ApiGetParamsSchema,
        async execute(params) {
            const parseResult = ApiGetParamsSchema.safeParse(params);
            if (!parseResult.success) {
                return { success: false, error: parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ') };
            }
            try {
                const response = await service.get(parseResult.data.id, { user_id });
                if (!response.success) {
                    return { success: false, error: response.error.message || 'Failed to get API source' };
                }
                const source = response.data.data;
                const content = `API Source: ${source.name} (${source.status}) — ${source.spec_version ?? 'unknown version'}`;
                return { success: true, data: { content, details: source } };
            }
            catch (error) {
                logger.error('api_get failed', { user_id, error: error instanceof Error ? error.message : String(error) });
                return { success: false, error: sanitizeErrorMessage(error) };
            }
        },
    };
}
