// @troykelly/openclaw-projects/dist/tools/context-search.js
/**
 * context_search tool implementation.
 * Provides unified cross-entity recall across memories, todos, projects, and messages.
 * Fans out to memory, work-item, and message search APIs in parallel, normalizes scores,
 * and returns a blended ranked list.
 *
 * Part of Issue #1219, #1222.
 */
import { z } from 'zod';
import { createBoundaryMarkers, sanitizeMetadataField, wrapExternalMessage } from '../utils/injection-protection.js';
import { sanitizeErrorMessage } from '../utils/sanitize.js';
/** Supported entity types for cross-entity search */
export const EntityType = z.enum(['memory', 'todo', 'project', 'message', 'dev_session']);
/** Parameters for context_search tool */
export const ContextSearchParamsSchema = z.object({
    query: z.string().min(1, 'Query cannot be empty').max(1000, 'Query must be 1000 characters or less'),
    entity_types: z.array(EntityType).optional(),
    limit: z.number().int().min(1).max(50).optional(),
    // TODO: location and location_weight params will be added when #1204 (geo-aware memory) ships
});
/**
 * Sanitize query input to prevent injection and remove control characters.
 */
function sanitizeQuery(query) {
    return query.replace(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/g, '').trim();
}
/**
 * Normalize an array of scores to 0-1 by dividing each by the max score in the set.
 * If maxScore <= 0, all scores are set to 0. All results are clamped to [0, 1].
 */
function normalizeScores(items) {
    if (items.length === 0)
        return;
    const maxScore = Math.max(...items.map((i) => i.score));
    if (maxScore <= 0) {
        for (const item of items) {
            item.score = 0;
        }
        return;
    }
    for (const item of items) {
        item.score = Math.max(0, Math.min(1, item.score / maxScore));
    }
}
/**
 * Determine entity_type for a work item based on its metadata.kind.
 */
function classifyWorkItem(kind) {
    return kind === 'project' ? 'project' : 'todo';
}
/**
 * Format context search results as a text list with entity type annotations.
 */
function formatResultsAsText(results) {
    if (results.length === 0) {
        return 'No matching results found across memories, todos, projects, messages, or dev sessions.';
    }
    return results
        .map((r) => {
        const scoreStr = r.score.toFixed(2);
        const snippetStr = r.snippet ? ` - ${r.snippet}` : '';
        // Include metadata annotations for messages (channel, timestamp)
        let metaStr = '';
        if (r.entity_type === 'message' && r.metadata) {
            const parts = [];
            if (r.metadata.channel)
                parts.push(r.metadata.channel);
            if (r.metadata.received_at)
                parts.push(r.metadata.received_at);
            if (parts.length > 0)
                metaStr = ` [${parts.join(', ')}]`;
        }
        return `- [${r.entity_type}] ${r.title}${snippetStr}${metaStr} (score: ${scoreStr})`;
    })
        .join('\n');
}
/**
 * Creates the context_search tool.
 */
export function createContextSearchTool(options) {
    const { client, logger, user_id, namespaces } = options;
    return {
        name: 'context_search',
        description: 'Search across memories, todos, projects, messages, and dev sessions simultaneously. Use when you need broad context about a topic, ' +
            'person, or project. Returns a blended ranked list from all entity types. Optionally filter by entity_types to narrow the search.',
        parameters: ContextSearchParamsSchema,
        async execute(params) {
            // Validate parameters
            const parseResult = ContextSearchParamsSchema.safeParse(params);
            if (!parseResult.success) {
                const errorMessage = parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ');
                return { success: false, error: errorMessage };
            }
            const { query, entity_types, limit = 10 } = parseResult.data;
            // Sanitize query
            const sanitizedQuery = sanitizeQuery(query);
            if (sanitizedQuery.length === 0) {
                return { success: false, error: 'Query cannot be empty after sanitization' };
            }
            // Determine which searches to run
            const searchMemories = !entity_types || entity_types.includes('memory');
            const searchWorkItems = !entity_types || entity_types.includes('todo') || entity_types.includes('project');
            const searchMessages = !entity_types || entity_types.includes('message');
            const searchDevSessions = !entity_types || entity_types.includes('dev_session');
            logger.info('context_search invoked', {
                user_id,
                queryLength: sanitizedQuery.length,
                limit,
                entity_types: entity_types ?? 'all',
            });
            const taggedPromises = [];
            if (searchMemories) {
                const memoryParams = new URLSearchParams({
                    q: sanitizedQuery,
                    limit: String(limit),
                    user_email: user_id,
                });
                // Namespace scoping for memory search (Issue #1834)
                if (namespaces && namespaces.length > 0) {
                    memoryParams.set('namespaces', namespaces.join(','));
                }
                taggedPromises.push({
                    tag: 'memory',
                    promise: client.get(`/memories/search?${memoryParams.toString()}`, { user_id }),
                });
            }
            if (searchWorkItems) {
                const searchParams = new URLSearchParams({
                    q: sanitizedQuery,
                    types: 'work_item',
                    limit: String(Math.min(limit * 3, 50)), // Over-fetch for client-side filtering
                    semantic: 'true',
                    user_email: user_id,
                });
                taggedPromises.push({
                    tag: 'work_item',
                    promise: client.get(`/search?${searchParams.toString()}`, { user_id }),
                });
            }
            if (searchMessages) {
                const messageParams = new URLSearchParams({
                    q: sanitizedQuery,
                    types: 'message',
                    limit: String(Math.min(limit * 2, 50)),
                    semantic: 'true',
                    user_email: user_id,
                });
                taggedPromises.push({
                    tag: 'message',
                    promise: client.get(`/search?${messageParams.toString()}`, { user_id }),
                });
            }
            if (searchDevSessions) {
                const devSessionParams = new URLSearchParams({
                    q: sanitizedQuery,
                    limit: String(limit),
                });
                taggedPromises.push({
                    tag: 'dev_session',
                    promise: client.get(`/dev-sessions/search?${devSessionParams.toString()}`, { user_id, user_email: user_id }),
                });
            }
            const settled = await Promise.allSettled(taggedPromises.map((tp) => tp.promise));
            // Process results
            const warnings = [];
            const memoryItems = [];
            const workItemResults = [];
            const messageItems = [];
            const devSessionItems = [];
            for (let i = 0; i < settled.length; i++) {
                const entry = settled[i];
                const tag = taggedPromises[i].tag;
                if (entry.status === 'rejected') {
                    const safeMsg = sanitizeErrorMessage(entry.reason);
                    warnings.push(`${tag} search failed: ${safeMsg}`);
                    logger.error(`context_search ${tag} search failed`, { user_id, error: safeMsg });
                    continue;
                }
                const response = entry.value;
                if (!response.success) {
                    const safeMsg = sanitizeErrorMessage(response.error?.message ?? 'Unknown error');
                    warnings.push(`${tag} search failed: ${safeMsg}`);
                    logger.error(`context_search ${tag} API error`, { user_id, error: safeMsg });
                    continue;
                }
                if (tag === 'memory') {
                    const data = response.data;
                    const rawResults = data.results ?? [];
                    for (const m of rawResults) {
                        memoryItems.push({
                            id: m.id,
                            entity_type: 'memory',
                            title: m.content,
                            snippet: '',
                            score: m.similarity ?? 0,
                        });
                    }
                }
                else if (tag === 'message') {
                    const data = response.data;
                    const rawResults = data.results ?? [];
                    // Generate a per-invocation nonce for boundary markers (#1255)
                    const { nonce: msgNonce } = createBoundaryMarkers();
                    for (const msg of rawResults) {
                        // Sanitize external message content to prevent prompt injection.
                        // Both title and snippet are wrapped since messages are external/untrusted content
                        // mixed alongside trusted entity results from memories/todos/projects.
                        const safeTitle = wrapExternalMessage(sanitizeMetadataField(msg.title ?? '', msgNonce), { nonce: msgNonce });
                        const safeSnippet = msg.snippet ? wrapExternalMessage(sanitizeMetadataField(msg.snippet.substring(0, 100), msgNonce), { nonce: msgNonce }) : '';
                        const meta = {};
                        if (msg.metadata?.channel)
                            meta.channel = sanitizeMetadataField(msg.metadata.channel, msgNonce);
                        if (msg.metadata?.received_at)
                            meta.received_at = sanitizeMetadataField(msg.metadata.received_at, msgNonce);
                        messageItems.push({
                            id: msg.id,
                            entity_type: 'message',
                            title: safeTitle,
                            snippet: safeSnippet,
                            score: msg.score ?? 0,
                            metadata: Object.keys(meta).length > 0 ? meta : undefined,
                        });
                    }
                }
                else if (tag === 'dev_session') {
                    const data = response.data;
                    const rawResults = data.items ?? [];
                    for (const ds of rawResults) {
                        devSessionItems.push({
                            id: ds.id,
                            entity_type: 'dev_session',
                            title: ds.session_name,
                            snippet: ds.task_summary || ds.completion_summary || '',
                            score: ds.similarity ?? 0,
                            metadata: {
                                ...(ds.status ? { status: ds.status } : {}),
                                ...(ds.branch ? { branch: ds.branch } : {}),
                            },
                        });
                    }
                }
                else {
                    const data = response.data;
                    const rawResults = data.results ?? [];
                    for (const w of rawResults) {
                        const entity_type = classifyWorkItem(w.metadata?.kind);
                        workItemResults.push({
                            id: w.id,
                            entity_type: entity_type,
                            title: w.title,
                            snippet: w.snippet,
                            score: w.score ?? 0,
                        });
                    }
                }
            }
            // If all searches failed, return error
            if (memoryItems.length === 0 && workItemResults.length === 0 && messageItems.length === 0 && devSessionItems.length === 0 && warnings.length === settled.length && settled.length > 0) {
                return {
                    success: false,
                    error: `All searches failed: ${warnings.join('; ')}`,
                };
            }
            // Normalize scores within each category
            normalizeScores(memoryItems);
            normalizeScores(workItemResults);
            normalizeScores(messageItems);
            normalizeScores(devSessionItems);
            // Merge all results
            let allResults = [...memoryItems, ...workItemResults, ...messageItems, ...devSessionItems];
            // Filter by entity_types if specified
            if (entity_types) {
                allResults = allResults.filter((r) => entity_types.includes(r.entity_type));
            }
            // Sort by normalized score descending
            allResults.sort((a, b) => b.score - a.score);
            // Apply limit
            allResults = allResults.slice(0, limit);
            // Format output
            const content = formatResultsAsText(allResults);
            logger.debug('context_search completed', {
                user_id,
                resultCount: allResults.length,
                warningCount: warnings.length,
            });
            return {
                success: true,
                data: {
                    content,
                    details: {
                        count: allResults.length,
                        results: allResults,
                        warnings: warnings.length > 0 ? warnings : undefined,
                        user_id,
                    },
                },
            };
        },
    };
}
