// @troykelly/openclaw-projects/dist/tools/notes.js
/**
 * Note tools for OpenClaw agents.
 * Part of Epic #339, Issues #359, #360, #361, #362
 *
 * Provides tools for:
 * - note_create: Create a new note
 * - note_get: Get a note by ID
 * - note_update: Update a note
 * - note_delete: Delete a note
 * - note_search: Search notes
 */
import { z } from 'zod';
import { sanitizeText, sanitizeErrorMessage, truncateForPreview } from '../utils/sanitize.js';
// ─────────────────────────────────────────────────────────────────────────────
// Common Types
// ─────────────────────────────────────────────────────────────────────────────
/** Note visibility levels */
export const NoteVisibility = z.enum(['private', 'shared', 'public']);
// ─────────────────────────────────────────────────────────────────────────────
// note_create Tool
// ─────────────────────────────────────────────────────────────────────────────
export const NoteCreateParamsSchema = z.object({
    title: z.string().min(1, 'Title cannot be empty').max(500, 'Title must be 500 characters or less'),
    content: z.string().min(1, 'Content cannot be empty').max(100000, 'Content must be 100,000 characters or less'),
    notebook_id: z.string().uuid().optional(),
    tags: z.array(z.string()).max(20).optional(),
    visibility: NoteVisibility.optional().default('private'),
    summary: z.string().max(1000).optional(),
});
export function createNoteCreateTool(options) {
    const { client, logger, config, user_id } = options;
    return {
        name: 'note_create',
        description: 'Creates a new note with markdown content. Use for long-form documents such as meeting notes, documentation, or research. For discrete facts or preferences, use memory_store instead. Persists the note with optional notebook assignment.',
        parameters: NoteCreateParamsSchema,
        async execute(params) {
            const parseResult = NoteCreateParamsSchema.safeParse(params);
            if (!parseResult.success) {
                const errorMessage = parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ');
                return { success: false, error: errorMessage };
            }
            const { title, content, notebook_id, tags, visibility, summary } = parseResult.data;
            const sanitizedTitle = sanitizeText(title);
            const sanitizedContent = sanitizeText(content);
            if (sanitizedTitle.length === 0) {
                return { success: false, error: 'Title cannot be empty after sanitization' };
            }
            if (sanitizedContent.length === 0) {
                return { success: false, error: 'Content cannot be empty after sanitization' };
            }
            logger.info('note_create invoked', {
                user_id,
                titleLength: sanitizedTitle.length,
                contentLength: sanitizedContent.length,
                notebook_id,
                visibility,
            });
            try {
                const response = await client.post('/notes', {
                    user_email: user_id,
                    title: sanitizedTitle,
                    content: sanitizedContent,
                    notebook_id: notebook_id,
                    tags,
                    visibility,
                    summary,
                }, { user_id, user_email: user_id });
                if (!response.success) {
                    logger.error('note_create API error', {
                        user_id,
                        status: response.error.status,
                        code: response.error.code,
                    });
                    return {
                        success: false,
                        error: response.error.message || 'Failed to create note',
                    };
                }
                const note = response.data;
                logger.debug('note_create completed', {
                    user_id,
                    noteId: note.id,
                });
                return {
                    success: true,
                    data: {
                        id: note.id,
                        title: note.title,
                        notebook_id: note.notebook_id,
                        visibility: note.visibility,
                        created_at: note.created_at,
                        ...(config.baseUrl ? { url: `${config.baseUrl}/notes/${note.id}` } : {}),
                    },
                };
            }
            catch (error) {
                logger.error('note_create failed', {
                    user_id,
                    error: error instanceof Error ? error.message : String(error),
                });
                return {
                    success: false,
                    error: sanitizeErrorMessage(error),
                };
            }
        },
    };
}
// ─────────────────────────────────────────────────────────────────────────────
// note_get Tool
// ─────────────────────────────────────────────────────────────────────────────
export const NoteGetParamsSchema = z.object({
    noteId: z.string().uuid('Note ID must be a valid UUID'),
    include_versions: z.boolean().optional().default(false),
});
export function createNoteGetTool(options) {
    const { client, logger, config, user_id } = options;
    return {
        name: 'note_get',
        description: 'Gets a note by its ID. Returns the full content, metadata, and optional version history. Only accessible if you have permission to view the note. Read-only.',
        parameters: NoteGetParamsSchema,
        async execute(params) {
            const parseResult = NoteGetParamsSchema.safeParse(params);
            if (!parseResult.success) {
                const errorMessage = parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ');
                return { success: false, error: errorMessage };
            }
            const { noteId, include_versions } = parseResult.data;
            logger.info('note_get invoked', {
                user_id,
                noteId,
                include_versions,
            });
            try {
                const queryParams = new URLSearchParams({ user_email: user_id });
                if (include_versions) {
                    queryParams.set('include_versions', 'true');
                }
                const response = await client.get(`/notes/${noteId}?${queryParams}`, { user_id, user_email: user_id });
                if (!response.success) {
                    if (response.error.status === 404) {
                        return { success: false, error: 'Note not found or access denied' };
                    }
                    logger.error('note_get API error', {
                        user_id,
                        noteId,
                        status: response.error.status,
                    });
                    return {
                        success: false,
                        error: response.error.message || 'Failed to get note',
                    };
                }
                const note = response.data;
                logger.debug('note_get completed', {
                    user_id,
                    noteId,
                });
                return {
                    success: true,
                    data: {
                        id: note.id,
                        title: note.title,
                        content: note.content,
                        notebook_id: note.notebook_id,
                        tags: note.tags,
                        visibility: note.visibility,
                        summary: note.summary,
                        is_pinned: note.is_pinned,
                        created_at: note.created_at,
                        updated_at: note.updated_at,
                        ...(config.baseUrl ? { url: `${config.baseUrl}/notes/${note.id}` } : {}),
                        version_count: note.version_count,
                    },
                };
            }
            catch (error) {
                logger.error('note_get failed', {
                    user_id,
                    noteId,
                    error: error instanceof Error ? error.message : String(error),
                });
                return {
                    success: false,
                    error: sanitizeErrorMessage(error),
                };
            }
        },
    };
}
// ─────────────────────────────────────────────────────────────────────────────
// note_update Tool
// ─────────────────────────────────────────────────────────────────────────────
export const NoteUpdateParamsSchema = z.object({
    noteId: z.string().uuid('Note ID must be a valid UUID'),
    title: z.string().min(1).max(500).optional(),
    content: z.string().min(1).max(100000).optional(),
    notebook_id: z.string().uuid().nullable().optional(),
    tags: z.array(z.string()).max(20).optional(),
    visibility: NoteVisibility.optional(),
    summary: z.string().max(1000).nullable().optional(),
    is_pinned: z.boolean().optional(),
});
export function createNoteUpdateTool(options) {
    const { client, logger, config, user_id } = options;
    return {
        name: 'note_update',
        description: 'Updates an existing note. Can modify title, content, tags, visibility, or move to a different notebook. Creates a version in history when content changes. Only the note owner can update.',
        parameters: NoteUpdateParamsSchema,
        async execute(params) {
            const parseResult = NoteUpdateParamsSchema.safeParse(params);
            if (!parseResult.success) {
                const errorMessage = parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ');
                return { success: false, error: errorMessage };
            }
            const { noteId, title, content, notebook_id, tags, visibility, summary, is_pinned } = parseResult.data;
            // Track what's being changed
            const changes = [];
            const updateData = { user_email: user_id };
            if (title !== undefined) {
                const sanitizedTitle = sanitizeText(title);
                if (sanitizedTitle.length === 0) {
                    return { success: false, error: 'Title cannot be empty after sanitization' };
                }
                updateData.title = sanitizedTitle;
                changes.push('title');
            }
            if (content !== undefined) {
                const sanitizedContent = sanitizeText(content);
                if (sanitizedContent.length === 0) {
                    return { success: false, error: 'Content cannot be empty after sanitization' };
                }
                updateData.content = sanitizedContent;
                changes.push('content');
            }
            if (notebook_id !== undefined) {
                updateData.notebook_id = notebook_id;
                changes.push('notebook');
            }
            if (tags !== undefined) {
                updateData.tags = tags;
                changes.push('tags');
            }
            if (visibility !== undefined) {
                updateData.visibility = visibility;
                changes.push('visibility');
            }
            if (summary !== undefined) {
                updateData.summary = summary;
                changes.push('summary');
            }
            if (is_pinned !== undefined) {
                updateData.is_pinned = is_pinned;
                changes.push('is_pinned');
            }
            if (changes.length === 0) {
                return { success: false, error: 'No changes specified' };
            }
            logger.info('note_update invoked', {
                user_id,
                noteId,
                changes,
            });
            try {
                const response = await client.put(`/notes/${noteId}`, updateData, { user_id, user_email: user_id });
                if (!response.success) {
                    if (response.error.status === 404) {
                        return { success: false, error: 'Note not found or access denied' };
                    }
                    if (response.error.status === 403) {
                        return { success: false, error: 'You do not have permission to update this note' };
                    }
                    logger.error('note_update API error', {
                        user_id,
                        noteId,
                        status: response.error.status,
                    });
                    return {
                        success: false,
                        error: response.error.message || 'Failed to update note',
                    };
                }
                const note = response.data;
                logger.debug('note_update completed', {
                    user_id,
                    noteId,
                    changes,
                });
                return {
                    success: true,
                    data: {
                        id: note.id,
                        title: note.title,
                        visibility: note.visibility,
                        updated_at: note.updated_at,
                        ...(config.baseUrl ? { url: `${config.baseUrl}/notes/${note.id}` } : {}),
                        changes,
                    },
                };
            }
            catch (error) {
                logger.error('note_update failed', {
                    user_id,
                    noteId,
                    error: error instanceof Error ? error.message : String(error),
                });
                return {
                    success: false,
                    error: sanitizeErrorMessage(error),
                };
            }
        },
    };
}
// ─────────────────────────────────────────────────────────────────────────────
// note_delete Tool
// ─────────────────────────────────────────────────────────────────────────────
export const NoteDeleteParamsSchema = z.object({
    noteId: z.string().uuid('Note ID must be a valid UUID'),
});
export function createNoteDeleteTool(options) {
    const { client, logger, user_id } = options;
    return {
        name: 'note_delete',
        description: 'Soft-deletes a note. The note can be restored later. Only the note owner can delete. Use note_search or note_get to find the note ID first.',
        parameters: NoteDeleteParamsSchema,
        async execute(params) {
            const parseResult = NoteDeleteParamsSchema.safeParse(params);
            if (!parseResult.success) {
                const errorMessage = parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ');
                return { success: false, error: errorMessage };
            }
            const { noteId } = parseResult.data;
            logger.info('note_delete invoked', {
                user_id,
                noteId,
            });
            try {
                const response = await client.delete(`/notes/${noteId}?user_email=${encodeURIComponent(user_id)}`, { user_id, user_email: user_id });
                if (!response.success) {
                    if (response.error.status === 404) {
                        return { success: false, error: 'Note not found' };
                    }
                    if (response.error.status === 403) {
                        return { success: false, error: 'Only the note owner can delete this note' };
                    }
                    logger.error('note_delete API error', {
                        user_id,
                        noteId,
                        status: response.error.status,
                    });
                    return {
                        success: false,
                        error: response.error.message || 'Failed to delete note',
                    };
                }
                logger.debug('note_delete completed', {
                    user_id,
                    noteId,
                });
                return {
                    success: true,
                    data: {
                        id: noteId,
                        message: 'Note deleted successfully',
                    },
                };
            }
            catch (error) {
                logger.error('note_delete failed', {
                    user_id,
                    noteId,
                    error: error instanceof Error ? error.message : String(error),
                });
                return {
                    success: false,
                    error: sanitizeErrorMessage(error),
                };
            }
        },
    };
}
// ─────────────────────────────────────────────────────────────────────────────
// note_search Tool
// ─────────────────────────────────────────────────────────────────────────────
export const NoteSearchParamsSchema = z.object({
    query: z.string().min(1, 'Search query cannot be empty').max(500),
    search_type: z.enum(['hybrid', 'text', 'semantic']).optional().default('hybrid'),
    notebook_id: z.string().uuid().optional(),
    tags: z.array(z.string()).optional(),
    visibility: NoteVisibility.optional(),
    limit: z.number().min(1).max(50).optional().default(20),
    offset: z.number().min(0).optional().default(0),
});
export function createNoteSearchTool(options) {
    const { client, logger, config, user_id } = options;
    return {
        name: 'note_search',
        description: 'Searches notes using text, semantic, or hybrid search. Searches notes only. For cross-entity search, prefer context_search. Respects privacy settings; private notes are only visible to their owner.',
        parameters: NoteSearchParamsSchema,
        async execute(params) {
            const parseResult = NoteSearchParamsSchema.safeParse(params);
            if (!parseResult.success) {
                const errorMessage = parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ');
                return { success: false, error: errorMessage };
            }
            const { query, search_type, notebook_id, tags, visibility, limit, offset } = parseResult.data;
            logger.info('note_search invoked', {
                user_id,
                queryLength: query.length,
                search_type,
                notebook_id,
                limit,
            });
            try {
                const queryParams = new URLSearchParams({
                    user_email: user_id,
                    q: query,
                    search_type,
                    limit: String(limit),
                    offset: String(offset),
                });
                if (notebook_id)
                    queryParams.set('notebook_id', notebook_id);
                if (tags && tags.length > 0)
                    queryParams.set('tags', tags.join(','));
                if (visibility)
                    queryParams.set('visibility', visibility);
                const response = await client.get(`/notes/search?${queryParams}`, { user_id, user_email: user_id, isAgent: true });
                if (!response.success) {
                    logger.error('note_search API error', {
                        user_id,
                        status: response.error.status,
                    });
                    return {
                        success: false,
                        error: response.error.message || 'Failed to search notes',
                    };
                }
                const searchResult = response.data;
                logger.debug('note_search completed', {
                    user_id,
                    resultsCount: searchResult.results.length,
                    total: searchResult.total,
                });
                return {
                    success: true,
                    data: {
                        query: searchResult.query,
                        search_type: searchResult.search_type,
                        results: searchResult.results.map((r) => ({
                            id: r.id,
                            title: r.title,
                            snippet: truncateForPreview(r.snippet, 200),
                            score: r.score,
                            tags: r.tags,
                            visibility: r.visibility,
                            ...(config.baseUrl ? { url: `${config.baseUrl}/notes/${r.id}` } : {}),
                        })),
                        total: searchResult.total,
                        limit: searchResult.limit,
                        offset: searchResult.offset,
                    },
                };
            }
            catch (error) {
                logger.error('note_search failed', {
                    user_id,
                    error: error instanceof Error ? error.message : String(error),
                });
                return {
                    success: false,
                    error: sanitizeErrorMessage(error),
                };
            }
        },
    };
}
