// @troykelly/openclaw-projects/dist/tools/projects.js
/**
 * Project management tools implementation.
 * Provides project_list, project_get, and project_create tools.
 */
import { z } from 'zod';
import { sanitizeErrorMessage } from '../utils/sanitize.js';
/** UUID validation regex */
const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
/** Project status enum */
export const ProjectStatus = z.enum(['active', 'completed', 'archived']);
// ==================== project_list ====================
/** Parameters for project_list tool */
export const ProjectListParamsSchema = z.object({
    status: ProjectStatus.optional(),
    limit: z.number().int().min(1).max(100).optional(),
    offset: z.number().int().min(0).optional(),
});
/**
 * Strip HTML tags from a string.
 * Also removes content inside script and style tags for security.
 */
function stripHtml(text) {
    return text
        .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')
        .replace(/<style\b[^<]*(?:(?!<\/style>)<[^<]*)*<\/style>/gi, '')
        .replace(/<[^>]*>/g, '')
        .trim();
}
/**
 * Truncate text for display.
 */
function truncate(text, maxLength = 100) {
    if (text.length <= maxLength)
        return text;
    return `${text.substring(0, maxLength)}...`;
}
/**
 * Creates the project_list tool.
 */
export function createProjectListTool(options) {
    const { client, logger, user_id } = options;
    return {
        name: 'project_list',
        description: 'List all projects. Optionally filter by status (active, completed, archived).',
        parameters: ProjectListParamsSchema,
        async execute(params) {
            const parseResult = ProjectListParamsSchema.safeParse(params);
            if (!parseResult.success) {
                const errorMessage = parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ');
                return { success: false, error: errorMessage };
            }
            const { status, limit = 20, offset = 0 } = parseResult.data;
            logger.info('project_list invoked', { user_id, status, limit, offset });
            try {
                const queryParams = new URLSearchParams({
                    item_type: 'project',
                    limit: String(limit),
                    offset: String(offset),
                });
                if (status) {
                    queryParams.set('status', status);
                }
                const response = await client.get(`/work-items?${queryParams.toString()}`, { user_id });
                if (!response.success) {
                    logger.error('project_list API error', {
                        user_id,
                        status: response.error.status,
                        code: response.error.code,
                    });
                    return {
                        success: false,
                        error: response.error.message || 'Failed to list projects',
                    };
                }
                const projects = response.data.items ?? [];
                const total = response.data.total ?? projects.length;
                if (projects.length === 0) {
                    return {
                        success: true,
                        data: {
                            content: 'No projects found.',
                            details: { projects: [], total: 0, user_id },
                        },
                    };
                }
                const content = projects
                    .map((p) => {
                    const name = p.name ?? p.title ?? 'Untitled';
                    const desc = p.description ? ` - ${truncate(p.description, 50)}` : '';
                    return `- **${name}** [${p.status ?? 'unknown'}]${desc}`;
                })
                    .join('\n');
                logger.debug('project_list completed', { user_id, count: projects.length });
                return {
                    success: true,
                    data: {
                        content,
                        details: { projects, total, user_id },
                    },
                };
            }
            catch (error) {
                logger.error('project_list failed', {
                    user_id,
                    error: error instanceof Error ? error.message : String(error),
                });
                return { success: false, error: sanitizeErrorMessage(error) };
            }
        },
    };
}
// ==================== project_get ====================
/** Parameters for project_get tool */
export const ProjectGetParamsSchema = z.object({
    id: z.string().min(1, 'Project ID is required'),
});
/**
 * Creates the project_get tool.
 */
export function createProjectGetTool(options) {
    const { client, logger, user_id } = options;
    return {
        name: 'project_get',
        description: 'Get details of a specific project by ID.',
        parameters: ProjectGetParamsSchema,
        async execute(params) {
            const parseResult = ProjectGetParamsSchema.safeParse(params);
            if (!parseResult.success) {
                const errorMessage = parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ');
                return { success: false, error: errorMessage };
            }
            const { id } = parseResult.data;
            // Validate UUID format
            if (!UUID_REGEX.test(id)) {
                return { success: false, error: 'Invalid project ID format. Expected UUID.' };
            }
            logger.info('project_get invoked', { user_id, project_id: id });
            try {
                const response = await client.get(`/work-items/${id}`, { user_id });
                if (!response.success) {
                    if (response.error.code === 'NOT_FOUND') {
                        return { success: false, error: 'Project not found.' };
                    }
                    logger.error('project_get API error', {
                        user_id,
                        project_id: id,
                        status: response.error.status,
                    });
                    return {
                        success: false,
                        error: response.error.message || 'Failed to get project',
                    };
                }
                const project = response.data;
                const name = project.name ?? project.title ?? 'Untitled';
                const content = `**${name}** [${project.status ?? 'unknown'}]\n\n${project.description ?? 'No description.'}`;
                logger.debug('project_get completed', { user_id, project_id: id });
                return {
                    success: true,
                    data: {
                        content,
                        details: { project, user_id },
                    },
                };
            }
            catch (error) {
                logger.error('project_get failed', {
                    user_id,
                    project_id: id,
                    error: error instanceof Error ? error.message : String(error),
                });
                return { success: false, error: sanitizeErrorMessage(error) };
            }
        },
    };
}
// ==================== project_create ====================
/** Parameters for project_create tool */
export const ProjectCreateParamsSchema = z.object({
    name: z.string().min(1, 'Project name is required').max(200, 'Project name must be 200 characters or less'),
    description: z.string().max(2000, 'Description must be 2000 characters or less').optional(),
});
/**
 * Creates the project_create tool.
 */
export function createProjectCreateTool(options) {
    const { client, logger, user_id } = options;
    return {
        name: 'project_create',
        description: 'Create a new project.',
        parameters: ProjectCreateParamsSchema,
        async execute(params) {
            const parseResult = ProjectCreateParamsSchema.safeParse(params);
            if (!parseResult.success) {
                const errorMessage = parseResult.error.errors.map((e) => `${e.path.join('.')}: ${e.message}`).join(', ');
                return { success: false, error: errorMessage };
            }
            const { name, description } = parseResult.data;
            // Sanitize input
            const sanitizedName = stripHtml(name);
            const sanitizedDescription = description ? stripHtml(description) : undefined;
            if (sanitizedName.length === 0) {
                return { success: false, error: 'Project name cannot be empty after sanitization' };
            }
            logger.info('project_create invoked', {
                user_id,
                nameLength: sanitizedName.length,
            });
            try {
                const response = await client.post('/work-items', {
                    title: sanitizedName,
                    description: sanitizedDescription,
                    item_type: 'project',
                }, { user_id });
                if (!response.success) {
                    logger.error('project_create API error', {
                        user_id,
                        status: response.error.status,
                        code: response.error.code,
                    });
                    return {
                        success: false,
                        error: response.error.message || 'Failed to create project',
                    };
                }
                const newProject = response.data;
                logger.debug('project_create completed', {
                    user_id,
                    project_id: newProject.id,
                });
                return {
                    success: true,
                    data: {
                        content: `Created project "${sanitizedName}" (ID: ${newProject.id})`,
                        details: {
                            id: newProject.id,
                            name: sanitizedName,
                            user_id,
                        },
                    },
                };
            }
            catch (error) {
                logger.error('project_create failed', {
                    user_id,
                    error: error instanceof Error ? error.message : String(error),
                });
                return { success: false, error: sanitizeErrorMessage(error) };
            }
        },
    };
}
