// @troykelly/openclaw-projects/dist/gateway/oauth-rpc-methods.js
/**
 * OAuth Gateway RPC methods for surfacing connected account data to agents.
 * Part of Issue #1054.
 *
 * Registers gateway methods that proxy requests to the openclaw-projects
 * backend API, exposing connected account data (contacts, email, files)
 * in an agent-friendly format with pagination, connection_label, and
 * available_actions metadata.
 */
// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------
/** Build available actions for an account listing based on features and permissions. */
function buildAccountActions(conn) {
    const actions = [];
    if (conn.enabled_features.includes('contacts')) {
        actions.push('list_contacts');
    }
    if (conn.enabled_features.includes('email')) {
        actions.push('list_emails', 'get_email');
        if (conn.permission_level === 'read_write') {
            actions.push('send_email', 'create_draft', 'update_email', 'delete_email');
        }
    }
    if (conn.enabled_features.includes('files')) {
        actions.push('list_files', 'search_files', 'get_file');
    }
    if (conn.enabled_features.includes('calendar')) {
        actions.push('list_events', 'sync_calendar');
        if (conn.permission_level === 'read_write') {
            actions.push('create_event', 'delete_event');
        }
    }
    return actions;
}
/** Build email-specific available actions based on permission level. */
function buildEmailActions(permission_level) {
    const actions = ['list_messages', 'get_message', 'list_threads', 'list_folders'];
    if (permission_level === 'read_write') {
        actions.push('send_email', 'create_draft', 'update_draft', 'update_message', 'delete_message');
    }
    return actions;
}
/** Build file-specific available actions. */
function buildFileActions() {
    return ['list_files', 'search_files', 'get_file'];
}
/** Build calendar-specific available actions based on permission level. */
function buildCalendarActions(permission_level) {
    const actions = ['list_events', 'sync_calendar'];
    if (permission_level === 'read_write') {
        actions.push('create_event', 'delete_event');
    }
    return actions;
}
/**
 * Resolve a connection by ID from the backend.
 * Returns null with logged warning when not found or inactive.
 */
async function resolveConnection(apiClient, user_id, logger, connection_id) {
    const response = await apiClient.get('/oauth/connections', { user_id });
    if (!response.success) {
        logger.error('Failed to fetch OAuth connections', { user_id, error: response.error.message });
        return null;
    }
    const conn = response.data.connections.find((c) => c.id === connection_id);
    if (!conn) {
        logger.debug('OAuth connection not found', { user_id, connection_id });
        return null;
    }
    if (!conn.is_active) {
        logger.debug('OAuth connection is disabled', { user_id, connection_id });
        return null;
    }
    return conn;
}
/**
 * Create OAuth gateway RPC method handlers.
 */
export function createOAuthGatewayMethods(options) {
    const { logger, apiClient, getAgentId } = options;
    return {
        async accountsList(params) {
            const user_id = getAgentId();
            logger.debug('oauth.accounts.list', { user_id });
            const qs = new URLSearchParams();
            if (params.user_email)
                qs.set('user_email', params.user_email);
            if (params.provider)
                qs.set('provider', params.provider);
            const qsStr = qs.toString();
            const path = `/oauth/connections${qsStr ? `?${qsStr}` : ''}`;
            const response = await apiClient.get(path, { user_id });
            if (!response.success) {
                throw new Error(response.error.message || 'Failed to list accounts');
            }
            return {
                accounts: response.data.connections.map((conn) => ({
                    connection_id: conn.id,
                    provider: conn.provider,
                    connection_label: conn.label ?? conn.provider,
                    provider_account_email: conn.provider_account_email,
                    permission_level: conn.permission_level,
                    enabled_features: conn.enabled_features,
                    is_active: conn.is_active,
                    last_sync_at: conn.last_sync_at,
                    sync_status: conn.sync_status,
                    available_actions: buildAccountActions(conn),
                })),
            };
        },
        async contactsList(params) {
            if (!params.connection_id)
                throw new Error('connection_id is required');
            const user_id = getAgentId();
            logger.debug('oauth.contacts.list', { user_id, connection_id: params.connection_id });
            const conn = await resolveConnection(apiClient, user_id, logger, params.connection_id);
            if (!conn)
                throw new Error('Connection not found or inactive');
            if (!conn.enabled_features.includes('contacts')) {
                throw new Error('Contacts feature is not enabled on this connection');
            }
            const qs = new URLSearchParams({ connection_id: params.connection_id });
            if (params.page_token)
                qs.set('page_token', params.page_token);
            const response = await apiClient.get(`/contacts?${qs}`, { user_id });
            if (!response.success) {
                throw new Error(response.error.message || 'Failed to list contacts');
            }
            return {
                connection_label: conn.label ?? conn.provider,
                contacts: response.data.contacts ?? [],
                next_page_token: response.data.next_page_token,
                available_actions: ['list_contacts'],
            };
        },
        async emailList(params) {
            if (!params.connection_id)
                throw new Error('connection_id is required');
            const user_id = getAgentId();
            logger.debug('oauth.email.list', { user_id, connection_id: params.connection_id });
            const conn = await resolveConnection(apiClient, user_id, logger, params.connection_id);
            if (!conn)
                throw new Error('Connection not found or inactive');
            if (!conn.enabled_features.includes('email')) {
                throw new Error('Email feature is not enabled on this connection');
            }
            const qs = new URLSearchParams({ connection_id: params.connection_id });
            if (params.folder_id)
                qs.set('folder_id', params.folder_id);
            if (params.q)
                qs.set('q', params.q);
            if (params.max_results)
                qs.set('max_results', String(params.max_results));
            if (params.page_token)
                qs.set('page_token', params.page_token);
            const response = await apiClient.get(`/email/messages?${qs}`, { user_id });
            if (!response.success) {
                throw new Error(response.error.message || 'Failed to list emails');
            }
            return {
                connection_label: conn.label ?? conn.provider,
                messages: response.data.messages ?? [],
                next_page_token: response.data.next_page_token,
                available_actions: buildEmailActions(conn.permission_level),
            };
        },
        async emailGet(params) {
            if (!params.connection_id)
                throw new Error('connection_id is required');
            if (!params.message_id)
                throw new Error('message_id is required');
            const user_id = getAgentId();
            logger.debug('oauth.email.get', { user_id, connection_id: params.connection_id, message_id: params.message_id });
            const conn = await resolveConnection(apiClient, user_id, logger, params.connection_id);
            if (!conn)
                throw new Error('Connection not found or inactive');
            if (!conn.enabled_features.includes('email')) {
                throw new Error('Email feature is not enabled on this connection');
            }
            const qs = new URLSearchParams({ connection_id: params.connection_id });
            const response = await apiClient.get(`/email/messages/${encodeURIComponent(params.message_id)}?${qs}`, { user_id });
            if (!response.success) {
                throw new Error(response.error.message || 'Failed to get email');
            }
            return {
                connection_label: conn.label ?? conn.provider,
                message: response.data,
                available_actions: buildEmailActions(conn.permission_level),
            };
        },
        async filesList(params) {
            if (!params.connection_id)
                throw new Error('connection_id is required');
            const user_id = getAgentId();
            logger.debug('oauth.files.list', { user_id, connection_id: params.connection_id });
            const conn = await resolveConnection(apiClient, user_id, logger, params.connection_id);
            if (!conn)
                throw new Error('Connection not found or inactive');
            if (!conn.enabled_features.includes('files')) {
                throw new Error('Files feature is not enabled on this connection');
            }
            const qs = new URLSearchParams({ connection_id: params.connection_id });
            if (params.folder_id)
                qs.set('folder_id', params.folder_id);
            if (params.page_token)
                qs.set('page_token', params.page_token);
            const response = await apiClient.get(`/drive/files?${qs}`, { user_id });
            if (!response.success) {
                throw new Error(response.error.message || 'Failed to list files');
            }
            return {
                connection_label: conn.label ?? conn.provider,
                files: response.data.files ?? [],
                next_page_token: response.data.next_page_token,
                available_actions: buildFileActions(),
            };
        },
        async filesSearch(params) {
            if (!params.connection_id)
                throw new Error('connection_id is required');
            if (!params.query)
                throw new Error('query is required');
            const user_id = getAgentId();
            logger.debug('oauth.files.search', { user_id, connection_id: params.connection_id });
            const conn = await resolveConnection(apiClient, user_id, logger, params.connection_id);
            if (!conn)
                throw new Error('Connection not found or inactive');
            if (!conn.enabled_features.includes('files')) {
                throw new Error('Files feature is not enabled on this connection');
            }
            const qs = new URLSearchParams({ connection_id: params.connection_id, q: params.query });
            if (params.page_token)
                qs.set('page_token', params.page_token);
            const response = await apiClient.get(`/drive/files/search?${qs}`, { user_id });
            if (!response.success) {
                throw new Error(response.error.message || 'Failed to search files');
            }
            return {
                connection_label: conn.label ?? conn.provider,
                files: response.data.files ?? [],
                next_page_token: response.data.next_page_token,
                available_actions: buildFileActions(),
            };
        },
        async filesGet(params) {
            if (!params.connection_id)
                throw new Error('connection_id is required');
            if (!params.file_id)
                throw new Error('file_id is required');
            const user_id = getAgentId();
            logger.debug('oauth.files.get', { user_id, connection_id: params.connection_id, file_id: params.file_id });
            const conn = await resolveConnection(apiClient, user_id, logger, params.connection_id);
            if (!conn)
                throw new Error('Connection not found or inactive');
            if (!conn.enabled_features.includes('files')) {
                throw new Error('Files feature is not enabled on this connection');
            }
            const qs = new URLSearchParams({ connection_id: params.connection_id });
            const response = await apiClient.get(`/drive/files/${encodeURIComponent(params.file_id)}?${qs}`, { user_id });
            if (!response.success) {
                throw new Error(response.error.message || 'Failed to get file');
            }
            return {
                connection_label: conn.label ?? conn.provider,
                file: response.data,
                available_actions: buildFileActions(),
            };
        },
        async calendarList(params) {
            if (!params.connection_id)
                throw new Error('connection_id is required');
            const user_id = getAgentId();
            logger.debug('oauth.calendar.list', { user_id, connection_id: params.connection_id });
            const conn = await resolveConnection(apiClient, user_id, logger, params.connection_id);
            if (!conn)
                throw new Error('Connection not found or inactive');
            if (!conn.enabled_features.includes('calendar')) {
                throw new Error('Calendar feature is not enabled on this connection');
            }
            const qs = new URLSearchParams({ connection_id: params.connection_id });
            if (params.time_min)
                qs.set('time_min', params.time_min);
            if (params.time_max)
                qs.set('time_max', params.time_max);
            if (params.max_results)
                qs.set('max_results', String(params.max_results));
            if (params.page_token)
                qs.set('page_token', params.page_token);
            const response = await apiClient.get(`/calendar/events/live?${qs}`, { user_id });
            if (!response.success) {
                throw new Error(response.error.message || 'Failed to list calendar events');
            }
            return {
                connection_label: conn.label ?? conn.provider,
                events: response.data.events ?? [],
                next_page_token: response.data.next_page_token,
                available_actions: buildCalendarActions(conn.permission_level),
            };
        },
        async calendarSync(params) {
            if (!params.connection_id)
                throw new Error('connection_id is required');
            const user_id = getAgentId();
            logger.debug('oauth.calendar.sync', { user_id, connection_id: params.connection_id });
            const conn = await resolveConnection(apiClient, user_id, logger, params.connection_id);
            if (!conn)
                throw new Error('Connection not found or inactive');
            if (!conn.enabled_features.includes('calendar')) {
                throw new Error('Calendar feature is not enabled on this connection');
            }
            const body = { connection_id: params.connection_id };
            if (params.time_min)
                body.time_min = params.time_min;
            if (params.time_max)
                body.time_max = params.time_max;
            const response = await apiClient.post('/sync/calendar', body, { user_id });
            if (!response.success) {
                throw new Error(response.error.message || 'Failed to sync calendar');
            }
            return {
                connection_label: conn.label ?? conn.provider,
                ...response.data,
            };
        },
        async calendarCreate(params) {
            if (!params.connection_id)
                throw new Error('connection_id is required');
            if (!params.title)
                throw new Error('title is required');
            const user_id = getAgentId();
            logger.debug('oauth.calendar.create', { user_id, connection_id: params.connection_id });
            const conn = await resolveConnection(apiClient, user_id, logger, params.connection_id);
            if (!conn)
                throw new Error('Connection not found or inactive');
            if (!conn.enabled_features.includes('calendar')) {
                throw new Error('Calendar feature is not enabled on this connection');
            }
            if (conn.permission_level !== 'read_write') {
                throw new Error('Write permission required to create events');
            }
            const body = {
                connection_id: params.connection_id,
                user_email: conn.provider_account_email ?? '',
                provider: conn.provider,
                title: params.title,
                start_time: params.start_time,
                end_time: params.end_time,
            };
            if (params.description !== undefined)
                body.description = params.description;
            if (params.location !== undefined)
                body.location = params.location;
            if (params.all_day !== undefined)
                body.all_day = params.all_day;
            const response = await apiClient.post('/calendar/events', body, { user_id });
            if (!response.success) {
                throw new Error(response.error.message || 'Failed to create calendar event');
            }
            return {
                connection_label: conn.label ?? conn.provider,
                event: response.data.event,
                available_actions: buildCalendarActions(conn.permission_level),
            };
        },
    };
}
/**
 * Wrap a typed method handler into a GatewayRequestHandler.
 *
 * The SDK's registerGatewayMethod expects a handler that receives
 * `(opts: GatewayRequestHandlerOptions)` and calls `opts.respond()`.
 * Our internal methods use `(params: T) => Promise<R>` for testability.
 * This wrapper bridges the two contracts. (#2031)
 */
function wrapOAuthMethodHandler(handler) {
    return async (opts) => {
        try {
            const result = await handler(opts.params);
            opts.respond(true, result);
        }
        catch (_error) {
            // Return generic error to clients to avoid leaking internal details
            opts.respond(false, undefined, {
                code: 'INTERNAL_ERROR',
                message: 'Internal error',
            });
        }
    };
}
/**
 * Register OAuth gateway RPC methods with the OpenClaw plugin API.
 *
 * Uses the SDK's GatewayRequestHandler contract: each handler receives
 * opts with { req, params, respond, client, context } and calls
 * opts.respond() to send the response. (#2031)
 */
export function registerOAuthGatewayRpcMethods(api, methods) {
    api.registerGatewayMethod('oauth.accounts.list', wrapOAuthMethodHandler(methods.accountsList));
    api.registerGatewayMethod('oauth.contacts.list', wrapOAuthMethodHandler(methods.contactsList));
    api.registerGatewayMethod('oauth.email.list', wrapOAuthMethodHandler(methods.emailList));
    api.registerGatewayMethod('oauth.email.get', wrapOAuthMethodHandler(methods.emailGet));
    api.registerGatewayMethod('oauth.files.list', wrapOAuthMethodHandler(methods.filesList));
    api.registerGatewayMethod('oauth.files.search', wrapOAuthMethodHandler(methods.filesSearch));
    api.registerGatewayMethod('oauth.files.get', wrapOAuthMethodHandler(methods.filesGet));
    api.registerGatewayMethod('oauth.calendar.list', wrapOAuthMethodHandler(methods.calendarList));
    api.registerGatewayMethod('oauth.calendar.sync', wrapOAuthMethodHandler(methods.calendarSync));
    api.registerGatewayMethod('oauth.calendar.create', wrapOAuthMethodHandler(methods.calendarCreate));
}
