// @troykelly/openclaw-projects/dist/gateway/rpc-methods.js
/**
 * Gateway RPC methods for real-time updates.
 * Part of Epic #310, Issue #324.
 *
 * Provides bidirectional communication between OpenClaw agents and
 * the openclaw-projects backend via the Gateway protocol.
 *
 * @see https://docs.openclaw.ai/gateway/protocol.md
 */
/** All valid event types for validation */
const VALID_EVENTS = ['message.new', 'message.delivered', 'task.due', 'task.overdue', 'memory.created'];
/**
 * Validate event types and filter to only valid ones.
 *
 * @param events - Array of event type strings
 * @returns Array of valid NotificationEvent types
 */
function filterValidEvents(events) {
    return events.filter((e) => VALID_EVENTS.includes(e));
}
/**
 * Create Gateway RPC method handlers.
 *
 * @param options - Configuration options
 * @returns Gateway method handlers
 */
export function createGatewayMethods(options) {
    const { logger, apiClient, getAgentId } = options;
    // Local subscription state
    let subscribedEvents = [];
    return {
        /**
         * Subscribe to event notifications.
         *
         * Registers the user to receive notifications for specified event types.
         * Invalid event types are filtered out.
         */
        async subscribe(params) {
            const user_id = getAgentId();
            const validEvents = filterValidEvents(params.events);
            // Update local subscription state
            subscribedEvents = validEvents;
            logger.info('Gateway subscribe', {
                user_id,
                requestedEvents: params.events,
                subscribedEvents: validEvents,
            });
            return {
                subscribed: validEvents,
            };
        },
        /**
         * Unsubscribe from all event notifications.
         */
        async unsubscribe(_params) {
            const user_id = getAgentId();
            const previousEvents = subscribedEvents;
            // Clear local subscription state
            subscribedEvents = [];
            logger.info('Gateway unsubscribe', {
                user_id,
                previousEvents,
            });
            return {
                unsubscribed: true,
            };
        },
        /**
         * Get pending notifications for the user.
         *
         * Fetches notifications from the backend, optionally filtered by timestamp.
         */
        async getNotifications(params) {
            const user_id = getAgentId();
            const { since, limit = 20 } = params;
            logger.debug('Gateway getNotifications', {
                user_id,
                since,
                limit,
            });
            try {
                // Build query parameters
                const queryParams = new URLSearchParams();
                queryParams.set('limit', String(limit));
                if (since) {
                    queryParams.set('since', since);
                }
                const response = await apiClient.get(`/notifications?${queryParams}`, { user_id });
                if (!response.success) {
                    logger.error('Gateway getNotifications API error', {
                        user_id,
                        error: response.error.message,
                    });
                    return {
                        notifications: [],
                        has_more: false,
                    };
                }
                const { notifications, total } = response.data;
                logger.debug('Gateway getNotifications success', {
                    user_id,
                    count: notifications.length,
                    total,
                });
                return {
                    notifications,
                    has_more: notifications.length >= limit && total > notifications.length,
                };
            }
            catch (error) {
                logger.error('Gateway getNotifications failed', {
                    user_id,
                    error: error instanceof Error ? error.message : String(error),
                });
                return {
                    notifications: [],
                    has_more: false,
                };
            }
        },
        /**
         * Get currently subscribed events.
         *
         * Internal method for checking subscription state.
         */
        getSubscribedEvents() {
            return [...subscribedEvents];
        },
    };
}
/**
 * Wrap a typed method handler into a GatewayRequestHandler.
 *
 * The SDK's registerGatewayMethod expects a handler that receives
 * `(opts: GatewayRequestHandlerOptions)` and calls `opts.respond()`.
 * Our internal methods use `(params: T) => Promise<R>` for testability.
 * This wrapper bridges the two contracts. (#2031)
 */
function wrapMethodHandler(handler) {
    return async (opts) => {
        try {
            const result = await handler(opts.params);
            opts.respond(true, result);
        }
        catch (_error) {
            // Return generic error to clients to avoid leaking internal details
            opts.respond(false, undefined, {
                code: 'INTERNAL_ERROR',
                message: 'Internal error',
            });
        }
    };
}
/**
 * Register Gateway RPC methods with the OpenClaw API.
 *
 * Uses the SDK's GatewayRequestHandler contract: each handler receives
 * opts with { req, params, respond, client, context } and calls
 * opts.respond() to send the response. (#2031)
 *
 * @param api - OpenClaw Plugin API
 * @param methods - Gateway method handlers
 */
export function registerGatewayRpcMethods(api, methods) {
    api.registerGatewayMethod('openclaw-projects.subscribe', wrapMethodHandler(methods.subscribe));
    api.registerGatewayMethod('openclaw-projects.unsubscribe', wrapMethodHandler(methods.unsubscribe));
    api.registerGatewayMethod('openclaw-projects.getNotifications', wrapMethodHandler(methods.getNotifications));
}
